{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "7#!/usr/bin/env python\n",
    "\n",
    "\"\"\"\n",
    "This script will download files from an amazon s3 box.\n",
    "First, it will look for any files (json) and download to a\n",
    "json_folder to store all jobs submitted.\n",
    "Then, it will download any raw files specified in the document\n",
    "and create an executable copy of each jobfile in a work_folder.\n",
    "\n",
    "\"\"\"\n",
    "import argparse\n",
    "import xml.etree.ElementTree as et \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "from xml.dom import minidom\n",
    "from collections import defaultdict\n",
    "import yaml\n",
    "\n",
    "from yaml.representer import Representer\n",
    "yaml.add_representer(defaultdict, Representer.represent_dict)\n",
    "\n",
    "import json\n",
    "\n",
    "def get_metadata_from_xml(input_file, transcriptome, module, module_version, module_script):\n",
    "    \"\"\"\n",
    "    Transforms the metadata XML file into a dictionary that can be dumped into 1 or more properly \n",
    "    formatted JSON files.\n",
    "    \n",
    "    input_file: string\n",
    "    transcriptome: string\n",
    "    chemistry: string\n",
    "    cr11: string\n",
    "    module: string\n",
    "    module_version: string\n",
    "    module_script: string\n",
    "    \"\"\"\n",
    "    my_expt_set_metadata = []    \n",
    "\n",
    "    xmldoc = minidom.parse(input_file)\n",
    "    expt_packages = xmldoc.getElementsByTagName('EXPERIMENT_PACKAGE')\n",
    "    for expt_package in expt_packages:\n",
    "        my_expt_metadata = defaultdict(list)\n",
    "        \n",
    "        ### Pipeline-specific. use function input to describe module/pipelines used. ###\n",
    "        my_expt_metadata['module'] = module\n",
    "        my_expt_metadata['module_version'] = module_version\n",
    "        my_expt_metadata['module_script'] = module_script\n",
    "        \n",
    "        ### EXPERIMENTAL METADATA. Describe experiment-level details. Broken up into 1) organization, 2) experiment 3) design 4) study ###\n",
    "        organization_metadata = expt_package.getElementsByTagName('Organization')\n",
    "        for metadata in organization_metadata:\n",
    "            my_expt_metadata['organization'].append(metadata.getElementsByTagName('Name')[0].firstChild.data)\n",
    "            my_expt_metadata['email'].append(metadata.getElementsByTagName('Contact')[0].attributes['email'].value)\n",
    "        my_expt_metadata['organization'] = ','.join(my_expt_metadata['organization'])\n",
    "        my_expt_metadata['email'] = ','.join(my_expt_metadata['email'])\n",
    "\n",
    "        expt_metadata = expt_package.getElementsByTagName('EXPERIMENT')\n",
    "        for metadata in expt_metadata:\n",
    "            title = metadata.getElementsByTagName('TITLE')[0].firstChild.data\n",
    "            my_expt_metadata['experiment_nickname'].append(metadata.attributes['accession'].value)\n",
    "            my_expt_metadata['experiment_summary'].append(title)\n",
    "        my_expt_metadata['experiment_nickname'] = ','.join(my_expt_metadata['experiment_nickname'])\n",
    "        my_expt_metadata['experiment_summary'] = ','.join(my_expt_metadata['experiment_summary'])\n",
    "\n",
    "        design_metadata = expt_package.getElementsByTagName('DESIGN')[0]\n",
    "        my_expt_metadata['library_layout'] = design_metadata.getElementsByTagName('LIBRARY_DESCRIPTOR')[0].getElementsByTagName('LIBRARY_LAYOUT')[0].firstChild.tagName\n",
    "        my_expt_metadata['library_source'] = design_metadata.getElementsByTagName('LIBRARY_DESCRIPTOR')[0].getElementsByTagName('LIBRARY_SOURCE')[0].firstChild.nodeValue\n",
    "        my_expt_metadata['library_description'] = design_metadata.getElementsByTagName('LIBRARY_DESCRIPTOR')[0].getElementsByTagName('LIBRARY_SELECTION')[0].firstChild.nodeValue        \n",
    "        \n",
    "        study_metadata = expt_package.getElementsByTagName('STUDY')[0]\n",
    "        my_expt_metadata['study_title'] = study_metadata.getElementsByTagName('STUDY_TITLE')[0].firstChild.nodeValue\n",
    "        my_expt_metadata['study_abstract'] = study_metadata.getElementsByTagName('STUDY_ABSTRACT')[0].firstChild.nodeValue\n",
    "        \n",
    "        external_ids = study_metadata.getElementsByTagName('EXTERNAL_ID')\n",
    "        for external_id in external_ids:\n",
    "            for attr in external_id.attributes.keys():\n",
    "                ddb = external_id.attributes[attr].value\n",
    "                did = external_id.firstChild.data\n",
    "                my_expt_metadata['database'].append({ddb:did})\n",
    "        study_links = study_metadata.getElementsByTagName('STUDY_LINKS')\n",
    "        \n",
    "        for link in study_links:\n",
    "            sdb = link.getElementsByTagName('XREF_LINK')[0].childNodes[0].firstChild.data\n",
    "            sid = link.getElementsByTagName('XREF_LINK')[0].childNodes[1].firstChild.data\n",
    "            my_expt_metadata['reference'].append({sdb:sid})\n",
    "            \n",
    "        ### RUN METADATA. Describe each individual sequencing run. ###\n",
    "        run_set = expt_package.getElementsByTagName('RUN_SET')\n",
    "        for run in run_set:\n",
    "            my_run_metadata = defaultdict(dict)\n",
    "\n",
    "            run_metadata = run.getElementsByTagName('RUN')\n",
    "            for metadata in run_metadata:\n",
    "                try:\n",
    "                    my_run_metadata['library_nickname'] = metadata.attributes['accession'].value\n",
    "                except KeyError:\n",
    "                    my_run_metadata['library_nickname'] = ''\n",
    "                try:\n",
    "                    my_run_metadata['library_prep'] = \"\"\n",
    "                except KeyError:\n",
    "                    my_run_metadata['library_prep'] = \"\"\n",
    "                try:\n",
    "                    my_run_metadata['sample_id'] = metadata.attributes['accession'].value\n",
    "                except KeyError:\n",
    "                    my_run_metadata['sample_id'] = \"\"\n",
    "                try:\n",
    "                    my_run_metadata['original_assembly'] = metadata.attributes['assembly'].value\n",
    "                except KeyError:\n",
    "                    my_run_metadata['original_assembly'] = \"\"\n",
    "\n",
    "            ### Comment out for now. Supposedly NCBI will move to AWS and will eventually have s3 links available. ###\n",
    "            # file_metadata = run.getElementsByTagName('SRAFile')\n",
    "            # for metadata in file_metadata:\n",
    "            #     if 'url' in metadata.attributes:\n",
    "            #         print(metadata.attributes['url'].value)\n",
    "                # filename = metadata.attributes['filename'].value\n",
    "                # url = metadata.attributes['url'].value\n",
    "                # print(url)\n",
    "\n",
    "            instrument_metadata = expt_package.getElementsByTagName('PLATFORM')\n",
    "            for metadata in instrument_metadata:\n",
    "                try:\n",
    "                    instrument = metadata.getElementsByTagName('INSTRUMENT_MODEL')[0].firstChild.nodeValue\n",
    "                except AttributeError:\n",
    "                    instrument = ''\n",
    "                my_run_metadata['instrument_model'] = instrument\n",
    "                my_run_metadata['transcriptome'] = {\n",
    "                    'class': 'Directory',\n",
    "                    'path': transcriptome\n",
    "                }\n",
    "\n",
    "            # This actually belong on the top level per-expt, but we apply characteristics to each file.\n",
    "            taxonomy_metadata = expt_package.getElementsByTagName('SAMPLE_NAME')\n",
    "            for metadata in taxonomy_metadata:\n",
    "                scientific_name = metadata.getElementsByTagName('SCIENTIFIC_NAME')[0].firstChild.data\n",
    "                my_expt_metadata['organism'] = scientific_name\n",
    "\n",
    "            sample_set = expt_package.getElementsByTagName('SAMPLE')\n",
    "            for sample in sample_set:\n",
    "                sample_metadata = sample.getElementsByTagName('SAMPLE_ATTRIBUTES')\n",
    "                characteristics_list = []\n",
    "                for metadata in sample_metadata:\n",
    "                    attributes = metadata.getElementsByTagName('SAMPLE_ATTRIBUTE')\n",
    "                    my_sample_metadata = defaultdict()\n",
    "                    for attribute in attributes:\n",
    "                        characteristics_tag = {\n",
    "                            'name': attribute.getElementsByTagName('TAG')[0].firstChild.data,\n",
    "                            'value': attribute.getElementsByTagName('VALUE')[0].firstChild.data\n",
    "                        }\n",
    "                        characteristics_list.append(characteristics_tag)\n",
    "            my_run_metadata['characteristics'] = characteristics_list\n",
    "            my_expt_metadata['samples'].append(my_run_metadata)\n",
    "\n",
    "        my_expt_set_metadata.append(my_expt_metadata)\n",
    "    return my_expt_set_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[defaultdict(list,\n",
       "             {'module': 'rnaseq',\n",
       "              'module_version': '0.0.3',\n",
       "              'module_script': 'rnaseq/0.0.3',\n",
       "              'organization': 'NCBI',\n",
       "              'email': 'geo-group@ncbi.nlm.nih.gov',\n",
       "              'experiment_nickname': 'SRX4004898',\n",
       "              'experiment_summary': 'GSM3119650: PSA-NCAM+ GFP- rep 2; Homo sapiens; RNA-Seq',\n",
       "              'library_layout': 'PAIRED',\n",
       "              'library_source': 'TRANSCRIPTOMIC',\n",
       "              'library_description': 'cDNA',\n",
       "              'study_title': 'Trans-differentiation of human adult peripheral blood T cells into neurons',\n",
       "              'study_abstract': 'Examining the transcriptomic changes during transdifferentiation of peripheral blood mononuclear cells to induced neuronal cells. Overall design: There are three different populations: PBMC (2 biological replicates, starting population), PSA-NCAM+GFP+ (2 biological replicates, induced neuronal cells) and PSA-NCAM+GFP- (2 biological replicates, induced neuronal cells).',\n",
       "              'database': [{'BioProject': 'PRJNA454016'},\n",
       "               {'primary': 'PRJNA454016'},\n",
       "               {'GEO': 'GSE113804'}],\n",
       "              'reference': [{'pubmed': '29866841'}],\n",
       "              'organism': 'Homo sapiens',\n",
       "              'samples': [defaultdict(dict,\n",
       "                           {'library_nickname': 'SRR7075128',\n",
       "                            'sample_id': 'SRR7075128',\n",
       "                            'instrument_model': 'NextSeq 500',\n",
       "                            'transcriptome': {'class': 'Directory',\n",
       "                             'path': 'hg19.transcriptome'},\n",
       "                            'characteristics': [{'name': 'source_name',\n",
       "                              'value': 'In vitro derived neuronal cells'},\n",
       "                             {'name': 'cell type',\n",
       "                              'value': 'induced neuronal cells'}]})]})]"
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_file = '/home/bay001/projects/codebase/bfx/notebooks/pilot_notebooks/SraExperimentPackage.xml'\n",
    "\n",
    "get_metadata_from_xml(input_file, transcriptome='hg19.transcriptome', module='rnaseq', module_version='0.0.3', module_script='rnaseq/0.0.3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = get_metadata_from_xml(\n",
    "    input_file=input_file, \n",
    "    transcriptome=transcriptome, \n",
    "    chemistry=chemistry, \n",
    "    cr11=cr11,\n",
    "    module=module,\n",
    "    module_version=module_version,\n",
    "    module_script=module_script\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3-essential-tscc",
   "language": "python",
   "name": "python3essentialtscc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
