{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is packaged in junction-counter\n",
    "https://github.com/byee4/junction-counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "import pysam\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(96, 12)"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_jxn_string(row):\n",
    "    return \"{}:{}-{}:{}\".format(\n",
    "        row['chrom'], row['low_exon_end'], int(row['low_exon_end'])+1, row['strand']\n",
    "    )\n",
    "\n",
    "test_file = '/home/elvannostrand/data/ENCODE/RNAseq/scripts/exon_junction_counts/test2'\n",
    "test_header = ['position','num_spliced_reads','num_inclusion_reads','dpsi']\n",
    "test_df = pd.read_table(test_file, names=test_header)\n",
    "test_df['chrom'] = test_df['position'].apply(lambda x: x.split(':')[0])\n",
    "test_df['strand'] = test_df['position'].apply(lambda x: x.split(':')[1])\n",
    "test_df['low_exon_start'] = test_df['position'].apply(lambda x: x.split(':')[2].split('|')[0].split('-')[0])\n",
    "test_df['low_exon_end'] = test_df['position'].apply(lambda x: x.split(':')[2].split('|')[0].split('-')[1])\n",
    "test_df['hi_exon_start'] = test_df['position'].apply(lambda x: x.split(':')[2].split('|')[1].split('-')[0])\n",
    "test_df['hi_exon_end'] = test_df['position'].apply(lambda x: x.split(':')[2].split('|')[1].split('-')[1])\n",
    "x = test_df.sort_values(by='num_spliced_reads').reset_index()\n",
    "# x = x[x['num_spliced_reads']!=0]\n",
    "# x = x[x['low_exon_end']=='57135600']\n",
    "# x = x.head(25)\n",
    "x['jxn_string'] = x.apply(get_jxn_string, axis=1)\n",
    "x.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_offset_m_basedon_n(\n",
    "    cigartuples, n, \n",
    "    include_jxn_span=True, \n",
    "    include_insertions=False, \n",
    "    include_deletions=True, \n",
    "    verbose=False\n",
    "):\n",
    "    \"\"\"\n",
    "    for junction # n (0 based), return the left and right offsets m.\n",
    "    So for example, if we have a cigar string such as 10M100N15M:\n",
    "    this function will return:\n",
    "    (if no flags): 10, 15\n",
    "    (if include_jxn_span): 110, 115\n",
    "    (if include_insertions): \n",
    "    \"\"\"\n",
    "    all_counter = 0\n",
    "    counter = 0\n",
    "    left_accumulated_m = 0\n",
    "    current_left_offset = 0\n",
    "    current_right_offset = 0\n",
    "    for t in cigartuples:\n",
    "        if verbose:\n",
    "            print(t)\n",
    "        if t[0] == 0:\n",
    "            left_accumulated_m += t[1]\n",
    "        elif t[0] == 1 and include_insertions == True: # insertion code\n",
    "            left_accumulated_m += t[1]\n",
    "        elif t[0] == 2 and include_deletions == True:\n",
    "            left_accumulated_m += t[1]\n",
    "        elif t[0] == 3:\n",
    "            if include_jxn_span:\n",
    "                current_left_offset = left_accumulated_m + t[1]\n",
    "            else:\n",
    "                current_left_offset = left_accumulated_m\n",
    "            counter += 1\n",
    "        \n",
    "        if counter >= n:\n",
    "            if verbose:\n",
    "                print(\"ALL COUNTER\", all_counter)\n",
    "            for tr in cigartuples[all_counter:]:\n",
    "                if verbose:\n",
    "                    print(\"TR\", tr)\n",
    "                if tr[0] == 0:\n",
    "                    current_right_offset += tr[1]\n",
    "                elif tr[0] == 1 and include_insertions == True:\n",
    "                    current_right_offset += tr[1]\n",
    "                elif tr[0] == 2: #  and include_deletions == True:\n",
    "                    current_right_offset += tr[1]\n",
    "                elif tr[0] == 3 and include_jxn_span == True:\n",
    "                    current_right_offset += tr[1]\n",
    "                \n",
    "            return [current_left_offset, current_right_offset]\n",
    "        all_counter += 1\n",
    "    return [current_left_offset, current_right_offset]\n",
    "\n",
    "def get_jxc_counts_from_cigartuples(cigartuples):\n",
    "    \"\"\"\n",
    "    From a list of cigar tuples, return the number of junctions found (N's)\n",
    "    \"\"\"\n",
    "    count = 0\n",
    "    for t in cigartuples:\n",
    "        if t[0] == 3:\n",
    "            count += 1\n",
    "    return count \n",
    "\n",
    "def parse_jxn_string(jxn_string):\n",
    "    \"\"\"\n",
    "    Parses a line like: chr:start-stop:strand to return\n",
    "    these proper fields.\n",
    "    \"\"\"\n",
    "    chrom, pos, strand = jxn_string.split(':')\n",
    "    start, stop = pos.split('-')\n",
    "    return chrom, int(start), int(stop), strand\n",
    "\n",
    "def return_spliced_junction_counts(jxn_string, bam_file, min_overlap = 0):\n",
    "    aligned_file = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "    chrom, start, stop, strand = parse_jxn_string(jxn_string)\n",
    "    \n",
    "    if start + 1 != stop:\n",
    "        print(\"WARNING: junction site is greater than 1 nt: {}\".format(jxn_string))\n",
    "        \n",
    "    depth=0  # initial number of reads supporting junction\n",
    "    names = ''  # string of 'comma delimited' names of all reads supporting junction\n",
    "    badflags = []  # flags of all the skipped reads, if failed QC (for debugging mostly)\n",
    "    \n",
    "    # jxn = '{}:{}-{}:{}'.format(row['chrom'],row['low_exon_end'],int(row['low_exon_end'])+1, row['strand'])\n",
    "    for read in aligned_file.fetch(chrom, start, stop): # get the reads which span this position\n",
    "        ### cigar_tuple codes:\n",
    "        # 0: match\n",
    "        # 3: skip\n",
    "\n",
    "        skip = False  # initially treat this read as supporting jxn\n",
    "        left_span = False  # assume that this read does not support any junction yet\n",
    "        right_span = False  # assume that this read does not support any junction yet\n",
    "\n",
    "        jxc_count = 0  # number of junctions in the read\n",
    "        if 'N' in read.cigarstring:  # only consider reads with jxns\n",
    "            \"\"\"\n",
    "            if read.is_reverse and read.is_read2:\n",
    "                if start==118838559:\n",
    "                    print('read2 and opposite', read.query_name)\n",
    "                skip = True\n",
    "            if not read.is_reverse and read.is_read1:\n",
    "                if start==118838559:\n",
    "                    print('read1 and same', read.query_name)\n",
    "                skip = True\n",
    "                \"\"\"\n",
    "            if (not read.is_proper_pair) or read.is_qcfail or read.is_secondary:\n",
    "                skip = True\n",
    "                badflags.append(read.flag)\n",
    "\n",
    "            jxc_count = get_jxc_counts_from_cigartuples(read.cigartuples)\n",
    "            ### \n",
    "            # Look at each junction that this read overlaps and see if any of them match\n",
    "            # the one in focus.\n",
    "            ###\n",
    "            for j in range(1, jxc_count+1):\n",
    "                left, right = get_offset_m_basedon_n(read.cigartuples, j, True, True, True)\n",
    "                \n",
    "                left_wo, right_wo = get_offset_m_basedon_n(\n",
    "                    read.cigartuples, j, False, False, False\n",
    "                )\n",
    "                if left_wo < min_overlap or right_wo < min_overlap:\n",
    "                    skip = True\n",
    "                # if read.query_name == 'D80KHJN1:241:C5HF7ACXX:6:1105:6615:43587':\n",
    "                #     print(\"RIGHT\", read.reference_end, right, int(row['low_exon_end']), read.flag, read.cigartuples)\n",
    "                #     print(\"LEFT\", read.reference_start, read.query_alignment_start, left, int(row['low_exon_end']), read.flag)\n",
    "                ### need to filter out read1 whose: 1) same strand \n",
    "                # left span refers to this junction 'L', while right_span refers to R\n",
    "                #   [L======R]-----[L=========R]----...\n",
    "                if read.reference_start + read.query_alignment_start + left == start:\n",
    "                    left_span = True # read supports a left jxc\n",
    "                if read.reference_end - right == start:\n",
    "                    right_span = True # read supports a right jxc\n",
    "                    \n",
    "            if left_span == False and right_span == False:\n",
    "                skip = True\n",
    "            if not skip:\n",
    "                names += read.query_name + ','\n",
    "                depth+=1\n",
    "    return depth, names[:-1]\n",
    "\n",
    "def get_junction_sites(jxn_list, bam_file, min_overlap):\n",
    "    jxn_dict = defaultdict(dict)\n",
    "    \n",
    "    for jxn in jxn_list:\n",
    "        depth, names = return_spliced_junction_counts(jxn, bam_file, min_overlap)\n",
    "        jxn_dict[jxn] = {'depth':depth, 'names':names}\n",
    "    return pd.DataFrame(jxn_dict).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "jxn_list = []\n",
    "bam_file = '/home/bay001/projects/encode/analysis/tests/eric_jxc_tests/ENCFF756RDZ.bam'\n",
    "o = open('/home/bay001/projects/codebase/junction-counter/data/jxnlist.txt', 'w')\n",
    "for jxn in x['jxn_string']:\n",
    "    jxn_list.append(jxn)\n",
    "    o.write('{}\\n'.format(jxn))\n",
    "o.close()\n",
    "get_junction_sites(jxn_list, bam_file, min_overlap=10).to_csv('/home/bay001/projects/encode/analysis/tests/eric_jxc_tests/ENCFF756RDZ.bam.jxc', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[67, 165]"
      ]
     },
     "execution_count": 287,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test_t = [(0, 1), (3, 169), (1, 2), (0, 10), (0, 92), (3, 1149), (0, 7)]\n",
    "test_t = [(0, 1), (3, 66), (0, 99)]\n",
    "\n",
    "def get_offset_m_basedon_n(cigartuples, n, include_jxn_span=True):\n",
    "    \"\"\"\n",
    "    for junction # n (0 based), return the left and right offsets m\n",
    "    \"\"\"\n",
    "    all_counter = 0\n",
    "    counter = 0\n",
    "    left_accumulated_m = 0\n",
    "    current_left_offset = 0\n",
    "    current_right_offset = 0\n",
    "    \n",
    "    for t in cigartuples:\n",
    "        if t[0] == 0:\n",
    "            left_accumulated_m += t[1]\n",
    "        elif t[0] == 3:\n",
    "            if include_jxn_span:\n",
    "                current_left_offset = left_accumulated_m + t[1]\n",
    "            else:\n",
    "                current_left_offset = left_accumulated_m\n",
    "            counter += 1\n",
    "        \n",
    "        \n",
    "        if counter >= n:\n",
    "            for tr in cigartuples[all_counter:]:\n",
    "                if tr[0] == 0:\n",
    "                    current_right_offset += tr[1]\n",
    "                elif tr[0] == 3 and include_jxn_span == True:\n",
    "                    current_right_offset += tr[1]\n",
    "                    \n",
    "            return [current_left_offset, current_right_offset]\n",
    "        all_counter += 1\n",
    "    return [current_left_offset, current_right_offset]\n",
    "    \n",
    "get_offset_m_basedon_n(test_t, 1)\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>rgb</th>\n",
       "      <th>block_count</th>\n",
       "      <th>block_sizes</th>\n",
       "      <th>block_locations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, stop, name, score, strand, thickStart, thickEnd, rgb, block_count, block_sizes, block_locations]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Read name = D80KHJN1:241:C5HF7ACXX:6:1105:6615:43587\n",
    "Read length = 100bp\n",
    "----------------------\n",
    "Mapping = Primary @ MAPQ 50\n",
    "Reference span = chr3:128,525,316-128,526,490 (+) = 1,175bp\n",
    "Cigar = 3M1075N97M\n",
    "Clipping = None\n",
    "----------------------\n",
    "Location = chr3:128,525,318\n",
    "Base = C @ QV 31\n",
    "----------------------\n",
    "Mate is mapped = yes\n",
    "Mate start = chr3:128532171 (-)\n",
    "Insert size = 6956\n",
    "Second in pair\n",
    "Pair orientation = F2R1\n",
    "----------------------\n",
    "XG = 0\n",
    "NH = 1\n",
    "NM = 0\n",
    "XM = 0\n",
    "XN = 0\n",
    "XO = 0\n",
    "AS = 0\n",
    "XS = +\n",
    "YT = UU\n",
    "-------------------\n",
    "Alignment start position = chr3:128525316\n",
    "CACAAAGCGGGCACAGGCCTGGTGCTACAGCAAAAACAACATTCCCTACTTTGAGACCAGTGCCAAGGAGGCCATCAACGTGGAGCAGGCGTTCCAGACG\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
