{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from rnaseq import rmats_inclevel_analysis as rmats\n",
    "from encode import manifest_helpers\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "import pybedtools\n",
    "\n",
    "pd.set_option('display.max_columns', 500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exprs = '/home/elvannostrand/data/clip/CLIPseq_analysis/RNAseq_expression/HepG2K562wnoncoding_ensg_onecelltypeonly.txt'\n",
    "wd = '/home/bay001/projects/encode/analysis/cell_specific_overlaps'\n",
    "splice_dir = '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check to see if there are no overlaps between U2AF1 K562 and HepG2 splice sites. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hepg2 = '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF1-BGHLV30-HepG2-SE.MATS.JunctionCountOnly.significant.txt'\n",
    "# k562 = '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF1-LV08-K562-SE.MATS.JunctionCountOnly.significant.txt'\n",
    "hepg2 = '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF2-BGHLV26-HepG2-SE.MATS.JunctionCountOnly.significant.txt'\n",
    "k562 = '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF2-LV08-K562-SE.MATS.JunctionCountOnly.significant.txt'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "\n",
    "def rmats2bedtool_all(annotations, exon_start, exon_end):\n",
    "    \"\"\"\n",
    "    Returns merged intervals that contain all regions from all \n",
    "    junction counts files. This ensures that regions will be non-overlapping\n",
    "    and can be used as 'keys' to regions on the epigenome browser.\n",
    "    \"\"\"\n",
    "    intervals = []\n",
    "    # progress = tnrange(len(annotations))\n",
    "    for pos in annotations:\n",
    "        nice_name = os.path.basename(pos)\n",
    "        nice_name = nice_name.replace('-SE.MATS.JunctionCountOnly','')\n",
    "        nice_name = nice_name.replace('nr.txt','')\n",
    "        # nice_name = os.path.basename(pos).replace('-SE.MATS.JunctionCountOnly.positive.nr.txt','')\n",
    "        # nice_name = nice_name.replace('-SE.MATS.JunctionCountOnly.negative.nr.txt','')\n",
    "        \n",
    "        df = pd.read_table(pos)\n",
    "        for col, row in df.iterrows():\n",
    "            intervals.append(\n",
    "                pybedtools.create_interval_from_list(\n",
    "                    [row['chr'], str(row[exon_start]), str(row[exon_end]), nice_name, str(row['IncLevelDifference']), row['strand']])\n",
    "            )\n",
    "        # progress.update(1)\n",
    "    bedtool_all_intervals = pybedtools.BedTool(intervals)\n",
    "    return bedtool_all_intervals.sort().merge()\n",
    "\n",
    "def rmats2bedtool(annotation, exon_start, exon_end):\n",
    "    \"\"\"\n",
    "    Returns a bedtool object from an rmats annotation\n",
    "    Uses the exonStart_0base and exonEnd as coordinates.\n",
    "    \"\"\"\n",
    "    intervals = []\n",
    "    df = pd.read_table(annotation)\n",
    "    # progress = tnrange(df.shape[0], leave=False)\n",
    "    for _, row in df.iterrows():\n",
    "        intervals.append(\n",
    "            pybedtools.create_interval_from_list(\n",
    "                [\n",
    "                    row['chr'], str(row[exon_start]), \n",
    "                    str(row[exon_end]), row['GeneID'], \n",
    "                    str(row['IncLevelDifference']), row['strand']\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        # progress.update(1)\n",
    "    \n",
    "    return pybedtools.BedTool(intervals)\n",
    "\n",
    "def transform_individual_rmats_positions(rmats_file, big_merged_bedtool, es, ee):\n",
    "    \"\"\"\n",
    "    turns individual rmats exon positions into something \n",
    "    common to those found in the big merged bedtool file.\n",
    "    \n",
    "    Parameters:\n",
    "    rmats_file : string\n",
    "        rmats JunctionCountsOnly.txt\n",
    "    big_merged_bedtool : pybedtools.BedTool\n",
    "        intervals containing all regions for all annotations being compared.\n",
    "    \n",
    "    Returns: dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    # just get the \"nice name\" (RBP_CELL without the extra stuff)\n",
    "    nice_name = os.path.basename(rmats_file)\n",
    "    \n",
    "    # nice_name = nice_name.replace('-SE.MATS.JunctionCountOnly..nr.txt','')\n",
    "    # nice_name = nice_name.replace('-SE.MATS.JunctionCountOnly.negative.nr.txt','')\n",
    "    nice_name = nice_name.split('-')\n",
    "    nice_name = '{}_{}'.format(nice_name[0], nice_name[2])\n",
    "    \n",
    "    # for each rmats file, intersect with the merged bedtool to bin regions into those that are common amongst all\n",
    "    individual_rmats_bedtool = rmats2bedtool(rmats_file, es, ee).sort()\n",
    "    intersected = individual_rmats_bedtool.intersect(big_merged_bedtool, wb=True).to_dataframe()\n",
    "    \n",
    "    # thickStart, thickEnd, itemRgb actually contain the 'key' common regions from the big_merged_bedtool intersection.\n",
    "    # intersected['chrom'] = intersected['thickStart']\n",
    "    # intersected['start'] = intersected['thickEnd']\n",
    "    # intersected['end'] = intersected['itemRgb']\n",
    "    \n",
    "    # re-format so that it's a proper dataframe, and re-name the 'score' column to be that of the name of the RBP. \n",
    "    # intersected = intersected[['chrom','start','end','name','score','strand']]\n",
    "    intersected.columns = ['chrom','start','end','name','{}'.format(nice_name),'strand', \n",
    "                           'original_chr_{}'.format(nice_name), 'original_start_{}'.format(nice_name), 'original_end_{}'.format(nice_name)]\n",
    "    intersected.set_index(['chrom','start','end','name','strand'], inplace=True)\n",
    "    return intersected\n",
    "\n",
    "def merge_all_rmats_transformed(all_annotations, big_merged_bedtool):\n",
    "    \"\"\"\n",
    "    merge all dpsi for common regions (as described in big_merged_bedtool) into one dataframe. \n",
    "    \"\"\"\n",
    "    progress = tnrange(len(all_annotations))\n",
    "    \n",
    "    # do this once to easily/automatically populate the index.\n",
    "    merged = transform_individual_rmats_positions(all_annotations[0], big_merged_bedtool)\n",
    "    progress.update(1)\n",
    "    \n",
    "    # foreach subsequent file, merge (outer join to not miss any) files into merged.\n",
    "    for annotation in all_annotations[1:]:\n",
    "        df = transform_individual_rmats_positions(annotation, big_merged_bedtool, es, ee)\n",
    "        merged = pd.merge(merged, df, how='outer', left_index=True, right_index=True)\n",
    "        progress.update(1)\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bay001/anaconda2/envs/brian/lib/python2.7/site-packages/ipykernel_launcher.py:3: DeprecationWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate_ix\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4068f08f12164effa215144a9192ae22"
      }
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ((21, 4), (133, 4)) ((8, 4), (123, 4)) ((30, 4), (119, 4)) ('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/XRCC6-BGKLV13-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/XRCC6-BGHLV14-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "((46, 4), (3, 4)) ((14, 4), (148, 4)) ('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF2-LV08-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/U2AF2-BGHLV26-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "((4, 4), (116, 4)) ((85, 4), (35, 4)) ('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/HNRNPU-BGKLV21-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/HNRNPU-BGHLV20-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "((24, 4), (55, 4)) ('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/HNRNPC-BGKLV21-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/HNRNPC-BGHLV20-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "((14, 4), (116, 4)) ('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/SF3B4-LV08-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/SF3B4-BGHLV12-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "('/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/SF3B1-BGKLV36-K562-SE.MATS.JunctionCountOnly.significant.txt', '/projects/ps-yeolab3/bay001/maps/current_normed_annotations/se/SF3B1-BGHLV35-HepG2-SE.MATS.JunctionCountOnly.significant.txt')\n",
      "((26, 4), (32, 4)) ((8, 4), (49, 4))\n"
     ]
    }
   ],
   "source": [
    "rbps = []\n",
    "tpm = pd.read_table(exprs)\n",
    "tpm.ix[0] = ['ENSG00000117308.10','ENSG00000117308.10',1.5,'ENSG00000117308.10',1.5]\n",
    "splice_files = glob.glob(os.path.join(splice_dir,'*.significant.txt'))\n",
    "for splice in splice_files:\n",
    "    rbps.append(os.path.basename(splice).split('-')[0])\n",
    "rbps = set(rbps)\n",
    "progress = tnrange(len(rbps))\n",
    "for rbp in rbps:\n",
    "    hepg2 = glob.glob(os.path.join(splice_dir,'{}-*-{}*significant.txt'.format(rbp,'HepG2')))\n",
    "    k562 = glob.glob(os.path.join(splice_dir,'{}-*-{}*significant.txt'.format(rbp,'K562')))\n",
    "    if(len(hepg2)==1) and (len(k562)==1):\n",
    "        hepg2 = hepg2[0]\n",
    "        k562 = k562[0]\n",
    "        annotations = [k562, hepg2]\n",
    "        \"\"\"big_bedtool = rmats2bedtool_all(annotations, 'upstreamES', 'downstreamEE')\n",
    "        x = transform_individual_rmats_positions(k562, big_bedtool,  'upstreamES', 'downstreamEE')\n",
    "        y = transform_individual_rmats_positions(hepg2, big_bedtool,  'upstreamES', 'downstreamEE')\"\"\"\n",
    "        big_bedtool = rmats2bedtool_all(annotations, 'exonStart_0base', 'exonEnd')\n",
    "        x = transform_individual_rmats_positions(k562, big_bedtool,  'exonStart_0base', 'exonEnd')\n",
    "        y = transform_individual_rmats_positions(hepg2, big_bedtool,  'exonStart_0base', 'exonEnd')\n",
    "        try:\n",
    "            merged = pd.merge(x, y, how='inner', left_index=True, right_index=True).reset_index()\n",
    "            result = pd.merge(merged, tpm, how='inner', left_on='name', right_on='ensg')\n",
    "            if result.shape[0] > 0:\n",
    "                print(k562, hepg2)\n",
    "        except TypeError:\n",
    "            print(x.shape, y.shape), \n",
    "    if len(hepg2) > 1:\n",
    "        pass\n",
    "    if len(k562) > 1:\n",
    "        pass\n",
    "    progress.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>U2AF2_K562</th>\n",
       "      <th>original_chr_U2AF2_K562</th>\n",
       "      <th>original_start_U2AF2_K562</th>\n",
       "      <th>original_end_U2AF2_K562</th>\n",
       "      <th>U2AF2_HepG2</th>\n",
       "      <th>original_chr_U2AF2_HepG2</th>\n",
       "      <th>original_start_U2AF2_HepG2</th>\n",
       "      <th>original_end_U2AF2_HepG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1190284</td>\n",
       "      <td>1192510</td>\n",
       "      <td>ENSG00000160087.16</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.119</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1189293</td>\n",
       "      <td>1192510</td>\n",
       "      <td>-0.128</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1189293</td>\n",
       "      <td>1192510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1653034</td>\n",
       "      <td>1654270</td>\n",
       "      <td>ENSG00000268575.1</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.120</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1653034</td>\n",
       "      <td>1654270</td>\n",
       "      <td>-0.146</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1653034</td>\n",
       "      <td>1654270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24112164</td>\n",
       "      <td>24114722</td>\n",
       "      <td>ENSG00000057757.5</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.169</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24112164</td>\n",
       "      <td>24114722</td>\n",
       "      <td>-0.496</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24112164</td>\n",
       "      <td>24114722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24122640</td>\n",
       "      <td>24123272</td>\n",
       "      <td>ENSG00000117308.10</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.147</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24122640</td>\n",
       "      <td>24123272</td>\n",
       "      <td>-0.289</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24122640</td>\n",
       "      <td>24123272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>24973157</td>\n",
       "      <td>24975520</td>\n",
       "      <td>ENSG00000133226.12</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.097</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24973157</td>\n",
       "      <td>24975520</td>\n",
       "      <td>-0.206</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24973157</td>\n",
       "      <td>24975520</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end                name strand  U2AF2_K562  \\\n",
       "0  chr1   1190284   1192510  ENSG00000160087.16      -      -0.119   \n",
       "1  chr1   1653034   1654270   ENSG00000268575.1      -      -0.120   \n",
       "2  chr1  24112164  24114722   ENSG00000057757.5      +      -0.169   \n",
       "3  chr1  24122640  24123272  ENSG00000117308.10      -      -0.147   \n",
       "4  chr1  24973157  24975520  ENSG00000133226.12      +      -0.097   \n",
       "\n",
       "  original_chr_U2AF2_K562  original_start_U2AF2_K562  original_end_U2AF2_K562  \\\n",
       "0                    chr1                    1189293                  1192510   \n",
       "1                    chr1                    1653034                  1654270   \n",
       "2                    chr1                   24112164                 24114722   \n",
       "3                    chr1                   24122640                 24123272   \n",
       "4                    chr1                   24973157                 24975520   \n",
       "\n",
       "   U2AF2_HepG2 original_chr_U2AF2_HepG2  original_start_U2AF2_HepG2  \\\n",
       "0       -0.128                     chr1                     1189293   \n",
       "1       -0.146                     chr1                     1653034   \n",
       "2       -0.496                     chr1                    24112164   \n",
       "3       -0.289                     chr1                    24122640   \n",
       "4       -0.206                     chr1                    24973157   \n",
       "\n",
       "   original_end_U2AF2_HepG2  \n",
       "0                   1192510  \n",
       "1                   1654270  \n",
       "2                  24114722  \n",
       "3                  24123272  \n",
       "4                  24975520  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2054, 5)\n"
     ]
    }
   ],
   "source": [
    "print(tpm.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>strand</th>\n",
       "      <th>U2AF2_K562</th>\n",
       "      <th>original_chr_U2AF2_K562</th>\n",
       "      <th>original_start_U2AF2_K562</th>\n",
       "      <th>original_end_U2AF2_K562</th>\n",
       "      <th>U2AF2_HepG2</th>\n",
       "      <th>original_chr_U2AF2_HepG2</th>\n",
       "      <th>original_start_U2AF2_HepG2</th>\n",
       "      <th>original_end_U2AF2_HepG2</th>\n",
       "      <th>ensg</th>\n",
       "      <th>K562_topENST</th>\n",
       "      <th>K562_TPM</th>\n",
       "      <th>HepG2_topENST</th>\n",
       "      <th>HepG2_TPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, name, strand, U2AF2_K562, original_chr_U2AF2_K562, original_start_U2AF2_K562, original_end_U2AF2_K562, U2AF2_HepG2, original_chr_U2AF2_HepG2, original_start_U2AF2_HepG2, original_end_U2AF2_HepG2, ensg, K562_topENST, K562_TPM, HepG2_topENST, HepG2_TPM]\n",
       "Index: []"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>K562_topENST</th>\n",
       "      <th>K562_TPM</th>\n",
       "      <th>HepG2_topENST</th>\n",
       "      <th>HepG2_TPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ensg, K562_topENST, K562_TPM, HepG2_topENST, HepG2_TPM]\n",
       "Index: []"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm[tpm['ensg']=='ENSG00000057757.5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
