{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Small cookbook for parsing fastq files that I've needed to use more than once.\n",
    "- just a quick reference to share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hsa-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "hsa-let-7g-5p\n",
      "UGAGGUAGUAGUUUGUACAGUU\n",
      "mmu-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "rno-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "sme-let-7b-5p\n",
      "UGAGGUAGAUUGUUGGAUGACU\n",
      "oan-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "cin-let-7b-5p\n",
      "UGAGGUAGUAGGUUAUGUUGUU\n",
      "mml-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "tgu-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "pma-let-7b-5p\n",
      "UGAGGUAGUAGGUUUUGUAGUU\n",
      "aca-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGU\n",
      "pol-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "ssa-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "chi-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n",
      "oha-let-7b-5p\n",
      "UGAGGUAGUAGGUUGUGUGGUU\n"
     ]
    }
   ],
   "source": [
    "\n",
    "### This parses a fasta file and searches for a particular set of ids (names_to_search), and printing\n",
    "\n",
    "names_to_search = ['hsa-let-7g-5p', 'let-7b-5p']\n",
    "from Bio import SeqIO\n",
    "handle = open(\"/projects/ps-yeolab/genomes/mirbase/release_21/mature.fa\", \"rU\")\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    for name in names_to_search:\n",
    "        if name in record.id:\n",
    "            print(record.id)\n",
    "            print(record.seq)\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] [2] [3] [4] [5] [6] [7] [8] [9] [10] [11] [12] [13] [14] [15] [16] [17] [18] [19] [20] [21] [22] [23] [24] [25] [26] [27] [28] [29] [30] [31] [32] [33] [34] [35] [36] [37] [38] [39] [40] [41] [42] [43] [44] [45] [46] [47] [48] [49] [50] [51] [52] [53] [54] [55] [56] [57] [58] [59] [60] [61] [62] [63] [64] [65] [66] [67] [68] [69] [70] [71] [72] [73] [74] [75] [76] [77] [78] [79] [80] [81] [82] [83] [84] [85] [86] [87] [88] [89] [90] [91] [92] [93] [94] [95] [96] [97] [98] [99] [100] [101] [102] [103] [104] [105] [106] [107] [108] [109] [110] [111] [112] [113]\n"
     ]
    }
   ],
   "source": [
    "### This parses a fasta file and searches for a particular set of ids (names_to_search), and writing to a separate file\n",
    "\n",
    "from Bio import SeqIO\n",
    "names_to_search = ['rna', 'RNA', 'alu']\n",
    "handle = open(\"/projects/ps-yeolab/genomes/RepBase18.05.fasta/all.ref\", \"rU\")\n",
    "outfile = \"/home/bay001/projects/encode/analysis/tests/eclip_tests/small_repelements/small_repelements.fa\"\n",
    "records = []\n",
    "x = 0\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    for name in names_to_search:\n",
    "        if name in record.id:\n",
    "            x = x + 1\n",
    "            print('[{}]'.format(x)),\n",
    "            records.append(record)\n",
    "SeqIO.write(records,outfile,\"fasta\")\n",
    "handle.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aae-bantam-3p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-bantam-5p</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-let-7</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-10</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-100</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1000</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-11-3p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-11-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1174</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1175-3p</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1175-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-12-3p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-12-5p</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-124</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-125-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-13-3p</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-13-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-133</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-137</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-14</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-184</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1889-3p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1889-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1890</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-1891</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-190</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-193</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-210</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aae-miR-219</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399c-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399d-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399d-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399e-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399e-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399f-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399f-5p</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399g-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399g-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399h-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399h-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399i-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399i-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399j-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR399j-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR408a</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR408b-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR408b-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR444a</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR444b</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR482-3p</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR482-5p</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR528a-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR528a-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR528b-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR528b-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR529-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR529-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR827-3p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma-miR827-5p</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35828 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 len\n",
       "aae-bantam-3p     22\n",
       "aae-bantam-5p     23\n",
       "aae-let-7         21\n",
       "aae-miR-1         22\n",
       "aae-miR-10        22\n",
       "aae-miR-100       22\n",
       "aae-miR-1000      21\n",
       "aae-miR-11-3p     22\n",
       "aae-miR-11-5p     22\n",
       "aae-miR-1174      21\n",
       "aae-miR-1175-3p   24\n",
       "aae-miR-1175-5p   22\n",
       "aae-miR-12-3p     22\n",
       "aae-miR-12-5p     23\n",
       "aae-miR-124       19\n",
       "aae-miR-125-5p    22\n",
       "aae-miR-13-3p     23\n",
       "aae-miR-13-5p     22\n",
       "aae-miR-133       22\n",
       "aae-miR-137       22\n",
       "aae-miR-14        22\n",
       "aae-miR-184       22\n",
       "aae-miR-1889-3p   22\n",
       "aae-miR-1889-5p   22\n",
       "aae-miR-1890      22\n",
       "aae-miR-1891      22\n",
       "aae-miR-190       24\n",
       "aae-miR-193       22\n",
       "aae-miR-210       19\n",
       "aae-miR-219       23\n",
       "...              ...\n",
       "zma-miR399c-5p    21\n",
       "zma-miR399d-3p    21\n",
       "zma-miR399d-5p    21\n",
       "zma-miR399e-3p    21\n",
       "zma-miR399e-5p    21\n",
       "zma-miR399f-3p    21\n",
       "zma-miR399f-5p    22\n",
       "zma-miR399g-3p    21\n",
       "zma-miR399g-5p    21\n",
       "zma-miR399h-3p    21\n",
       "zma-miR399h-5p    21\n",
       "zma-miR399i-3p    21\n",
       "zma-miR399i-5p    21\n",
       "zma-miR399j-3p    21\n",
       "zma-miR399j-5p    21\n",
       "zma-miR408a       21\n",
       "zma-miR408b-3p    21\n",
       "zma-miR408b-5p    21\n",
       "zma-miR444a       21\n",
       "zma-miR444b       21\n",
       "zma-miR482-3p     20\n",
       "zma-miR482-5p     19\n",
       "zma-miR528a-3p    21\n",
       "zma-miR528a-5p    21\n",
       "zma-miR528b-3p    21\n",
       "zma-miR528b-5p    21\n",
       "zma-miR529-3p     21\n",
       "zma-miR529-5p     21\n",
       "zma-miR827-3p     21\n",
       "zma-miR827-5p     21\n",
       "\n",
       "[35828 rows x 1 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### This parses a fasta file and searches for a particular set of ids (names_to_search), and printing\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "lengths = {}\n",
    "handle = open(\"/projects/ps-yeolab/genomes/mirbase/release_21/mature.fa\", \"rU\")\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    lengths[record.id] = len(record.seq)\n",
    "handle.close()\n",
    "pd.DataFrame(lengths,index=['len']).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### This parses a fasta file renames the ID\n",
    "\n",
    "from Bio import SeqIO\n",
    "handle = open(\"/projects/ps-yeolab/genomes/ce11/ws245_genes.ucsctable.fa\", \"rU\")\n",
    "outfile = \"/projects/ps-yeolab/genomes/ce11/ws245_genes.ucsctable.fix_genenames.fa\"\n",
    "records = []\n",
    "x = 0\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    record.id = record.id.split('_')[2]\n",
    "    records.append(record)\n",
    "SeqIO.write(records,outfile,\"fasta\")\n",
    "handle.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15072423 chrI\n",
      "15279345 chrII\n",
      "13783700 chrIII\n",
      "17493793 chrIV\n",
      "13794 chrM\n",
      "20924149 chrV\n",
      "17718866 chrX\n"
     ]
    }
   ],
   "source": [
    "### This gets len of each sequence\n",
    "\n",
    "from Bio import SeqIO\n",
    "handle = open(\"/projects/ps-yeolab/genomes/ce10/ce10.fa\", \"rU\")\n",
    "\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    print len(record.seq), record.name\n",
    "handle.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15072423 chrI\n",
      "15279345 chrII\n",
      "13783700 chrIII\n",
      "17493793 chrIV\n",
      "13794 chrM\n",
      "20924149 chrV\n",
      "17718866 chrX\n"
     ]
    }
   ],
   "source": [
    "### This gets len of each sequence\n",
    "\n",
    "from Bio import SeqIO\n",
    "handle = open(\"/projects/ps-yeolab/genomes/ce10/chromosomes/all.fa\", \"rU\")\n",
    "\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    print len(record.seq), record.name\n",
    "handle.close()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': 249250621,\n",
       " 'chr10': 135534747,\n",
       " 'chr11': 135006516,\n",
       " 'chr12': 133851895,\n",
       " 'chr13': 115169878,\n",
       " 'chr14': 107349540,\n",
       " 'chr15': 102531392,\n",
       " 'chr16': 90354753,\n",
       " 'chr17': 81195210,\n",
       " 'chr18': 78077248,\n",
       " 'chr19': 59128983,\n",
       " 'chr2': 243199373,\n",
       " 'chr20': 63025520,\n",
       " 'chr21': 48129895,\n",
       " 'chr22': 51304566,\n",
       " 'chr3': 198022430,\n",
       " 'chr4': 191154276,\n",
       " 'chr5': 180915260,\n",
       " 'chr6': 171115067,\n",
       " 'chr7': 159138663,\n",
       " 'chr8': 146364022,\n",
       " 'chr9': 141213431,\n",
       " 'chrM': 16571,\n",
       " 'chrX': 155270560,\n",
       " 'chrY': 59373566}"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "def get_seq_dict_from_file(f, seq_ids=[], file_type='fasta', equal=True):\n",
    "    \"\"\"\n",
    "    Returns dictionary of {name : sequence}\n",
    "     \n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    f : basestring\n",
    "        file location of a fasta file\n",
    "    seq_ids : list\n",
    "        list of sequence ids to search. Empty field returns all sequences\n",
    "    equal : bool\n",
    "        True if seq_id needs to be identical\n",
    "        False if we just have partial identifier\n",
    "    file_type : basestring\n",
    "        default \"fasta\" type file\n",
    "    Returns\n",
    "    -------\n",
    "    records : dict\n",
    "        {name:sequence}\n",
    "    \"\"\"\n",
    "    records = {}\n",
    "    for record in SeqIO.parse(f, file_type):\n",
    "        if len(seq_ids) > 0:\n",
    "            for name in seq_ids:\n",
    "                if equal:\n",
    "                    if name == record.id:\n",
    "                        records[record.id] = record.seq\n",
    "                else:\n",
    "                    if name in record.id:\n",
    "                        records[record.id] = record.seq\n",
    "        else:\n",
    "            records[record.id] = record.seq\n",
    "    return records\n",
    "\n",
    "def get_seq_sizes(f, seq_ids=[], file_type='fasta', equal=True):\n",
    "    \"\"\"\n",
    "    Returns dictionary of {name : seqsize}\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    f\n",
    "    seq_ids\n",
    "    equal\n",
    "    file_type\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    \"\"\"\n",
    "    lengths = {}\n",
    "    records = get_seq_dict_from_file(f, seq_ids, file_type, equal)\n",
    "    \n",
    "    for seq_id, sequence in records.iteritems():\n",
    "        lengths[seq_id] = len(sequence)\n",
    "    return lengths\n",
    "\n",
    "f = \"/projects/ps-yeolab/genomes/hg38/gencode/v24/GRCh38.primary_assembly.genome.fa\"\n",
    "seq_ids = []\n",
    "\n",
    "d = get_seq_sizes(f, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000000\n"
     ]
    }
   ],
   "source": [
    "### this subsets the fasta file to just contain XYZ number of bases\n",
    "\n",
    "from Bio import SeqIO\n",
    "handle = open(\"/projects/ps-yeolab3/cellrangerdatasets/chr19.fa\", \"rU\")\n",
    "outfile = \"/projects/ps-yeolab3/cellrangerdatasets/hg19chr19kbp550_CELLRANGER_REFERENCE/chr19.1M.fa\"\n",
    "records = []\n",
    "for record in SeqIO.parse(handle, \"fasta\"):\n",
    "    record.seq = record.seq[:1000000]\n",
    "    records.append(record)\n",
    "    print(len(record.seq))\n",
    "SeqIO.write(records,outfile,\"fasta\")\n",
    "handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
