{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import seaborn as sns\n",
    "import gffutils\n",
    "from pathway import GO  # look at bfx/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = 'inputs/'\n",
    "output_dir = 'outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = os.path.join(output_dir, 'GO.txt')\n",
    "bg_list = os.path.join(output_dir, 'BG.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_file = '/projects/ps-yeolab3/bay001/annotations/hg19_ensg_to_go.txt'\n",
    "hg19go = GO.GO(go_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/bay001/anaconda2/envs/brian/lib/python2.7/site-packages/gffutils/interface.py:161: UserWarning: It appears that this database has not had the ANALYZE sqlite3 command run on it. Doing so can dramatically speed up queries, and is done by default for databases created with gffutils >0.8.7.1 (this database was created with version 0.8.1) Consider calling the analyze() method of this object.\n",
      "  \"method of this object.\" % self.version)\n"
     ]
    }
   ],
   "source": [
    "# get the background (all genes)\n",
    "db_file = '/projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf.db'\n",
    "DATABASE = gffutils.FeatureDB(db_file)\n",
    "\n",
    "def find_protein_coding_genes(db):\n",
    "    \"\"\"\n",
    "    Finds the number of protein coding genes in the database.\n",
    "    \"\"\"\n",
    "    all_pc = []\n",
    "    pc_count = 0\n",
    "    genes = db.features_of_type('gene')\n",
    "    for gene in genes:\n",
    "        pc = False\n",
    "        for gene_type in gene.attributes['gene_type']:\n",
    "            if gene_type == 'protein_coding':\n",
    "                for geneid in gene.attributes['gene_id']:\n",
    "                    all_pc.append(geneid.split('.')[0]) # just get the ensembl-style IDs, not gencode\n",
    "    return set(all_pc)\n",
    "\n",
    "protein_coding_genes = find_protein_coding_genes(DATABASE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20345\n"
     ]
    }
   ],
   "source": [
    "print(len(protein_coding_genes)) # number of protein coding genes. make sure it's reasonable (~20,000)\n",
    "with open(bg_list, 'w') as f:\n",
    "    for geneid in protein_coding_genes:\n",
    "        f.write('{}\\n'.format(geneid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>region</th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>138665777</td>\n",
       "      <td>138665826</td>\n",
       "      <td>5.290644</td>\n",
       "      <td>3.686779</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000015479.13</td>\n",
       "      <td>ENSG00000015479.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>30824074</td>\n",
       "      <td>30824134</td>\n",
       "      <td>3.887475</td>\n",
       "      <td>5.240236</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000101346.7</td>\n",
       "      <td>ENSG00000101346.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>141667691</td>\n",
       "      <td>141667722</td>\n",
       "      <td>5.369356</td>\n",
       "      <td>4.129040</td>\n",
       "      <td>-</td>\n",
       "      <td>3utr|contained|ENSG00000114126.13</td>\n",
       "      <td>ENSG00000114126.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113667717</td>\n",
       "      <td>113667786</td>\n",
       "      <td>8.662595</td>\n",
       "      <td>4.978746</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113667654</td>\n",
       "      <td>113667717</td>\n",
       "      <td>5.413653</td>\n",
       "      <td>4.076449</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end      name     score strand  \\\n",
       "0   chr5  138665777  138665826  5.290644  3.686779      +   \n",
       "1  chr20   30824074   30824134  3.887475  5.240236      +   \n",
       "2   chr3  141667691  141667722  5.369356  4.129040      -   \n",
       "3   chr1  113667717  113667786  8.662595  4.978746      +   \n",
       "4   chr1  113667654  113667717  5.413653  4.076449      +   \n",
       "\n",
       "                              region              geneid  \n",
       "0  3utr|contained|ENSG00000015479.13  ENSG00000015479.13  \n",
       "1   3utr|contained|ENSG00000101346.7   ENSG00000101346.7  \n",
       "2  3utr|contained|ENSG00000114126.13  ENSG00000114126.13  \n",
       "3   3utr|contained|ENSG00000198799.7   ENSG00000198799.7  \n",
       "4   3utr|contained|ENSG00000198799.7   ENSG00000198799.7  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_table(\n",
    "    os.path.join(input_dir, '764.01v02.IDR.out.0102merged.bed.annotated'),\n",
    "    names=['chrom','start','end','name','score','strand','region','geneid']\n",
    ")\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>region</th>\n",
       "      <th>geneid</th>\n",
       "      <th>ensembl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>138665777</td>\n",
       "      <td>138665826</td>\n",
       "      <td>5.290644</td>\n",
       "      <td>3.686779</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000015479.13</td>\n",
       "      <td>ENSG00000015479.13</td>\n",
       "      <td>ENSG00000015479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>30824074</td>\n",
       "      <td>30824134</td>\n",
       "      <td>3.887475</td>\n",
       "      <td>5.240236</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000101346.7</td>\n",
       "      <td>ENSG00000101346.7</td>\n",
       "      <td>ENSG00000101346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>141667691</td>\n",
       "      <td>141667722</td>\n",
       "      <td>5.369356</td>\n",
       "      <td>4.129040</td>\n",
       "      <td>-</td>\n",
       "      <td>3utr|contained|ENSG00000114126.13</td>\n",
       "      <td>ENSG00000114126.13</td>\n",
       "      <td>ENSG00000114126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113667717</td>\n",
       "      <td>113667786</td>\n",
       "      <td>8.662595</td>\n",
       "      <td>4.978746</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113667654</td>\n",
       "      <td>113667717</td>\n",
       "      <td>5.413653</td>\n",
       "      <td>4.076449</td>\n",
       "      <td>+</td>\n",
       "      <td>3utr|contained|ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799.7</td>\n",
       "      <td>ENSG00000198799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end      name     score strand  \\\n",
       "0   chr5  138665777  138665826  5.290644  3.686779      +   \n",
       "1  chr20   30824074   30824134  3.887475  5.240236      +   \n",
       "2   chr3  141667691  141667722  5.369356  4.129040      -   \n",
       "3   chr1  113667717  113667786  8.662595  4.978746      +   \n",
       "4   chr1  113667654  113667717  5.413653  4.076449      +   \n",
       "\n",
       "                              region              geneid          ensembl  \n",
       "0  3utr|contained|ENSG00000015479.13  ENSG00000015479.13  ENSG00000015479  \n",
       "1   3utr|contained|ENSG00000101346.7   ENSG00000101346.7  ENSG00000101346  \n",
       "2  3utr|contained|ENSG00000114126.13  ENSG00000114126.13  ENSG00000114126  \n",
       "3   3utr|contained|ENSG00000198799.7   ENSG00000198799.7  ENSG00000198799  \n",
       "4   3utr|contained|ENSG00000198799.7   ENSG00000198799.7  ENSG00000198799  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_ensembl_id(row):\n",
    "    try:\n",
    "        return row['geneid'].split('.')[0]\n",
    "    except AttributeError:\n",
    "        return 'intergenic'\n",
    "    \n",
    "genes['ensembl'] = genes.apply(get_ensembl_id, axis=1)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(gene_list, 'w') as f:\n",
    "    for geneid in set(genes['ensembl']):\n",
    "        if geneid != 'intergenic':\n",
    "            f.write('{}\\n'.format(geneid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                      Gene stable ID  \\\n",
      "GO term accession                                                      \n",
      "GO:0005829         {ENSG00000167693, ENSG00000171408, ENSG0000006...   \n",
      "GO:0005515         {ENSG00000167693, ENSG00000064703, ENSG0000011...   \n",
      "GO:0005634         {ENSG00000167693, ENSG00000064703, ENSG0000019...   \n",
      "GO:0005654         {ENSG00000102054, ENSG00000198157, ENSG0000023...   \n",
      "GO:0005737         {ENSG00000124209, ENSG00000261772, ENSG0000018...   \n",
      "\n",
      "                                                           Gene name  \\\n",
      "GO term accession                                                      \n",
      "GO:0005829         {HIF3A, PGM2L1, CEL, FHIT, NCBP2, HSPA8, AGL, ...   \n",
      "GO:0005515         {RNF14, DUOXA1, RNF17, RNF10, RNF11, RNF13, UB...   \n",
      "GO:0005634         {HIF3A, HIST1H4B, HIST1H4A, RNF17, HIST1H4G, R...   \n",
      "GO:0005654         {HIF3A, HIST1H4B, HIST1H4A, C17orf70, HIST1H4F...   \n",
      "GO:0005737         {HIF3A, RNF14, ELANE, RNF17, RNF10, FHIT, NCBP...   \n",
      "\n",
      "                        GO term name  \\\n",
      "GO term accession                      \n",
      "GO:0005829                 {cytosol}   \n",
      "GO:0005515         {protein binding}   \n",
      "GO:0005634                 {nucleus}   \n",
      "GO:0005654             {nucleoplasm}   \n",
      "GO:0005737               {cytoplasm}   \n",
      "\n",
      "                                                  GO term definition  \\\n",
      "GO term accession                                                      \n",
      "GO:0005829         {The part of the cytoplasm that does not conta...   \n",
      "GO:0005515         {Interacting selectively and non-covalently wi...   \n",
      "GO:0005634         {A membrane-bounded organelle of eukaryotic ce...   \n",
      "GO:0005654         {That part of the nuclear content other than t...   \n",
      "GO:0005737         {All of the contents of a cell excluding the p...   \n",
      "\n",
      "                                          GO term evidence code  \\\n",
      "GO term accession                                                 \n",
      "GO:0005829              {TAS, ISS, IBA, NAS, IEA, IC, IMP, IDA}   \n",
      "GO:0005515                                           {IEA, IPI}   \n",
      "GO:0005634         {TAS, ISS, IBA, IPI, IMP, NAS, IEA, IC, IDA}   \n",
      "GO:0005654                  {NAS, ISS, IBA, TAS, IMP, IEA, IDA}   \n",
      "GO:0005737         {TAS, ISS, IBA, IPI, IMP, NAS, IEA, IC, IDA}   \n",
      "\n",
      "                            GO domain  nGenes  inBoth  expressedGOGenes  \\\n",
      "GO term accession                                                         \n",
      "GO:0005829         cellular_component    2766     540              2548   \n",
      "GO:0005515         molecular_function   10450    1401              9086   \n",
      "GO:0005634         cellular_component    5809     889              5323   \n",
      "GO:0005654         cellular_component    1119     242              1022   \n",
      "GO:0005737         cellular_component    5205     772              4794   \n",
      "\n",
      "                   Hypergeometric p-Value  \\\n",
      "GO term accession                           \n",
      "GO:0005829                   2.907186e-45   \n",
      "GO:0005515                   1.573878e-39   \n",
      "GO:0005634                   5.605998e-32   \n",
      "GO:0005654                   1.064598e-26   \n",
      "GO:0005737                   8.224785e-22   \n",
      "\n",
      "                   Bonferroni-corrected Hypergeometric p-Value  \\\n",
      "GO term accession                                                \n",
      "GO:0005829                                        4.264842e-42   \n",
      "GO:0005515                                        2.308879e-36   \n",
      "GO:0005634                                        8.223999e-29   \n",
      "GO:0005654                                        1.561765e-23   \n",
      "GO:0005737                                        1.206576e-18   \n",
      "\n",
      "                                            Ensembl Gene IDs in List  \\\n",
      "GO term accession                                                      \n",
      "GO:0005829         ENSG00000233927,ENSG00000075643,ENSG0000010052...   \n",
      "GO:0005515         ENSG00000124209,ENSG00000103507,ENSG0000010655...   \n",
      "GO:0005634         ENSG00000124209,ENSG00000123091,ENSG0000017921...   \n",
      "GO:0005654         ENSG00000111358,ENSG00000131508,ENSG0000017393...   \n",
      "GO:0005737         ENSG00000124209,ENSG00000173933,ENSG0000017921...   \n",
      "\n",
      "                                                Gene symbols in List  \\\n",
      "GO term accession                                                      \n",
      "GO:0005829         NCBP1,CRKL,AP1M1,FKBP4,SMG5,BLOC1S6,SMG7,PPP3R...   \n",
      "GO:0005515         RNF14,UBE2Q2,NCBP1,HSPA5,RNF13,NELFB,RNF11,PPP...   \n",
      "GO:0005634         RNF14,SMARCC1,RNF11,HSPA5,NELFB,MNT,SUMO3,STK1...   \n",
      "GO:0005654         KHSRP,NCBP1,PPP3R1,ZFPM2,SUMO3,ZFPM1,AKT1,SRSF...   \n",
      "GO:0005737         RNF14,UBE2Q2,NELFB,SUMO3,STK11,FKBP4,FKBP5,HIP...   \n",
      "\n",
      "                  GO Term Description  N Genes in GO category  \\\n",
      "GO term accession                                               \n",
      "GO:0005829                    cytosol                    2766   \n",
      "GO:0005515            protein binding                   10450   \n",
      "GO:0005634                    nucleus                    5809   \n",
      "GO:0005654                nucleoplasm                    1119   \n",
      "GO:0005737                  cytoplasm                    5205   \n",
      "\n",
      "                   N Expressed Genes in GO Category  \\\n",
      "GO term accession                                     \n",
      "GO:0005829                                     2548   \n",
      "GO:0005515                                     9086   \n",
      "GO:0005634                                     5323   \n",
      "GO:0005654                                     1022   \n",
      "GO:0005737                                     4794   \n",
      "\n",
      "                   N Genes in List and GO Category  \n",
      "GO term accession                                   \n",
      "GO:0005829                                     540  \n",
      "GO:0005515                                    1401  \n",
      "GO:0005634                                     889  \n",
      "GO:0005654                                     242  \n",
      "GO:0005737                                     772  \n"
     ]
    }
   ],
   "source": [
    "df = hg19go.enrichment(list(set(genes['ensembl'])), background=list(protein_coding_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
