{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import functools\n",
    "from encode import encode_helpers\n",
    "# from gscripts.encode import encode_helpers # this is from gscripts, but having trouble with import after install :/\n",
    "\n",
    "actual_peaks_dir = \"/projects/ps-yeolab3/encode/analysis/Eric_IDR/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fetch all eCLIP file locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "merged_data = encode_helpers.get_merged_data()\n",
    "\n",
    "#For Peak Analysis we only want to analyze datasets that have been submitted \n",
    "#(we can talk about other ways to filter if you want, but these are the most confident datasets)\n",
    "merged_data = merged_data[merged_data.submitted]\n",
    "\n",
    "\n",
    "#merged_data = merged_data[['CLIP', 'INPUT', 'input_norm']]\n",
    "make_and_filter_clipper_moderate = functools.partial(encode_helpers.make_and_filter_clipper, l2fc=3, pval=3)\n",
    "merged_data['filtered_peaks'] = merged_data['input_norm'].apply(make_and_filter_clipper_moderate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Column Information for data frame\n",
    "\n",
    "- uID - internal unique ID for expermental triple (rep1, rep2, size matched input)\n",
    "- Cell line - cell line experiment was done in\n",
    "- RBP - RNA Binding Protein clipped\n",
    "â€“ CLIP - bam file for the clip rep\n",
    "- INPUt - bam file for the input.  For both reps this will be the same\n",
    "- input norm - input normalized (but not filtered) bed file.  Contains all peaks no matter signifance value\n",
    "- filtered_peaks - peaks that pass signifiance threshold (log2 fold change > 3 and -log10 p-value > 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>CLIP</th>\n",
       "      <th>CLIP</th>\n",
       "      <th></th>\n",
       "      <th>INPUT</th>\n",
       "      <th>Submitted Date</th>\n",
       "      <th>annotation</th>\n",
       "      <th>family</th>\n",
       "      <th>family_bound</th>\n",
       "      <th>family_mapping_submitable</th>\n",
       "      <th>is_qced</th>\n",
       "      <th>note</th>\n",
       "      <th>notes</th>\n",
       "      <th>peaks_submitable</th>\n",
       "      <th>reference</th>\n",
       "      <th>submitted</th>\n",
       "      <th>generally_submittable</th>\n",
       "      <th>input_norm</th>\n",
       "      <th>filtered_peaks</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uID</th>\n",
       "      <th>Cell line</th>\n",
       "      <th>RBP</th>\n",
       "      <th>rep</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">203</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">HepG2</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">HNRNPC</th>\n",
       "      <th>rep1</th>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-09-16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGU motif? binds alus</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep2</th>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-09-16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGU motif? binds alus</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">204</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">HepG2</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">RBFOX2</th>\n",
       "      <th>rep1</th>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-18</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGCAUG motif</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep2</th>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-18</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGCAUG motif</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>IGF2BP1</th>\n",
       "      <th>rep1</th>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-18</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>CDS/3UTR</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CLIP                                                                     CLIP  \\\n",
       "uID Cell line RBP     rep                                                       \n",
       "203 HepG2     HNRNPC  rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "204 HepG2     RBFOX2  rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "205 HepG2     IGF2BP1 rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "\n",
       "CLIP                          \\\n",
       "uID Cell line RBP     rep      \n",
       "203 HepG2     HNRNPC  rep1     \n",
       "                      rep2     \n",
       "204 HepG2     RBFOX2  rep1     \n",
       "                      rep2     \n",
       "205 HepG2     IGF2BP1 rep1     \n",
       "\n",
       "CLIP                                                                    INPUT  \\\n",
       "uID Cell line RBP     rep                                                       \n",
       "203 HepG2     HNRNPC  rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "204 HepG2     RBFOX2  rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "205 HepG2     IGF2BP1 rep1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "\n",
       "CLIP                       Submitted Date annotation family family_bound  \\\n",
       "uID Cell line RBP     rep                                                  \n",
       "203 HepG2     HNRNPC  rep1     2015-09-16                          FALSE   \n",
       "                      rep2     2015-09-16                          FALSE   \n",
       "204 HepG2     RBFOX2  rep1     2015-02-18                          FALSE   \n",
       "                      rep2     2015-02-18                          FALSE   \n",
       "205 HepG2     IGF2BP1 rep1     2015-02-18                          FALSE   \n",
       "\n",
       "CLIP                       family_mapping_submitable is_qced note  \\\n",
       "uID Cell line RBP     rep                                           \n",
       "203 HepG2     HNRNPC  rep1                     False    True        \n",
       "                      rep2                     False    True        \n",
       "204 HepG2     RBFOX2  rep1                     False    True        \n",
       "                      rep2                     False    True        \n",
       "205 HepG2     IGF2BP1 rep1                     False    True        \n",
       "\n",
       "CLIP                                                  notes peaks_submitable  \\\n",
       "uID Cell line RBP     rep                                                      \n",
       "203 HepG2     HNRNPC  rep1  Intronic, UGU motif? binds alus             True   \n",
       "                      rep2  Intronic, UGU motif? binds alus             True   \n",
       "204 HepG2     RBFOX2  rep1           Intronic, UGCAUG motif             True   \n",
       "                      rep2           Intronic, UGCAUG motif             True   \n",
       "205 HepG2     IGF2BP1 rep1                         CDS/3UTR             True   \n",
       "\n",
       "CLIP                       reference submitted generally_submittable  \\\n",
       "uID Cell line RBP     rep                                              \n",
       "203 HepG2     HNRNPC  rep1                True                  True   \n",
       "                      rep2                True                  True   \n",
       "204 HepG2     RBFOX2  rep1                True                  True   \n",
       "                      rep2                True                  True   \n",
       "205 HepG2     IGF2BP1 rep1                True                  True   \n",
       "\n",
       "CLIP                                                               input_norm  \\\n",
       "uID Cell line RBP     rep                                                       \n",
       "203 HepG2     HNRNPC  rep1  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...   \n",
       "204 HepG2     RBFOX2  rep1  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...   \n",
       "                      rep2  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...   \n",
       "205 HepG2     IGF2BP1 rep1  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...   \n",
       "\n",
       "CLIP                                                           filtered_peaks  \n",
       "uID Cell line RBP     rep                                                      \n",
       "203 HepG2     HNRNPC  rep1  /home/gpratt/projects/encode/analysis/peak_rea...  \n",
       "                      rep2  /home/gpratt/projects/encode/analysis/peak_rea...  \n",
       "204 HepG2     RBFOX2  rep1  /home/gpratt/projects/encode/analysis/peak_rea...  \n",
       "                      rep2  /home/gpratt/projects/encode/analysis/peak_rea...  \n",
       "205 HepG2     IGF2BP1 rep1  /home/gpratt/projects/encode/analysis/peak_rea...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Format of eCLIP bed file\n",
    "\n",
    "- Col 1: chrom\n",
    "- Col 2: start\n",
    "- Col 3: stop\n",
    "- Col 4: log2 fold change\n",
    "- Col 5: -log10 p-value\n",
    "- Col 6: strand\n",
    "- Col 7: annotation\n",
    "- Col 8: gene ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr7\t138151219\t138151300\t23.6215112519682\t3.45382390435692\t+\tintron|contained|ENSG00000122779.12\tENSG00000122779.12\r\n",
      "chr7\t139051735\t139051798\t15.3672998138034\t3.7517921003001\t+\tintron|contained|ENSG00000146963.13:ENSG00000269955.2\tENSG00000146963.13|ENSG00000269955.2\r\n"
     ]
    }
   ],
   "source": [
    "foo = merged_data['input_norm'][0]\n",
    "\n",
    "!head -n 2 $foo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct IDR manifest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rep_information = ['CLIP', 'input_norm']\n",
    "temp_index = list(merged_data.columns.difference(rep_information))\n",
    "idr_manifest = merged_data.xs(\"rep1\", level=\"rep\")\n",
    "idr_manifest = idr_manifest.set_index(temp_index, append=True)\n",
    "idr_manifest = idr_manifest.reset_index()\n",
    "idr_manifest = idr_manifest.set_index(merged_data.index.names[:-1])\n",
    "idr_manifest['uID'] = idr_manifest.index.get_level_values(\"uID\")\n",
    "\n",
    "idr_manifest['replicate_merged'] = idr_manifest.uID.apply(lambda x: os.path.join(actual_peaks_dir, x + \".01v02.IDR.out.0102merged.bed\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Column information for dataframe \n",
    "good luck its similar to the other one, just with one col per experiment. \n",
    "\n",
    "replicate_merged_v2 is the important column its got the IDR peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>CLIP</th>\n",
       "      <th></th>\n",
       "      <th>INPUT</th>\n",
       "      <th>Submitted Date</th>\n",
       "      <th>annotation</th>\n",
       "      <th>family</th>\n",
       "      <th>family_bound</th>\n",
       "      <th>family_mapping_submitable</th>\n",
       "      <th>filtered_peaks</th>\n",
       "      <th>generally_submittable</th>\n",
       "      <th>is_qced</th>\n",
       "      <th>note</th>\n",
       "      <th>notes</th>\n",
       "      <th>peaks_submitable</th>\n",
       "      <th>reference</th>\n",
       "      <th>submitted</th>\n",
       "      <th>CLIP</th>\n",
       "      <th>input_norm</th>\n",
       "      <th>uID</th>\n",
       "      <th>replicate_merged</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uID</th>\n",
       "      <th>Cell line</th>\n",
       "      <th>RBP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>HNRNPC</th>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-09-16</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGU motif? binds alus</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>203</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_IDR/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>RBFOX2</th>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-18</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic, UGCAUG motif</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>204</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_IDR/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>IGF2BP1</th>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-18</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>CDS/3UTR</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>205</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_IDR/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>HNRNPK</th>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-23</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>Intronic (dist+prox), CCC motif</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>206</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_IDR/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>SRSF7</th>\n",
       "      <td></td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>2015-02-23</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>False</td>\n",
       "      <td>/home/gpratt/projects/encode/analysis/peak_rea...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>top binding site is 3' utr extension?</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/encode_ma...</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_Inpu...</td>\n",
       "      <td>209</td>\n",
       "      <td>/projects/ps-yeolab3/encode/analysis/Eric_IDR/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "CLIP                                                                 INPUT  \\\n",
       "uID Cell line RBP                                                            \n",
       "203 HepG2     HNRNPC     /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "204 HepG2     RBFOX2     /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "205 HepG2     IGF2BP1    /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "206 HepG2     HNRNPK     /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "209 HepG2     SRSF7      /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "\n",
       "CLIP                  Submitted Date annotation family family_bound  \\\n",
       "uID Cell line RBP                                                     \n",
       "203 HepG2     HNRNPC      2015-09-16                          FALSE   \n",
       "204 HepG2     RBFOX2      2015-02-18                          FALSE   \n",
       "205 HepG2     IGF2BP1     2015-02-18                          FALSE   \n",
       "206 HepG2     HNRNPK      2015-02-23                          FALSE   \n",
       "209 HepG2     SRSF7       2015-02-23                          FALSE   \n",
       "\n",
       "CLIP                  family_mapping_submitable  \\\n",
       "uID Cell line RBP                                 \n",
       "203 HepG2     HNRNPC                      False   \n",
       "204 HepG2     RBFOX2                      False   \n",
       "205 HepG2     IGF2BP1                     False   \n",
       "206 HepG2     HNRNPK                      False   \n",
       "209 HepG2     SRSF7                       False   \n",
       "\n",
       "CLIP                                                      filtered_peaks  \\\n",
       "uID Cell line RBP                                                          \n",
       "203 HepG2     HNRNPC   /home/gpratt/projects/encode/analysis/peak_rea...   \n",
       "204 HepG2     RBFOX2   /home/gpratt/projects/encode/analysis/peak_rea...   \n",
       "205 HepG2     IGF2BP1  /home/gpratt/projects/encode/analysis/peak_rea...   \n",
       "206 HepG2     HNRNPK   /home/gpratt/projects/encode/analysis/peak_rea...   \n",
       "209 HepG2     SRSF7    /home/gpratt/projects/encode/analysis/peak_rea...   \n",
       "\n",
       "CLIP                  generally_submittable is_qced note  \\\n",
       "uID Cell line RBP                                          \n",
       "203 HepG2     HNRNPC                   True    True        \n",
       "204 HepG2     RBFOX2                   True    True        \n",
       "205 HepG2     IGF2BP1                  True    True        \n",
       "206 HepG2     HNRNPK                   True    True        \n",
       "209 HepG2     SRSF7                    True    True        \n",
       "\n",
       "CLIP                                                   notes peaks_submitable  \\\n",
       "uID Cell line RBP                                                               \n",
       "203 HepG2     HNRNPC         Intronic, UGU motif? binds alus             True   \n",
       "204 HepG2     RBFOX2                  Intronic, UGCAUG motif             True   \n",
       "205 HepG2     IGF2BP1                               CDS/3UTR             True   \n",
       "206 HepG2     HNRNPK         Intronic (dist+prox), CCC motif             True   \n",
       "209 HepG2     SRSF7    top binding site is 3' utr extension?             True   \n",
       "\n",
       "CLIP                  reference submitted  \\\n",
       "uID Cell line RBP                           \n",
       "203 HepG2     HNRNPC                 True   \n",
       "204 HepG2     RBFOX2                 True   \n",
       "205 HepG2     IGF2BP1                True   \n",
       "206 HepG2     HNRNPK                 True   \n",
       "209 HepG2     SRSF7                  True   \n",
       "\n",
       "CLIP                                                                CLIP  \\\n",
       "uID Cell line RBP                                                          \n",
       "203 HepG2     HNRNPC   /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "204 HepG2     RBFOX2   /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "205 HepG2     IGF2BP1  /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "206 HepG2     HNRNPK   /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "209 HepG2     SRSF7    /projects/ps-yeolab3/encode/analysis/encode_ma...   \n",
       "\n",
       "CLIP                                                          input_norm  uID  \\\n",
       "uID Cell line RBP                                                               \n",
       "203 HepG2     HNRNPC   /projects/ps-yeolab3/encode/analysis/Eric_Inpu...  203   \n",
       "204 HepG2     RBFOX2   /projects/ps-yeolab3/encode/analysis/Eric_Inpu...  204   \n",
       "205 HepG2     IGF2BP1  /projects/ps-yeolab3/encode/analysis/Eric_Inpu...  205   \n",
       "206 HepG2     HNRNPK   /projects/ps-yeolab3/encode/analysis/Eric_Inpu...  206   \n",
       "209 HepG2     SRSF7    /projects/ps-yeolab3/encode/analysis/Eric_Inpu...  209   \n",
       "\n",
       "CLIP                                                    replicate_merged  \n",
       "uID Cell line RBP                                                         \n",
       "203 HepG2     HNRNPC   /projects/ps-yeolab3/encode/analysis/Eric_IDR/...  \n",
       "204 HepG2     RBFOX2   /projects/ps-yeolab3/encode/analysis/Eric_IDR/...  \n",
       "205 HepG2     IGF2BP1  /projects/ps-yeolab3/encode/analysis/Eric_IDR/...  \n",
       "206 HepG2     HNRNPK   /projects/ps-yeolab3/encode/analysis/Eric_IDR/...  \n",
       "209 HepG2     SRSF7    /projects/ps-yeolab3/encode/analysis/Eric_IDR/...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idr_manifest.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Format of IDR bed file\n",
    "\n",
    "- Col 1: chrom\n",
    "- Col 2: start\n",
    "- Col 3: stop\n",
    "- Col 4: log2 fold change\n",
    "- Col 5: -log10 p-value\n",
    "- Col 6: strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t247086670\t247086708\t5.15157944416774\t3.53270919042856\t-\r\n",
      "chr1\t247086708\t247086720\t3.40206294870218\t3.09777048796747\t-\r\n"
     ]
    }
   ],
   "source": [
    "foo = idr_manifest['replicate_merged'][0]\n",
    "\n",
    "!head -n 2 $foo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Save the IDR manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "del idr_manifest['uID']\n",
    "idr_manifest.reset_index(inplace=True)\n",
    "idr_manifest.to_csv(\n",
    "    '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS.txt', \n",
    "    sep='\\t', header=True, index=False, encoding='utf-8'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
