{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic download from encode script.\n",
    "- Current stats as of 2-28:\n",
    "    - 526 Hg19 bams total (254 HepG2, 272 K562).  \n",
    "    - 474 Hg19 bams (230 HepG2, 244 K562)\n",
    "    - 52 control Hg19 bams (24 HepG2, 28 K562)\n",
    "    - 2 missing possible controls (TIA1 K562 and PPIL4 K562)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import urllib\n",
    "import json\n",
    "import requests\n",
    "import os\n",
    "import glob\n",
    "from collections import defaultdict\n",
    "import qtools\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "\n",
    "pd.set_option(\"display.max_columns\",500)\n",
    "host = 'https://www.encodeproject.org'\n",
    "experiments = \"https://www.encodeproject.org/experiments/\"\n",
    "date = '3-9-2017'\n",
    "annotated_manifest_hepg2 = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_HepG2.csv'\n",
    "annotated_manifest_k562 = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_K562.csv'\n",
    "downloaded_dir = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Do some checks on how many samples we have from the two lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('number of k562 rnaseq: ', 224)\n",
      "('number of RBPs in K562', 224)\n",
      "('number of Expts in K562', 224)\n",
      "('number of Expts in Hepg2', 228)\n",
      "('number of hepg2 rnaseq: ', 228)\n",
      "('number of RBPs in Hepg2', 228)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Official_RBP</th>\n",
       "      <th>GENCODE_v19_ID</th>\n",
       "      <th>target</th>\n",
       "      <th>RBP</th>\n",
       "      <th>RNA-Seq on K562</th>\n",
       "      <th>EXP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>ENSG00000090861.11</td>\n",
       "      <td>AARS-human</td>\n",
       "      <td>AARS</td>\n",
       "      <td>ENCSR599UDS</td>\n",
       "      <td>AARS-BGKLV21-k562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>ENSG00000108270.6</td>\n",
       "      <td>AATF-human</td>\n",
       "      <td>AATF</td>\n",
       "      <td>ENCSR973QSV</td>\n",
       "      <td>AATF-BGKLV13-k562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCF1</td>\n",
       "      <td>ENSG00000204574.8</td>\n",
       "      <td>ABCF1-human</td>\n",
       "      <td>ABCF1</td>\n",
       "      <td>ENCSR721MXZ</td>\n",
       "      <td>ABCF1-BGKLV29-K562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>ENSG00000160710.11</td>\n",
       "      <td>ADAR-human</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>ENCSR164TLB</td>\n",
       "      <td>ADAR-BGKLV24-K562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGGF1</td>\n",
       "      <td>ENSG00000164252.8</td>\n",
       "      <td>AGGF1-human</td>\n",
       "      <td>AGGF1</td>\n",
       "      <td>ENCSR812TLY</td>\n",
       "      <td>AGGF1-BGKLV32-K562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Official_RBP      GENCODE_v19_ID       target    RBP RNA-Seq on K562  \\\n",
       "0         AARS  ENSG00000090861.11   AARS-human   AARS     ENCSR599UDS   \n",
       "1         AATF   ENSG00000108270.6   AATF-human   AATF     ENCSR973QSV   \n",
       "2        ABCF1   ENSG00000204574.8  ABCF1-human  ABCF1     ENCSR721MXZ   \n",
       "3         ADAR  ENSG00000160710.11   ADAR-human   ADAR     ENCSR164TLB   \n",
       "4        AGGF1   ENSG00000164252.8  AGGF1-human  AGGF1     ENCSR812TLY   \n",
       "\n",
       "                  EXP  \n",
       "0   AARS-BGKLV21-k562  \n",
       "1   AATF-BGKLV13-k562  \n",
       "2  ABCF1-BGKLV29-K562  \n",
       "3   ADAR-BGKLV24-K562  \n",
       "4  AGGF1-BGKLV32-K562  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hepg2_df = pd.read_table(annotated_manifest_hepg2)\n",
    "k562_df = pd.read_table(annotated_manifest_k562)\n",
    "\n",
    "print(\"number of k562 rnaseq: \",k562_df.shape[0])\n",
    "print(\"number of RBPs in K562\",len(set(k562_df['Official_RBP'])))\n",
    "print(\"number of Expts in K562\",len(set(k562_df['RNA-Seq on K562'])))\n",
    "print(\"number of Expts in Hepg2\",len(set(hepg2_df['RNA-Seq on HepG2'])))\n",
    "print(\"number of hepg2 rnaseq: \",hepg2_df.shape[0])\n",
    "print(\"number of RBPs in Hepg2\",len(set(hepg2_df['Official_RBP'])))\n",
    "k562_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_expt_from_rbp(manifest_df, rbp, official=False, cell='HepG2'):\n",
    "    \"\"\"\n",
    "    From the manifest generated by xintao, return the expt_id given an RBP name\n",
    "    \"\"\"\n",
    "    colname = 'Official_RBP' if official else 'RBP'\n",
    "    cellcol = 'RNA-Seq on HepG2' if cell == 'HepG2' else 'RNA-Seq on K562'\n",
    "    return manifest_df[manifest_df[colname]==rbp][cellcol].to_string(index=False)\n",
    "    \n",
    "def star_or_tophat(link):\n",
    "    \"\"\"\n",
    "    from the link, reads and returns whether or not STAR or TOPHAT are in the filename.\n",
    "    \"\"\"\n",
    "    # print(\"LINK\",os.path.basename(link))\n",
    "    if os.path.basename(link).find('star') > -1:\n",
    "        return 'STAR'\n",
    "    if os.path.basename(link).find('tophat') > -1:\n",
    "        return 'TOPHAT'\n",
    "    return 'ambiguous'\n",
    "\n",
    "def exists(fpath, current):\n",
    "    \"\"\"\n",
    "    returns whether or not the file (fpath) exists in the directory(current)\n",
    "    \"\"\"\n",
    "    base = os.path.basename(fpath.rstrip())\n",
    "    if not os.path.exists(os.path.join(current,fpath)):\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "def exists_and_return_fullpath(fpath, current):\n",
    "    \"\"\"\n",
    "    returns whether or not the file (fpath) exists in the directory(current)\n",
    "    \"\"\"\n",
    "    base = os.path.basename(fpath.rstrip())\n",
    "    if not os.path.exists(os.path.join(current,fpath)):\n",
    "        return False\n",
    "    else:\n",
    "        return os.path.join(current,fpath)\n",
    "    \n",
    "def get_bams_from_expt_id(\n",
    "    expt_id, assembly, aligner, lab\n",
    "):\n",
    "    \"\"\"\n",
    "    Given an expt id, return a list: [rbp name, [rep1bam, rep2bam], control_expt_id]\n",
    "    \n",
    "    control_expt_id is None if the expt_id given to this function is itself a control.\n",
    "    \n",
    "    \"\"\"\n",
    "    sample_bams = []\n",
    "    control_expts = []\n",
    "    \n",
    "    experiments = \"https://www.encodeproject.org/experiments/\"\n",
    "    url = experiments+expt_id+\"/?format=json\"\n",
    "    response = urllib.urlopen(url)\n",
    "    data = json.loads(response.read())\n",
    "    if 'code' in data.keys():\n",
    "        next\n",
    "    else:\n",
    "        for i in range(0,len(data['files'])):\n",
    "            \n",
    "            if (\n",
    "                (host+data['files'][i]['href']).endswith('bam') & \n",
    "                (data['files'][i]['output_type'] == u'alignments') &\n",
    "                (data['files'][i]['lab'][u'name'] == lab) &\n",
    "                (star_or_tophat(data['files'][i]['submitted_file_name']) == aligner)\n",
    "            ):\n",
    "                if(data['files'][i]['assembly'] == assembly):\n",
    "                    sample_bams.append(\n",
    "                        {\n",
    "                            os.path.basename(data['files'][i]['href']):data['files'][i]['md5sum']\n",
    "                        }\n",
    "                    )\n",
    "        try:\n",
    "            control_expts.append(data['possible_controls'][0]['accession'])\n",
    "        except IndexError:\n",
    "            pass\n",
    "            # print(\"this is a control\")\n",
    "    if(len(control_expts) > 1):\n",
    "        print(\"Warning, this expt {} has more than 1 associated control expt\".format(expt_id))\n",
    "    return data['target']['label'], sample_bams, control_expts\n",
    "\n",
    "def make_expt_dataframe_from_expt_list(expt_list, assembly='hg19', aligner='STAR', lab='encode-processing-pipeline'):\n",
    "    \"\"\"\n",
    "    Given a list of expt ids, return the reps and controls (and md5sums in case we need to check)\n",
    "    \"\"\"\n",
    "    x = tnrange(len(expt_list))\n",
    "    samples = defaultdict(dict)\n",
    "    for h in expt_list:\n",
    "        x.update(1)\n",
    "        sname, s, control_expt = get_bams_from_expt_id(\n",
    "            h, assembly, aligner, lab\n",
    "        ) # gets the sample name, sample, and control expt id\n",
    "        cname, c, _ = get_bams_from_expt_id(\n",
    "            control_expt[0], assembly, aligner, lab\n",
    "        ) # gets the control name, control\n",
    "        for i in range(0,len(s)):\n",
    "            samples_key = 'expt_rep{}'.format(i+1)\n",
    "            samples_md5sum_key = 'expt_rep{}_md5sum'.format(i+1)\n",
    "            control_key = 'control_rep{}'.format(i+1)\n",
    "            controls_md5sum_key = 'control_rep{}_md5sum'.format(i+1)\n",
    "            name_key = 'name'\n",
    "            samples[h][name_key] = sname\n",
    "            samples[h][samples_key] = s[i].keys()[0]\n",
    "            samples[h][control_key] = c[i].keys()[0]\n",
    "            samples[h][samples_md5sum_key] = s[i].values()[0]\n",
    "            samples[h][controls_md5sum_key] = c[i].values()[0]\n",
    "    bams = pd.DataFrame(samples).T\n",
    "    return bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make a list of expt IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hepg2_expts = list(set(hepg2_df['RNA-Seq on HepG2']))\n",
    "k562_expts = list(set(k562_df['RNA-Seq on K562']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner = 'ambiguous' # graveley lab pipeline doesn't adhere to the same file name structure as encode processing pipeline, but we trust that it's all tophat-aligned.\n",
    "lab = 'brenton-graveley' # or 'encode-processing-pipeline'\n",
    "\n",
    "hepg2_bams = make_expt_dataframe_from_expt_list(hepg2_expts, assembly='hg19', aligner=aligner, lab=lab)\n",
    "k562_bams = make_expt_dataframe_from_expt_list(k562_expts, assembly='hg19', aligner=aligner, lab=lab)\n",
    "print(hepg2_bams.shape)\n",
    "print(k562_bams.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hepg2_bams.to_csv(\n",
    "    '/projects/ps-yeolab3/encode/hepg2_{}_{}_bams_for_integrated_analysis.txt'.format(lab, aligner),sep='\\t'\n",
    ")\n",
    "k562_bams.to_csv(\n",
    "    '/projects/ps-yeolab3/encode/k562_{}_{}_bams_for_integrated_analysis.txt'.format(lab, aligner),sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Check md5sum and file existance.\n",
    "- If I've done this once, I don't need to do it again. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def already_checked_md5sum(filepath, md5sum_checked_file):\n",
    "    \"\"\"\n",
    "    given filepaths list, check the md5sum_checked_file to see if any file has been checked.\n",
    "    Parameters\n",
    "    ----------\n",
    "    filepath : list\n",
    "        list of files\n",
    "    md5sum_checked_file : file\n",
    "        dataframe of files\n",
    "        \n",
    "    returns list of files that haven't been checked.\n",
    "    \"\"\"\n",
    "    exists_in_file = ! grep $filepath $md5sum_checked_file\n",
    "    if len(exists_in_file) > 0:\n",
    "        if \"no such file\" not in exists_in_file[0]:\n",
    "            # print(exists_in_file)\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def is_md5sum_equal(row, rep, downloaded_dir):\n",
    "    \"\"\"\n",
    "    Takes a row in a bams dataframe and the rep prefix and determines\n",
    "    whether or not the md5sum is equal to what has been downloaded.\n",
    "    \"\"\"\n",
    "    website_md5sum = row['{}_md5sum'.format(rep)]\n",
    "    filepath = exists_and_return_fullpath(row[rep],downloaded_dir)\n",
    "    calculated_md5sum = ! md5sum $filepath\n",
    "    if website_md5sum != calculated_md5sum[0].split(' ')[0]:\n",
    "        print('TROUBLE: {}, {}, {}'.format(filepath, website_md5sum, calculated_md5sum))\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "def check_all_bams_for_existance(bams, downloaded_dir, to_download_file='download.sh'):\n",
    "    \"\"\"\n",
    "    Writes and prints the files that don't exist to a file\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bams : pandas.DataFrame\n",
    "        dataframe of bams with columns defined in cols.\n",
    "    to_download_file : string\n",
    "        file to download (download.sh)\n",
    "        \n",
    "    \"\"\"\n",
    "    encodeproject_prefix = 'https://www.encodeproject.org/files/'\n",
    "    o = open(to_download_file,'w')\n",
    "    cols = ['expt_rep1','expt_rep2','control_rep1','control_rep2']\n",
    "    for unused,row in bams.iterrows():\n",
    "        for c in cols:\n",
    "            if not(exists_and_return_fullpath(row[c],downloaded_dir)):\n",
    "                print(row[c], \" doesn't exist in {}.\".format(downloaded_dir))\n",
    "                # print(\"mv /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/{} ./\".format(row[c]))\n",
    "                o.write(\n",
    "                    encodeproject_prefix + \"{}/@@download/{}\\n\".format(\n",
    "                        os.path.splitext(row[c])[0],\n",
    "                        row[c]\n",
    "                    )\n",
    "                )\n",
    "    o.close()\n",
    "    \n",
    "def check_md5sums(bams, downloaded_dir, md5sums_checked_file):\n",
    "    \"\"\"\n",
    "    Checks the md5sums of the bams file\n",
    "    \"\"\"\n",
    "    cols = ['expt_rep1','expt_rep2','control_rep1','control_rep2']\n",
    "    x = tnrange(bams.shape[0])\n",
    "    \n",
    "    for _, row in bams.iterrows(): # for each row (RBP)\n",
    "        x.update(1)\n",
    "        for t in tnrange(len(cols), leave=False): # for each column (rep1, rep2, ctrl1, ctrl2)\n",
    "            c = cols[t]\n",
    "            filepath = exists_and_return_fullpath(row[c], downloaded_dir)\n",
    "            if filepath:\n",
    "                if not already_checked_md5sum(\n",
    "                    os.path.join(downloaded_dir,row[c]),\n",
    "                    md5sums_checked_file\n",
    "                ): # if not already checked, check the md5sum\n",
    "                    if is_md5sum_equal(row, c, downloaded_dir):\n",
    "                        with open(md5sums_checked_file,'a') as o:\n",
    "                            o.write(os.path.join(downloaded_dir,row[c]) + '\\n')\n",
    "                    else:\n",
    "                        print(row[c], \"needs to be downloaded again.\")\n",
    "            else:\n",
    "                print(row[c], \"needs to be downloaded again.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for existence of file, otherwise create a list to download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use: xargs -n 1 curl -O -L < FILE to download these...\n",
    "def check_all_bams():\n",
    "    download_hepg2_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/to_download_{}_hepg2_{}_{}'.format(aligner,lab,date)\n",
    "    download_k562_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/to_download_{}_k562_{}_{}'.format(aligner,lab,date)\n",
    "    check_all_bams_for_existance(hepg2_bams, downloaded_dir, download_hepg2_file)\n",
    "    check_all_bams_for_existance(k562_bams, downloaded_dir, download_k562_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "check_all_bams()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check md5sum of the ones we need to download (or just all of them)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# periods indicate row progress, commas indicate rep progress\n",
    "bams = hepg2_bams\n",
    "md5sums_checked_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/md5sums.graveley-tophat.checked'\n",
    "check_md5sums(bams, downloaded_dir, md5sums_checked_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run FeatureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_k562_bams = set(\n",
    "    k562_bams['control_rep1'].append(\n",
    "        k562_bams['control_rep2'].append(\n",
    "            k562_bams['expt_rep1'].append(\n",
    "                k562_bams['expt_rep2']\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")\n",
    "all_hepg2_bams = set(\n",
    "    hepg2_bams['control_rep1'].append(\n",
    "        hepg2_bams['control_rep2'].append(\n",
    "            hepg2_bams['expt_rep1'].append(\n",
    "                hepg2_bams['expt_rep2']\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")\n",
    "print(len(all_k562_bams))\n",
    "print(len(all_hepg2_bams))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "jobs = ['_k562','_hepg2']\n",
    "# jobs = ['_hepg2']\n",
    "for bams in [k562_bams, hepg2_bams]:\n",
    "    cmds = []\n",
    "    for unused,row in bams.iterrows():\n",
    "        p = '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts'\n",
    "        s = '/home/bay001/projects/encode/analysis/featureCounts/individual_expts/'\n",
    "        o = '/home/bay001/projects/encode/analysis/featureCounts/individual_expts/{}{}.counts.txt'.format(row['name'],jobs[i])\n",
    "        a = '/projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf'\n",
    "        ctrl_rep1 = exists_and_return_fullpath(row['control_rep1'],downloaded_dir)\n",
    "        ctrl_rep2 = exists_and_return_fullpath(row['control_rep2'],downloaded_dir)\n",
    "        expt_rep1 = exists_and_return_fullpath(row['expt_rep1'],downloaded_dir)\n",
    "        expt_rep2 = exists_and_return_fullpath(row['expt_rep2'],downloaded_dir)\n",
    "        cmd = '{} -s 2 --tmpDir {} -p -a {} -o {} {} {} {} {}'.format(\n",
    "            p, s, a, o, ctrl_rep1, ctrl_rep2, expt_rep1, expt_rep2\n",
    "        )\n",
    "        cmds.append(cmd)\n",
    "    qtools.Submitter(\n",
    "        cmds, jobs[i], array=True, nodes=1, ppn=1, walltime='4:00:00', submit=True, queue='home-scrm', sh=s+'{}.sh'.format(jobs[i])\n",
    "    )\n",
    "    i = i + 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create normalized bedgraph files from RNASEQ bams\n",
    "- Temporarily do this here, eventually pull these functions out into a script or something"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "rbp_list = ['HNRNPC','HNRNPK','SRSF1','U2AF2','U2AF1','PUF60','EIF4A3','MAGOH','PTBP1','MATR3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "expt_list = []\n",
    "for rbp in rbp_list:\n",
    "    expt_list.append(get_expt_from_rbp(hepg2_df,rbp,official=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = make_expt_dataframe_from_expt_list(expt_list, assembly='hg19', aligner='ambiguous', lab='brenton-graveley')\n",
    "df['Cell line'] = 'HepG2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_bams_to_transform = set(\n",
    "    pd.concat(\n",
    "        [df['expt_rep1'], df['expt_rep2'], df['control_rep1'], df['control_rep2']]\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_sizes = '/projects/ps-yeolab/genomes/hg19/hg19.chrom.sizes'\n",
    "output_temp_bedgraph_dir = '/home/bay001/projects/encode/analysis/rnaseq_bedgraphs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmds = []\n",
    "for bam in list_of_bams_to_transform:\n",
    "    bam_fullpath = exists_and_return_fullpath(bam,downloaded_dir)\n",
    "    output_file = os.path.join(output_temp_bedgraph_dir, '{}.bg'.format(os.path.splitext(bam)[0]))\n",
    "    cmd = 'bedtools genomecov '\n",
    "    cmd = cmd + '-ibam {} '.format(bam_fullpath)\n",
    "    cmd = cmd + '-bg '\n",
    "    cmd = cmd + '-g {} '.format(chrom_sizes)\n",
    "    cmd = cmd + '> {}'.format(output_file)\n",
    "    cmds.append(cmd)\n",
    "qtools.Submitter(\n",
    "    cmds, 'make_bedgraphs_rnaseq', array=True, nodes=1, ppn=1, walltime='4:00:00', submit=True, queue='home-scrm', sh=output_temp_bedgraph_dir+'make_bedgraphs.sh'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "unnormalized_bedgraphs = glob.glob(output_temp_bedgraph_dir + \"*.bg\")\n",
    "progress = tnrange(len(unnormalized_bedgraphs))\n",
    "\n",
    "for bg in unnormalized_bedgraphs:\n",
    "    bam = os.path.join(downloaded_dir, os.path.splitext(os.path.basename(bg))[0] + '.bam')\n",
    "    output_file = os.path.splitext(bg)[0] + '.norm.bg'\n",
    "    cmd = 'normalize_bedGraph.py '\n",
    "    cmd = cmd + '--bg {} '.format(bg)\n",
    "    cmd = cmd + '--bam {} '.format(bam)\n",
    "    cmd = cmd + '> {}'.format(output_file)\n",
    "    ! samtools index $bam\n",
    "    ! $cmd\n",
    "    progress.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for bam in list_of_bams_to_transform:\n",
    "    bam_fullpath = exists_and_return_fullpath(bam,downloaded_dir)\n",
    "    bam_softlink = os.path.join(output_temp_bedgraph_dir, os.path.basename(bam))\n",
    "    ! ln -s $bam_fullpath $bam_softlink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for bam in list_of_bams_to_transform:\n",
    "    bam_fullpath = exists_and_return_fullpath(bam,downloaded_dir).replace('.bam','.bam.bai')\n",
    "    bam_softlink = os.path.join(output_temp_bedgraph_dir, os.path.basename(bam)).replace('.bam','.bam.bai')\n",
    "    ! ln -s $bam_fullpath $bam_softlink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedgraph_dir = '/home/bay001/projects/encode/analysis/rnaseq_bedgraphs'\n",
    "def add_bedgraph_dir_r1(row):\n",
    "    return os.path.join(bedgraph_dir, row['expt_rep1'])\n",
    "def add_bedgraph_dir_input(row):\n",
    "    return os.path.join(bedgraph_dir, row['control_rep1'])\n",
    "\n",
    "dfx = pd.concat([df['name'], df['Cell line'], df['expt_rep1'], df['control_rep1']], axis=1).reset_index()\n",
    "dfx['CLIP'] = dfx.apply(add_bedgraph_dir_r1, axis=1)\n",
    "dfx['INPUT'] = dfx.apply(add_bedgraph_dir_input, axis=1)\n",
    "del dfx['expt_rep1']\n",
    "del dfx['control_rep1']\n",
    "dfx.columns = ['uID','RBP','Cell line', 'CLIP','INPUT']\n",
    "dfx.to_csv(os.path.join(bedgraph_dir + \"/input_normish_manifest.tsv\"), sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control_rep1</th>\n",
       "      <th>control_rep1_md5sum</th>\n",
       "      <th>control_rep2</th>\n",
       "      <th>control_rep2_md5sum</th>\n",
       "      <th>expt_rep1</th>\n",
       "      <th>expt_rep1_md5sum</th>\n",
       "      <th>expt_rep2</th>\n",
       "      <th>expt_rep2_md5sum</th>\n",
       "      <th>name</th>\n",
       "      <th>Cell line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCSR052IYH</th>\n",
       "      <td>ENCFF460AAE.bam</td>\n",
       "      <td>b691efe23fdd3873e7b5d5e6a9b20ba4</td>\n",
       "      <td>ENCFF058QDT.bam</td>\n",
       "      <td>d8f3f5606b74eccd282f560e20b8c893</td>\n",
       "      <td>ENCFF031QZG.bam</td>\n",
       "      <td>eddbeb519d034412d304f5ef2aa687d2</td>\n",
       "      <td>ENCFF531AUK.bam</td>\n",
       "      <td>9c3f52480237457fe97298e7cb0b30be</td>\n",
       "      <td>HNRNPC</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR064DXG</th>\n",
       "      <td>ENCFF571ARP.bam</td>\n",
       "      <td>ca8e470d0d42e42ee79162a245c647ec</td>\n",
       "      <td>ENCFF755GKJ.bam</td>\n",
       "      <td>d05681b0af8f612e9a16b07aa8a0e492</td>\n",
       "      <td>ENCFF155FIK.bam</td>\n",
       "      <td>7fe0ac1e954e9153f75bb15364f9e99d</td>\n",
       "      <td>ENCFF527FPF.bam</td>\n",
       "      <td>fd05926a2b33db5c2312a562a2a28715</td>\n",
       "      <td>PTBP1</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR094KBY</th>\n",
       "      <td>ENCFF571ARP.bam</td>\n",
       "      <td>ca8e470d0d42e42ee79162a245c647ec</td>\n",
       "      <td>ENCFF755GKJ.bam</td>\n",
       "      <td>d05681b0af8f612e9a16b07aa8a0e492</td>\n",
       "      <td>ENCFF707GZQ.bam</td>\n",
       "      <td>36790eca0dbdb238f3be857942435a7e</td>\n",
       "      <td>ENCFF223XHU.bam</td>\n",
       "      <td>94dfa343e8fe2590612ea958fcdd90dc</td>\n",
       "      <td>SRSF1</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR372UWV</th>\n",
       "      <td>ENCFF124DWE.bam</td>\n",
       "      <td>edd78ccc3f9bdc9d399b7c109f0daac6</td>\n",
       "      <td>ENCFF774FAK.bam</td>\n",
       "      <td>77ca88c3500f177ce1d5b5803714c41f</td>\n",
       "      <td>ENCFF882SMH.bam</td>\n",
       "      <td>7abc08fbcc2c8702465bd69ea715403a</td>\n",
       "      <td>ENCFF811FWH.bam</td>\n",
       "      <td>86b24fd4889e3f99b456bad33669f33b</td>\n",
       "      <td>U2AF1</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR492UFS</th>\n",
       "      <td>ENCFF052HTH.bam</td>\n",
       "      <td>e6d81c60dffc32e98f491f1ee02236b8</td>\n",
       "      <td>ENCFF857QSU.bam</td>\n",
       "      <td>2dd2e67ec047cb195b6da4207d0cc26d</td>\n",
       "      <td>ENCFF387CQS.bam</td>\n",
       "      <td>da9d045cf89512ac4acb030671964a26</td>\n",
       "      <td>ENCFF878ELZ.bam</td>\n",
       "      <td>6cf17c1e12dc069e4bd99f6a11a7fcbc</td>\n",
       "      <td>MATR3</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR622MCX</th>\n",
       "      <td>ENCFF499UUZ.bam</td>\n",
       "      <td>b0fdad0ee704046b511099cf04e7eb5e</td>\n",
       "      <td>ENCFF326XKY.bam</td>\n",
       "      <td>ecd206b9d417019fa0c50604afef3094</td>\n",
       "      <td>ENCFF992OHI.bam</td>\n",
       "      <td>8a9fbdba7091cfe89469aeca3f728e2b</td>\n",
       "      <td>ENCFF057UPR.bam</td>\n",
       "      <td>5f81c5e73450692b20ca0fdebd7d2555</td>\n",
       "      <td>U2AF2</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR648BSC</th>\n",
       "      <td>ENCFF829TUN.bam</td>\n",
       "      <td>95f6b60d575d3dd430860c48488129c7</td>\n",
       "      <td>ENCFF350VVR.bam</td>\n",
       "      <td>b9b9274301c8a890e860b19e59ee66ea</td>\n",
       "      <td>ENCFF450QLW.bam</td>\n",
       "      <td>838145dc05a19d800ec951f02311aae7</td>\n",
       "      <td>ENCFF584LUS.bam</td>\n",
       "      <td>9b00a546db42aaff9d5299ee03b9b39d</td>\n",
       "      <td>PUF60</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR746EKS</th>\n",
       "      <td>ENCFF052HTH.bam</td>\n",
       "      <td>e6d81c60dffc32e98f491f1ee02236b8</td>\n",
       "      <td>ENCFF857QSU.bam</td>\n",
       "      <td>2dd2e67ec047cb195b6da4207d0cc26d</td>\n",
       "      <td>ENCFF809RUA.bam</td>\n",
       "      <td>8d7288d20dd32eaaaa69a324d2b90b63</td>\n",
       "      <td>ENCFF629OTS.bam</td>\n",
       "      <td>7f5288bc43dbb1ebab149051ab1318fe</td>\n",
       "      <td>MAGOH</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR853ZJS</th>\n",
       "      <td>ENCFF052HTH.bam</td>\n",
       "      <td>e6d81c60dffc32e98f491f1ee02236b8</td>\n",
       "      <td>ENCFF857QSU.bam</td>\n",
       "      <td>2dd2e67ec047cb195b6da4207d0cc26d</td>\n",
       "      <td>ENCFF589WIS.bam</td>\n",
       "      <td>43f010e32ab0923fc8c21bd551c1a5bb</td>\n",
       "      <td>ENCFF033EZX.bam</td>\n",
       "      <td>61391a78e41d707880d6f78b80226db6</td>\n",
       "      <td>HNRNPK</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR957EEG</th>\n",
       "      <td>ENCFF052HTH.bam</td>\n",
       "      <td>e6d81c60dffc32e98f491f1ee02236b8</td>\n",
       "      <td>ENCFF857QSU.bam</td>\n",
       "      <td>2dd2e67ec047cb195b6da4207d0cc26d</td>\n",
       "      <td>ENCFF224ALJ.bam</td>\n",
       "      <td>34753ae6cad39fa00f40a133a0e49f5e</td>\n",
       "      <td>ENCFF037BCP.bam</td>\n",
       "      <td>768fe465ee8ae4ce498f2cbba2a37f63</td>\n",
       "      <td>EIF4A3</td>\n",
       "      <td>HepG2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                control_rep1               control_rep1_md5sum  \\\n",
       "ENCSR052IYH  ENCFF460AAE.bam  b691efe23fdd3873e7b5d5e6a9b20ba4   \n",
       "ENCSR064DXG  ENCFF571ARP.bam  ca8e470d0d42e42ee79162a245c647ec   \n",
       "ENCSR094KBY  ENCFF571ARP.bam  ca8e470d0d42e42ee79162a245c647ec   \n",
       "ENCSR372UWV  ENCFF124DWE.bam  edd78ccc3f9bdc9d399b7c109f0daac6   \n",
       "ENCSR492UFS  ENCFF052HTH.bam  e6d81c60dffc32e98f491f1ee02236b8   \n",
       "ENCSR622MCX  ENCFF499UUZ.bam  b0fdad0ee704046b511099cf04e7eb5e   \n",
       "ENCSR648BSC  ENCFF829TUN.bam  95f6b60d575d3dd430860c48488129c7   \n",
       "ENCSR746EKS  ENCFF052HTH.bam  e6d81c60dffc32e98f491f1ee02236b8   \n",
       "ENCSR853ZJS  ENCFF052HTH.bam  e6d81c60dffc32e98f491f1ee02236b8   \n",
       "ENCSR957EEG  ENCFF052HTH.bam  e6d81c60dffc32e98f491f1ee02236b8   \n",
       "\n",
       "                control_rep2               control_rep2_md5sum  \\\n",
       "ENCSR052IYH  ENCFF058QDT.bam  d8f3f5606b74eccd282f560e20b8c893   \n",
       "ENCSR064DXG  ENCFF755GKJ.bam  d05681b0af8f612e9a16b07aa8a0e492   \n",
       "ENCSR094KBY  ENCFF755GKJ.bam  d05681b0af8f612e9a16b07aa8a0e492   \n",
       "ENCSR372UWV  ENCFF774FAK.bam  77ca88c3500f177ce1d5b5803714c41f   \n",
       "ENCSR492UFS  ENCFF857QSU.bam  2dd2e67ec047cb195b6da4207d0cc26d   \n",
       "ENCSR622MCX  ENCFF326XKY.bam  ecd206b9d417019fa0c50604afef3094   \n",
       "ENCSR648BSC  ENCFF350VVR.bam  b9b9274301c8a890e860b19e59ee66ea   \n",
       "ENCSR746EKS  ENCFF857QSU.bam  2dd2e67ec047cb195b6da4207d0cc26d   \n",
       "ENCSR853ZJS  ENCFF857QSU.bam  2dd2e67ec047cb195b6da4207d0cc26d   \n",
       "ENCSR957EEG  ENCFF857QSU.bam  2dd2e67ec047cb195b6da4207d0cc26d   \n",
       "\n",
       "                   expt_rep1                  expt_rep1_md5sum  \\\n",
       "ENCSR052IYH  ENCFF031QZG.bam  eddbeb519d034412d304f5ef2aa687d2   \n",
       "ENCSR064DXG  ENCFF155FIK.bam  7fe0ac1e954e9153f75bb15364f9e99d   \n",
       "ENCSR094KBY  ENCFF707GZQ.bam  36790eca0dbdb238f3be857942435a7e   \n",
       "ENCSR372UWV  ENCFF882SMH.bam  7abc08fbcc2c8702465bd69ea715403a   \n",
       "ENCSR492UFS  ENCFF387CQS.bam  da9d045cf89512ac4acb030671964a26   \n",
       "ENCSR622MCX  ENCFF992OHI.bam  8a9fbdba7091cfe89469aeca3f728e2b   \n",
       "ENCSR648BSC  ENCFF450QLW.bam  838145dc05a19d800ec951f02311aae7   \n",
       "ENCSR746EKS  ENCFF809RUA.bam  8d7288d20dd32eaaaa69a324d2b90b63   \n",
       "ENCSR853ZJS  ENCFF589WIS.bam  43f010e32ab0923fc8c21bd551c1a5bb   \n",
       "ENCSR957EEG  ENCFF224ALJ.bam  34753ae6cad39fa00f40a133a0e49f5e   \n",
       "\n",
       "                   expt_rep2                  expt_rep2_md5sum    name  \\\n",
       "ENCSR052IYH  ENCFF531AUK.bam  9c3f52480237457fe97298e7cb0b30be  HNRNPC   \n",
       "ENCSR064DXG  ENCFF527FPF.bam  fd05926a2b33db5c2312a562a2a28715   PTBP1   \n",
       "ENCSR094KBY  ENCFF223XHU.bam  94dfa343e8fe2590612ea958fcdd90dc   SRSF1   \n",
       "ENCSR372UWV  ENCFF811FWH.bam  86b24fd4889e3f99b456bad33669f33b   U2AF1   \n",
       "ENCSR492UFS  ENCFF878ELZ.bam  6cf17c1e12dc069e4bd99f6a11a7fcbc   MATR3   \n",
       "ENCSR622MCX  ENCFF057UPR.bam  5f81c5e73450692b20ca0fdebd7d2555   U2AF2   \n",
       "ENCSR648BSC  ENCFF584LUS.bam  9b00a546db42aaff9d5299ee03b9b39d   PUF60   \n",
       "ENCSR746EKS  ENCFF629OTS.bam  7f5288bc43dbb1ebab149051ab1318fe   MAGOH   \n",
       "ENCSR853ZJS  ENCFF033EZX.bam  61391a78e41d707880d6f78b80226db6  HNRNPK   \n",
       "ENCSR957EEG  ENCFF037BCP.bam  768fe465ee8ae4ce498f2cbba2a37f63  EIF4A3   \n",
       "\n",
       "            Cell line  \n",
       "ENCSR052IYH     HepG2  \n",
       "ENCSR064DXG     HepG2  \n",
       "ENCSR094KBY     HepG2  \n",
       "ENCSR372UWV     HepG2  \n",
       "ENCSR492UFS     HepG2  \n",
       "ENCSR622MCX     HepG2  \n",
       "ENCSR648BSC     HepG2  \n",
       "ENCSR746EKS     HepG2  \n",
       "ENCSR853ZJS     HepG2  \n",
       "ENCSR957EEG     HepG2  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "widgets": {
   "state": {
    "24e206ee73604e4db6753c7072c58432": {
     "views": [
      {
       "cell_index": 23
      }
     ]
    },
    "7fd1558867f24909b0ab7585ef79e31e": {
     "views": [
      {
       "cell_index": 27
      }
     ]
    }
   },
   "version": "1.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
