{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wrapper notebook for submitting the RBP maps script to TSCC\n",
    "- split these up into their respective events notebooks (easier to read that way)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import yaml\n",
    "import glob\n",
    "import rethinkdb as r\n",
    "from collections import defaultdict\n",
    "from qtools import Submitter\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "pd.set_option(\"display.max_colwidth\", 10000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# r.connect(host='metadata.yeolab.sdsc.edu',port=8080,user='rbpmaps',password='rbpmaps')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define manifests, directories, etc.\n",
    "- SBDS-BGKLV24-K562 and PPIL4-BGKLV24-K562 were changed in the K562.csv list from SBDS-BGKLV24_2-K562 and PPIL4-BGKLV24_2-K562 per an email discussion from xintao."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "current_date = '5-12-2017'\n",
    "prefix = ''\n",
    "hepg2_rnaseq_manifest = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_HepG2.csv'\n",
    "k562_rnaseq_manifest = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_K562.csv'\n",
    "rnaseq_manifests = {'HepG2':hepg2_rnaseq_manifest, 'K562':k562_rnaseq_manifest}\n",
    "chrom_sizes = '/projects/ps-yeolab/genomes/hg19/hg19.chrom.sizes'\n",
    "density_runner = '/home/bay001/projects/codebase/stable/rbp-maps/maps/plot_density.py'\n",
    "peak_runner = '/home/bay001/projects/codebase/stable/rbp-maps/maps/plot_peak.py'\n",
    "\n",
    "\n",
    "def split_uid_and_rep(name):\n",
    "    \"\"\"\n",
    "    Splits the name (ie. 204_01_RBFOX2) into: [204, 01, RBFOX2] \n",
    "    (all strings)\n",
    "    \"\"\"\n",
    "    name = name.split('_')\n",
    "    uid = name[0]\n",
    "    rep = name[1]\n",
    "    others = '_'.join(name[2:])\n",
    "    return uid, rep, others\n",
    "\n",
    "def get_clip_file_from_uid(clip, rbp_uid):\n",
    "    \"\"\"\n",
    "    Returns attributes from the submitted CLIP manifest given an id\n",
    "    \"\"\"\n",
    "    # clip = pd.read_table(manifest_file)\n",
    "    rbp_df = clip[clip['uID']==rbp_uid]\n",
    "    clip_rbp = rbp_df['RBP'].to_string(index=False, header=False)\n",
    "    clip_celltype = rbp_df['Cell line'].to_string(index=False, header=False)\n",
    "    clip_rep1 = rbp_df['CLIP_rep1'].to_string(index=False, header=False)\n",
    "    clip_rep2 = rbp_df['CLIP_rep2'].to_string(index=False, header=False)\n",
    "    input_rep = rbp_df['INPUT'].to_string(index=False, header=False)\n",
    "    return clip_rep1, clip_rep2, input_rep, clip_rbp, clip_celltype\n",
    "\n",
    "def get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp_name, cell_type):\n",
    "    \"\"\"\n",
    "    Returns the experiment ID (ie. RBFOX2-BGHLVsomething-K562) given an RBP name \n",
    "    and cell type.\n",
    "    \"\"\"\n",
    "    manifest_filename = rnaseq_manifests[cell_type]\n",
    "    df = pd.read_table(manifest_filename)\n",
    "    exp = df[df['Official_RBP']==rbp_name]['EXP']\n",
    "    if exp.shape[0] == 0:\n",
    "        return \"NO_RNASEQ\"\n",
    "    return exp.to_string(index=False, header=False)\n",
    "\n",
    "def get_annotations_from_splicing_prefix(\n",
    "    annotation_dir, splicing_prefix, \n",
    "    pos_splicing_suffix = '-included-upon-knockdown',\n",
    "    neg_splicing_suffix = '-excluded-upon-knockdown'\n",
    "):\n",
    "    \"\"\"\n",
    "    For each RBP, we're looking specifically for positive and negative annotation.\n",
    "    For PNGs, since we're putting this on a website, we wanted to name these annotations properly.\n",
    "    \n",
    "    Valid suffixes so far:\n",
    "    \n",
    "    PNG: -included-upon-knockdown / -excluded-upon-knockdown\n",
    "    SVG: .positive.nr.txt / .negative.nr.txt\n",
    "    PEAK: .positive.nr.miso / .negative.nr.miso\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    positive = glob.glob(\n",
    "        os.path.join(annotation_dir, splicing_prefix) + \"*\" + pos_splicing_suffix\n",
    "    )\n",
    "    negative = glob.glob(\n",
    "        os.path.join(annotation_dir, splicing_prefix) + \"*\" + neg_splicing_suffix\n",
    "    )\n",
    "    \n",
    "    if (len(positive)==1 and len(negative)==1):\n",
    "        return positive[0], negative[0]\n",
    "    else:\n",
    "        return None, None\n",
    "\n",
    "def get_peak_annotations_from_splicing_prefix(annotation_dir, splicing_prefix):\n",
    "    return get_annotations_from_splicing_prefix(\n",
    "        annotation_dir,\n",
    "        splicing_prefix,\n",
    "        '.positive.nr.miso',\n",
    "        '.negative.nr.miso'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot SE Density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"pos_splicing_suffix = '.positive.nr.txt'\n",
    "neg_splicing_suffix = '.negative.nr.txt'\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "background_control_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr{}/'.format(prefix)\n",
    "bash_scripts_dir = '/home/bay001/projects/maps_20160420/scripts/bash_scripts/'\n",
    "\n",
    "k562_background_control_file = 'K562-constitutive-exons' # 'Brent_RNASEQlist_fromDCC.tsv.file_accessions_K562.hg19_V19.20161207.tsv.strict_CE_all'\n",
    "hepg2_background_control_file = 'HepG2-constitutive-exons' # 'Brent_RNASEQlist_fromDCC.tsv.file_accessions_HepG2.hg19_V19.20161207.tsv.strict_CE_all'\n",
    "\n",
    "k562_background_ce = os.path.join(background_control_dir, k562_background_control_file)\n",
    "k562_background_ase = None\n",
    "k562_background_nse = None\n",
    "hepg2_background_ce = os.path.join(background_control_dir, hepg2_background_control_file)\n",
    "hepg2_background_ase = None\n",
    "hepg2_background_nse = None\n",
    "\n",
    "img_extension = 'png'\n",
    "event = 'se'\n",
    "\n",
    "no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "cmds = []\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "for uid in clip_df['uID']:\n",
    "    \n",
    "    metadata = defaultdict()\n",
    "    metadata[uid] = defaultdict()\n",
    "    metadata[uid]['rep1'] = defaultdict()\n",
    "    metadata[uid]['rep2'] = defaultdict()\n",
    "    \n",
    "    \n",
    "    r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "    \n",
    "    if cell == 'K562':\n",
    "        background = k562_background_ce\n",
    "    else:\n",
    "        background = hepg2_background_ce\n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(\n",
    "        rnaseq_manifests, rbp, cell\n",
    "    )\n",
    "    if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "        no_rnaseq.append(uid)\n",
    "    else:\n",
    "        positive, negative = get_annotations_from_splicing_prefix(\n",
    "            annotation_dir, splicing_prefix, \n",
    "            pos_splicing_suffix, neg_splicing_suffix\n",
    "        )\n",
    "    if(positive == None or negative == None):\n",
    "        no_rnaseq_yet.append(uid)\n",
    "    else:\n",
    "        for r in [r1, r2]:\n",
    "            name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "            output_filename = os.path.join(\n",
    "                output_dir,\n",
    "                name\n",
    "            )\n",
    "\n",
    "            # Build the cmd line\n",
    "            cmd = \"python \" + density_runner\n",
    "            cmd = cmd + \" --event {}\".format(event)\n",
    "            cmd = cmd + \" --ipbam {}\".format(r)\n",
    "            cmd = cmd + \" --inputbam {}\".format(i)\n",
    "            cmd = cmd + \" --output {}\".format(output_filename)\n",
    "            cmd = cmd + \" --annotations {} {} {}\".format(\n",
    "                positive, negative, background\n",
    "            )\n",
    "            cmd = cmd + \" --annotation_type {} {} {}\".format(\n",
    "                'rmats', 'rmats', 'eric',\n",
    "            )\n",
    "            cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "            # cmd = cmd + \" --unflip\"\n",
    "            cmds.append(cmd)\n",
    "\n",
    "            # Build the json dict\n",
    "            rep = 'rep1' if r == r1 else 'rep2'\n",
    "\n",
    "            metadata[uid][rep]['annotations'] = list()\n",
    "            metadata[uid][rep]['annotation_type'] = list()\n",
    "\n",
    "            metadata[uid][rep]['cell'] = cell\n",
    "            metadata[uid][rep]['rbp'] = rbp\n",
    "            metadata[uid][rep]['date'] = current_date\n",
    "            metadata[uid][rep]['ip_bam'] = {'class':'File', 'path':r}\n",
    "            # metadata[uid][rep]['ip_pos_bw'] = r.replace('.bam','.norm.pos.bw') # prog decides this for us.\n",
    "            # metadata[uid][rep]['ip_neg_bw'] = r.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "            metadata[uid][rep]['input_bam'] = {'class':'File', 'path':i}\n",
    "            # metadata[uid][rep]['input_pos_bw'] = i.replace('.bam','.norm.pos.bw')\n",
    "            # metadata[uid][rep]['input_neg_bw'] = i.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "            metadata[uid][rep]['chrom_sizes'] = {'class':'File', 'path':chrom_sizes}\n",
    "            metadata[uid][rep]['unflip'] = False\n",
    "            metadata[uid][rep]['event'] = 'se'\n",
    "            metadata[uid][rep]['exon_offset'] = 50\n",
    "            metadata[uid][rep]['intron_offset'] = 300\n",
    "\n",
    "            metadata[uid][rep]['annotations'].append({'class':'File', 'path':positive})\n",
    "            metadata[uid][rep]['annotations'].append({'class':'File', 'path':negative})\n",
    "            metadata[uid][rep]['annotations'].append({'class':'File', 'path':background})\n",
    "\n",
    "            metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "            metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "            metadata[uid][rep]['annotation_type'].append('eric')\n",
    "\n",
    "            with open(\n",
    "                os.path.join(\n",
    "                    bash_scripts_dir,\"{}.{}.{}.{}.{}.{}.json\".format(\n",
    "                        uid, rep, rbp, cell, current_date, event\n",
    "                    )\n",
    "                ),\n",
    "                'w'\n",
    "            ) as f:\n",
    "                f.write('#!/usr/bin/env make_rbp_map.cwl\\n')\n",
    "                json.dump(metadata[uid][rep], f)\n",
    "Submitter(\n",
    "    cmds, \n",
    "    \"SE_NR_{}\".format(img_extension), \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/SE_NR_{}.sh'.format(img_extension),\n",
    "    submit=True,\n",
    "    array=True,\n",
    "    walltime='2:00:00'\n",
    ")\n",
    "print(\"uIDs for which we don't have splicing data for: {}\".format(\n",
    "        len(no_rnaseq))\n",
    "     )\n",
    "print(\"uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(\n",
    "        len(no_rnaseq_yet))\n",
    "     )\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot SVGs (using all backgrounds this time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "is_5p = True\n",
    "img_extension = 'svg'\n",
    "pos_splicing_suffix = '.positive.nr.txt'\n",
    "neg_splicing_suffix = '.negative.nr.txt'\n",
    "\n",
    "\n",
    "annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "background_control_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "bash_scripts_dir = '/home/bay001/projects/maps_20160420/scripts/bash_scripts/'\n",
    "\n",
    "if is_5p == True:\n",
    "    output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr_5p/'\n",
    "    clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.5p.txt'\n",
    "    prefix = '5p'\n",
    "else:\n",
    "    output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr/'\n",
    "    clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "    prefix = 'wholeread'\n",
    "\n",
    "k562_background_ce = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.strict_CE_all_20170401')\n",
    "k562_background_nse_all = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEall_0.5_20170401')\n",
    "k562_background_nse_inc = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEincl_0.5_20170401')\n",
    "k562_background_nse_exc = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEexcl_0.5_20170401')\n",
    "\n",
    "hepg2_background_ce = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.strict_CE_all_20170401')\n",
    "hepg2_background_nse_all = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEall_0.5_20170401')\n",
    "hepg2_background_nse_inc = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEincl_0.5_20170401')\n",
    "hepg2_background_nse_exc = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEexcl_0.5_20170401')\n",
    "\n",
    "event = 'se'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet (or there were no significant splice events)\n",
    "cmds = []\n",
    "\n",
    "for uid in clip_df['uID']:\n",
    "    \n",
    "    metadata = defaultdict()\n",
    "    metadata[uid] = defaultdict()\n",
    "    metadata[uid]['rep1'] = defaultdict()\n",
    "    metadata[uid]['rep2'] = defaultdict()\n",
    "    \n",
    "    \n",
    "    r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "    \n",
    "    \n",
    "    if cell == 'K562':\n",
    "        background1 = k562_background_ce\n",
    "        background2 = k562_background_nse_all\n",
    "        background3 = k562_background_nse_inc\n",
    "        background4 = k562_background_nse_exc\n",
    "    elif cell == 'HepG2':\n",
    "        background1 = hepg2_background_ce\n",
    "        background2 = hepg2_background_nse_all\n",
    "        background3 = hepg2_background_nse_inc\n",
    "        background4 = hepg2_background_nse_exc\n",
    "    else:\n",
    "        print(cell)\n",
    "    \n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(\n",
    "        rnaseq_manifests, rbp, cell\n",
    "    )\n",
    "    if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "        no_rnaseq.append(uid)\n",
    "    else:\n",
    "        positive, negative = get_annotations_from_splicing_prefix(\n",
    "            annotation_dir, splicing_prefix, \n",
    "            pos_splicing_suffix, neg_splicing_suffix\n",
    "        )\n",
    "        if(positive == None or negative == None):\n",
    "            no_rnaseq_yet.append(uid)\n",
    "        else:\n",
    "            if not (rbp in positive and rbp in negative):\n",
    "                print(\n",
    "                    'warning, these dont match: {}, {}, {}'.format(\n",
    "                        rbp, \n",
    "                        os.path.basename(positive),\n",
    "                        os.path.basename(negative)\n",
    "                    )\n",
    "                )\n",
    "            pos_prefix = os.path.basename(positive).split('-')[0]\n",
    "            neg_prefix = os.path.basename(negative).split('-')[0]\n",
    "            if not (pos_prefix in rbp and neg_prefix in rbp):\n",
    "                print(\n",
    "                    'warning, these dont match: {}, {}, {}'.format(\n",
    "                        rbp, \n",
    "                        os.path.basename(positive),\n",
    "                        os.path.basename(negative)\n",
    "                    )\n",
    "                )\n",
    "            for r in [r1, r2]:\n",
    "                name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "                output_filename = os.path.join(\n",
    "                    output_dir,\n",
    "                    name\n",
    "                )\n",
    "\n",
    "                # Build the cmd line\n",
    "                cmd = \"python \" + density_runner\n",
    "                cmd = cmd + \" --event {}\".format(event)\n",
    "                cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                cmd = cmd + \" --annotations {} {} {} {} {} {}\".format(\n",
    "                    positive, negative, background1, background2, background3, background4\n",
    "                )\n",
    "                cmd = cmd + \" --annotation_type {} {} {} {} {} {}\".format(\n",
    "                    'rmats', 'rmats', 'eric', 'eric', 'eric', 'eric' \n",
    "                )\n",
    "                cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                # if is_5p == True:\n",
    "                #     cmd = cmd + \" --unflip\"\n",
    "                \n",
    "                cmds.append(cmd)\n",
    "\n",
    "                # Build the json dict\n",
    "                rep = 'rep1' if r == r1 else 'rep2'\n",
    "\n",
    "                metadata[uid][rep]['annotations'] = list()\n",
    "                metadata[uid][rep]['annotation_type'] = list()\n",
    "\n",
    "                metadata[uid][rep]['cell'] = cell\n",
    "                metadata[uid][rep]['rbp'] = rbp\n",
    "                metadata[uid][rep]['date'] = current_date\n",
    "                metadata[uid][rep]['ip_bam'] = {'class':'File', 'path':r}\n",
    "                # metadata[uid][rep]['ip_pos_bw'] = r.replace('.bam','.norm.pos.bw') # prog decides this for us.\n",
    "                # metadata[uid][rep]['ip_neg_bw'] = r.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "                metadata[uid][rep]['input_bam'] = {'class':'File', 'path':i}\n",
    "                # metadata[uid][rep]['input_pos_bw'] = i.replace('.bam','.norm.pos.bw')\n",
    "                # metadata[uid][rep]['input_neg_bw'] = i.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "                metadata[uid][rep]['chrom_sizes'] = {'class':'File', 'path':chrom_sizes}\n",
    "                metadata[uid][rep]['unflip'] = False\n",
    "                metadata[uid][rep]['event'] = 'se'\n",
    "                metadata[uid][rep]['exon_offset'] = 50\n",
    "                metadata[uid][rep]['intron_offset'] = 300\n",
    "\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background1})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background2})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background3})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':positive})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':negative})\n",
    "\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "\n",
    "\n",
    "                with open(\n",
    "                    os.path.join(\n",
    "                        bash_scripts_dir,\"{}.{}.{}.{}.{}.{}.json\".format(\n",
    "                            uid, rep, rbp, cell, current_date, event\n",
    "                        )\n",
    "                    ),\n",
    "                    'w'\n",
    "                ) as f:\n",
    "                    f.write('#!/usr/bin/env make_rbp_map.cwl\\n')\n",
    "                    json.dump(metadata[uid][rep], f)\n",
    "Submitter(\n",
    "    cmds, \n",
    "    \"{}-SE_NR_{}\".format(prefix, img_extension), \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}-SE_NR_{}.sh'.format(prefix, img_extension),\n",
    "    submit=True,\n",
    "    array=True,\n",
    "    walltime='3:00:00'\n",
    ")\n",
    "print(\"uIDs for which we don't have splicing data for: {}\".format(\n",
    "        len(no_rnaseq))\n",
    "     )\n",
    "print(\"uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(\n",
    "        len(no_rnaseq_yet))\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "is_5p = False\n",
    "img_extension = 'png'\n",
    "pos_splicing_suffix = '-included-upon-knockdown'\n",
    "neg_splicing_suffix = '-excluded-upon-knockdown'\n",
    "\n",
    "\n",
    "annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se_renamed/'\n",
    "background_control_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se_renamed/'\n",
    "\n",
    "bash_scripts_dir = '/home/bay001/projects/maps_20160420/scripts/bash_scripts/{}'.format(current_date)\n",
    "! mkdir $bash_scripts_dir\n",
    "if is_5p == True:\n",
    "    output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr_5p/'\n",
    "    clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.5p.txt'\n",
    "    prefix = '5p'\n",
    "else:\n",
    "    output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr/'\n",
    "    clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "    prefix = 'wholeread'\n",
    "\n",
    "k562_background_ce = os.path.join(background_control_dir, 'K562_constitutive_exons')\n",
    "k562_background_nse_all = os.path.join(background_control_dir, 'K562_native_cassette_exons')\n",
    "k562_background_nse_inc = os.path.join(background_control_dir, 'K562_natively_included_exons')\n",
    "k562_background_nse_exc = os.path.join(background_control_dir, 'K562_natively_excluded_exons')\n",
    "\n",
    "hepg2_background_ce = os.path.join(background_control_dir, 'HepG2_constitutive_exons')\n",
    "hepg2_background_nse_all = os.path.join(background_control_dir, 'HepG2_native_cassette_exons')\n",
    "hepg2_background_nse_inc = os.path.join(background_control_dir, 'HepG2_natively_included_exons')\n",
    "hepg2_background_nse_exc = os.path.join(background_control_dir, 'HepG2_natively_excluded_exons')\n",
    "\n",
    "event = 'se'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet (or there were no significant splice events)\n",
    "cmds = []\n",
    "\n",
    "for uid in clip_df['uID']:\n",
    "    \n",
    "    metadata = defaultdict()\n",
    "    metadata[uid] = defaultdict()\n",
    "    metadata[uid]['rep1'] = defaultdict()\n",
    "    metadata[uid]['rep2'] = defaultdict()\n",
    "    \n",
    "    \n",
    "    r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "    \n",
    "    \n",
    "    if cell == 'K562':\n",
    "        background1 = k562_background_ce\n",
    "        background2 = k562_background_nse_all\n",
    "        background3 = k562_background_nse_inc\n",
    "        background4 = k562_background_nse_exc\n",
    "    elif cell == 'HepG2':\n",
    "        background1 = hepg2_background_ce\n",
    "        background2 = hepg2_background_nse_all\n",
    "        background3 = hepg2_background_nse_inc\n",
    "        background4 = hepg2_background_nse_exc\n",
    "    else:\n",
    "        print(cell)\n",
    "    \n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(\n",
    "        rnaseq_manifests, rbp, cell\n",
    "    )\n",
    "    if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "        no_rnaseq.append(uid)\n",
    "    else:\n",
    "        positive, negative = get_annotations_from_splicing_prefix(\n",
    "            annotation_dir, splicing_prefix, \n",
    "            pos_splicing_suffix, neg_splicing_suffix\n",
    "        )\n",
    "        if(positive == None or negative == None):\n",
    "            no_rnaseq_yet.append(uid)\n",
    "        else:\n",
    "            if not (rbp in positive and rbp in negative):\n",
    "                print(\n",
    "                    'warning, these dont match: {}, {}, {}'.format(\n",
    "                        rbp, \n",
    "                        os.path.basename(positive),\n",
    "                        os.path.basename(negative)\n",
    "                    )\n",
    "                )\n",
    "            pos_prefix = os.path.basename(positive).split('-')[0]\n",
    "            neg_prefix = os.path.basename(negative).split('-')[0]\n",
    "            if not (pos_prefix in rbp and neg_prefix in rbp):\n",
    "                print(\n",
    "                    'warning, these dont match: {}, {}, {}'.format(\n",
    "                        rbp, \n",
    "                        os.path.basename(positive),\n",
    "                        os.path.basename(negative)\n",
    "                    )\n",
    "                )\n",
    "            for r in [r1, r2]:\n",
    "                name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "                output_filename = os.path.join(\n",
    "                    output_dir,\n",
    "                    name\n",
    "                )\n",
    "\n",
    "                # Build the cmd line\n",
    "                cmd = \"python \" + density_runner\n",
    "                cmd = cmd + \" --event {}\".format(event)\n",
    "                cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                cmd = cmd + \" --annotations {} {} {} {} {} {}\".format(\n",
    "                    positive, negative, background1, background2, background3, background4\n",
    "                )\n",
    "                cmd = cmd + \" --annotation_type {} {} {} {} {} {}\".format(\n",
    "                    'rmats', 'rmats', 'eric', 'eric', 'eric', 'eric' \n",
    "                )\n",
    "                cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                # if is_5p == True:\n",
    "                #     cmd = cmd + \" --unflip\"\n",
    "                \n",
    "                cmds.append(cmd)\n",
    "\n",
    "                # Build the json dict\n",
    "                rep = 'rep1' if r == r1 else 'rep2'\n",
    "\n",
    "                metadata[uid][rep]['annotations'] = list()\n",
    "                metadata[uid][rep]['annotation_type'] = list()\n",
    "\n",
    "                metadata[uid][rep]['cell'] = cell\n",
    "                metadata[uid][rep]['rbp'] = rbp\n",
    "                metadata[uid][rep]['date'] = current_date\n",
    "                metadata[uid][rep]['ip_bam'] = {'class':'File', 'path':r}\n",
    "                # metadata[uid][rep]['ip_pos_bw'] = r.replace('.bam','.norm.pos.bw') # prog decides this for us.\n",
    "                # metadata[uid][rep]['ip_neg_bw'] = r.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "                metadata[uid][rep]['input_bam'] = {'class':'File', 'path':i}\n",
    "                # metadata[uid][rep]['input_pos_bw'] = i.replace('.bam','.norm.pos.bw')\n",
    "                # metadata[uid][rep]['input_neg_bw'] = i.replace('.bam','.norm.neg.bw')\n",
    "\n",
    "                metadata[uid][rep]['chrom_sizes'] = {'class':'File', 'path':chrom_sizes}\n",
    "                metadata[uid][rep]['unflip'] = False\n",
    "                metadata[uid][rep]['event'] = 'se'\n",
    "                metadata[uid][rep]['exon_offset'] = 50\n",
    "                metadata[uid][rep]['intron_offset'] = 300\n",
    "\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background1})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background2})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background3})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':positive})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':negative})\n",
    "\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "\n",
    "\n",
    "                with open(\n",
    "                    os.path.join(\n",
    "                        bash_scripts_dir,\"{}.{}.{}.{}.{}.{}.json\".format(\n",
    "                            uid, rep, rbp, cell, current_date, event\n",
    "                        )\n",
    "                    ),\n",
    "                    'w'\n",
    "                ) as f:\n",
    "                    f.write('#!/usr/bin/env make_rbp_map.cwl\\n')\n",
    "                    json.dump(metadata[uid][rep], f)\n",
    "Submitter(\n",
    "    cmds, \n",
    "    \"{}-SE_NR_{}\".format(prefix, img_extension), \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}/{}-SE_NR_{}.sh'.format(current_date, prefix, img_extension),\n",
    "    submit=False,\n",
    "    array=True,\n",
    "    walltime='3:00:00'\n",
    ")\n",
    "print(\"uIDs for which we don't have splicing data for: {}\".format(\n",
    "        len(no_rnaseq))\n",
    "     )\n",
    "print(\"uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(\n",
    "        len(no_rnaseq_yet))\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# output the csv to \n",
    "pd.Series(no_rnaseq_yet).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/se-missing.txt',index=None)\n",
    "pd.Series(no_rnaseq).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/se-nonexistant.txt',index=None)\n",
    "no_rnaseq_yet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the 5p SE densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"pos_splicing_suffix = '.positive.nr.txt'\n",
    "neg_splicing_suffix = '.negative.nr.txt'\n",
    "\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.5p.txt'\n",
    "annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "background_control_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/se/'\n",
    "output_dir = '/projects/ps-yeolab3/bay001/maps/current/se_nr_5p/'\n",
    "bash_scripts_dir = '/home/bay001/projects/maps_20160420/scripts/bash_scripts/'\n",
    "\n",
    "k562_background_ce = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.strict_CE_all_20170401')\n",
    "k562_background_nse_all = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEall_0.5_20170401')\n",
    "k562_background_nse_inc = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEincl_0.5_20170401')\n",
    "k562_background_nse_exc = os.path.join(background_control_dir, 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEexcl_0.5_20170401')\n",
    "\n",
    "hepg2_background_ce = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.strict_CE_all_20170401')\n",
    "hepg2_background_nse_all = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEall_0.5_20170401')\n",
    "hepg2_background_nse_inc = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEincl_0.5_20170401')\n",
    "hepg2_background_nse_exc = os.path.join(background_control_dir, 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nSEexcl_0.5_20170401')\n",
    "\n",
    "img_ext = 'png'\n",
    "event = 'se'\n",
    "\n",
    "no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "cmds = []\n",
    "\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "for uid in clip_df['uID']:\n",
    "    \n",
    "    metadata = defaultdict()\n",
    "    metadata[uid] = defaultdict()\n",
    "    metadata[uid]['rep1'] = defaultdict()\n",
    "    metadata[uid]['rep2'] = defaultdict()\n",
    "    \n",
    "    \n",
    "    r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "    \n",
    "    if cell == 'K562':\n",
    "        background = k562_background_ce\n",
    "    elif cell == 'HepG2':\n",
    "        background = hepg2_background_ce\n",
    "    else:\n",
    "        print(cell)\n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(\n",
    "        rnaseq_manifests, rbp, cell\n",
    "    )\n",
    "    if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "        no_rnaseq.append(uid)\n",
    "    else:\n",
    "        positive, negative = get_annotations_from_splicing_prefix(\n",
    "            annotation_dir, splicing_prefix, \n",
    "            pos_splicing_suffix, neg_splicing_suffix\n",
    "        )\n",
    "        if(positive == None or negative == None):\n",
    "            no_rnaseq_yet.append(uid)\n",
    "        else:\n",
    "            for r in [r1, r2]:\n",
    "                name = os.path.basename(r).replace('.bam','.{}'.format(img_ext))\n",
    "                output_filename = os.path.join(\n",
    "                    output_dir,\n",
    "                    name\n",
    "                )\n",
    "\n",
    "                # Build the cmd line\n",
    "                cmd = \"python \" + density_runner\n",
    "                cmd = cmd + \" --event {}\".format(event)\n",
    "                cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                cmd = cmd + \" --annotations {} {} {}\".format(\n",
    "                    positive, negative, background\n",
    "                )\n",
    "                cmd = cmd + \" --annotation_type {} {} {}\".format(\n",
    "                    'rmats', 'rmats', 'eric',\n",
    "                )\n",
    "                cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                cmd = cmd + \" --unflip\"\n",
    "                cmds.append(cmd)\n",
    "\n",
    "                # Build the json dict\n",
    "                rep = 'rep1' if r == r1 else 'rep2'\n",
    "\n",
    "                metadata[uid][rep]['annotations'] = list()\n",
    "                metadata[uid][rep]['annotation_type'] = list()\n",
    "\n",
    "                metadata[uid][rep]['cell'] = cell\n",
    "                metadata[uid][rep]['rbp'] = rbp\n",
    "                metadata[uid][rep]['date'] = current_date\n",
    "                metadata[uid][rep]['ip_bam'] = {'class':'File', 'path':r}\n",
    "\n",
    "                metadata[uid][rep]['input_bam'] = {'class':'File', 'path':i}\n",
    "\n",
    "                metadata[uid][rep]['chrom_sizes'] = {'class':'File', 'path':chrom_sizes}\n",
    "                metadata[uid][rep]['unflip'] = False\n",
    "                metadata[uid][rep]['event'] = 'se'\n",
    "                metadata[uid][rep]['exon_offset'] = 50\n",
    "                metadata[uid][rep]['intron_offset'] = 300\n",
    "\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':positive})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':negative})\n",
    "                metadata[uid][rep]['annotations'].append({'class':'File', 'path':background})\n",
    "\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "                metadata[uid][rep]['annotation_type'].append('rmats')\n",
    "                metadata[uid][rep]['annotation_type'].append('eric')\n",
    "\n",
    "                with open(\n",
    "                    os.path.join(\n",
    "                        bash_scripts_dir,\"{}.{}.{}.{}.{}.{}.json\".format(\n",
    "                            uid, rep, rbp, cell, current_date, event\n",
    "                        )\n",
    "                    ),\n",
    "                    'w'\n",
    "                ) as f:\n",
    "                    f.write('#!/usr/bin/env make_rbp_map.cwl\\n')\n",
    "                    json.dump(metadata[uid][rep], f)\n",
    "Submitter(\n",
    "    cmds, \n",
    "    \"SE_NR_5p_{}\".format(img_extension), \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/SE_NR_5p_{}.sh'.format(img_extension),\n",
    "    submit=False,\n",
    "    array=True,\n",
    "    walltime='2:00:00'\n",
    ")\n",
    "print(\"uIDs for which we don't have splicing data for: {}\".format(\n",
    "        len(no_rnaseq))\n",
    "     )\n",
    "print(\"uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(\n",
    "        len(no_rnaseq_yet))\n",
    "     )\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peak_dir = '/home/elvannostrand/data/clip/CLIPseq_analysis/ENCODE_CLIPperv2_20161120/'\n",
    "miso_annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/as_miso_renamed'\n",
    "program_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_peak.py'\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "\n",
    "outpath = '/projects/ps-yeolab3/bay001/maps/current/se_nr_peak/'\n",
    "img_ext = 'png'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "event = 'se'\n",
    "clip_df = pd.read_table(clip_manifest)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8355403[].tscc-mgr.local\n",
      "uIDs for which we don't have splicing data for: 46\n",
      "uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: 14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 46 tasks as an array-job.\n",
      "Wrote commands to /projects/ps-yeolab3/bay001/maps/bash_scripts/SE_PEAK_PNGS2.sh.\n",
      "Submitted script to queue home.\n",
      " Job ID: 8355403\n"
     ]
    }
   ],
   "source": [
    "all_peaks = glob.glob(os.path.join(peak_dir,'*.compressed.bed'))\n",
    "no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "cmds = []\n",
    "\n",
    "for peak in all_peaks:\n",
    "    outfile = os.path.join(\n",
    "        outpath,\n",
    "        os.path.basename(peak).replace('.compressed.bed','.compressed.{}'.format(img_ext))\n",
    "    )\n",
    "    uid, rep, other = split_uid_and_rep(os.path.basename(peak).split('.')[0])\n",
    "    _, _, _, rbp, cell = get_clip_file_from_uid(\n",
    "        clip_df, uid\n",
    "    )\n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(\n",
    "        rnaseq_manifests, rbp, cell\n",
    "    )\n",
    "    \n",
    "    if cell == 'HepG2':\n",
    "        ce_background = os.path.join(miso_annotation_dir,'HepG2-constitutive-exons.miso')\n",
    "        nc_background = os.path.join(miso_annotation_dir,'HepG2-native-cassette-exons.miso')\n",
    "        ni_background = os.path.join(miso_annotation_dir,'HepG2-native-included-exons.miso')\n",
    "        ne_background = os.path.join(miso_annotation_dir,'HepG2-native-excluded-exons.miso')\n",
    "    elif cell == 'K562':\n",
    "        ce_background = os.path.join(miso_annotation_dir,'K562-constitutive-exons.miso')\n",
    "        nc_background = os.path.join(miso_annotation_dir,'K562-native-cassette-exons.miso')\n",
    "        ni_background = os.path.join(miso_annotation_dir,'K562-native-included-exons.miso')\n",
    "        ne_background = os.path.join(miso_annotation_dir,'K562-native-excluded-exons.miso')\n",
    "    else:\n",
    "        print('error')\n",
    "        \n",
    "    if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "        no_rnaseq.append(uid)\n",
    "    else:\n",
    "        positive, negative = get_annotations_from_splicing_prefix(\n",
    "            miso_annotation_dir, splicing_prefix, \n",
    "            \n",
    "        )\n",
    "    if(positive == None or negative == None):\n",
    "        no_rnaseq_yet.append(uid)\n",
    "    elif not os.path.exists(outfile):\n",
    "        cmd = \"python {} -i {} -o {} -m {} {} {} {} {} {}\".format(\n",
    "            program_runner,\n",
    "            peak,\n",
    "            outfile,\n",
    "            positive,\n",
    "            negative,\n",
    "            ce_background,\n",
    "            nc_background,\n",
    "            ni_background,\n",
    "            ne_background\n",
    "        )\n",
    "        cmds.append(cmd)\n",
    "\n",
    "Submitter(\n",
    "    cmds, \n",
    "    \"SE_PEAK_PNGS\", \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/SE_PEAK_PNGS2.sh',\n",
    "    submit=True,\n",
    "    array=True,\n",
    "    walltime='2:00:00'\n",
    ")\n",
    "print(\"uIDs for which we don't have splicing data for: {}\".format(\n",
    "        len(no_rnaseq))\n",
    "     )\n",
    "print(\"uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(\n",
    "        len(no_rnaseq_yet))\n",
    "     )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the rest of the splice events\n",
    "- The first cell plots all the background events + positive (included) and negative (excluded) events separately.\n",
    "- The second cell plots a single background event + positive (included) + negative (excluded) events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"density_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_density.py'\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "events = {\n",
    "    # 'mxe':'/projects/ps-yeolab3/bay001/maps/current_annotations/mxe/',\n",
    "    'a3ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a3ss/',\n",
    "    'a5ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a5ss/',\n",
    "    # 'ri':'/projects/ps-yeolab3/bay001/maps/current_annotations/ri/'\n",
    "}\n",
    "\n",
    "\n",
    "img_extension = 'png'\n",
    "out_base = '/projects/ps-yeolab3/bay001/maps/current/'\n",
    "pos_splicing_suffix = 'positive.nr.txt'\n",
    "neg_splicing_suffix = 'negative.nr.txt'\n",
    "\n",
    "for event, annotation_dir in events.iteritems():\n",
    "    no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "    no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "    cmds = []\n",
    "    output_dir = os.path.join(out_base, '{}_nr'.format(event))\n",
    "    \n",
    "    for uid in clip_df['uID']:\n",
    "        r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "\n",
    "        if cell == 'K562':\n",
    "            background_basic = os.path.join(annotation_dir, 'K562-background-basic.txt')\n",
    "            background_center = os.path.join(annotation_dir, 'K562-background-center.txt')\n",
    "            background_extension = os.path.join(annotation_dir, 'K562-background-extension.txt')\n",
    "        elif cell == 'HepG2':\n",
    "            background_basic = os.path.join(annotation_dir, 'HepG2-background-basic.txt')\n",
    "            background_center = os.path.join(annotation_dir, 'HepG2-background-center.txt')\n",
    "            background_extension = os.path.join(annotation_dir, 'HepG2-background-extension.txt')\n",
    "        else:\n",
    "            print(cell)\n",
    "            \n",
    "        splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp, cell)\n",
    "        if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "            no_rnaseq.append(uid)\n",
    "        else:\n",
    "            positive, negative = get_annotations_from_splicing_prefix(\n",
    "                annotation_dir, \n",
    "                splicing_prefix,\n",
    "                pos_splicing_suffix=pos_splicing_suffix,\n",
    "                neg_splicing_suffix=neg_splicing_suffix\n",
    "            )\n",
    "\n",
    "        for r in [r1, r2]:\n",
    "            name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "            output_filename = os.path.join(\n",
    "                output_dir,\n",
    "                name\n",
    "            )\n",
    "            cmd = \"python \" + density_runner\n",
    "            cmd = cmd + \" --event {}\".format(event)\n",
    "            cmd = cmd + \" --ipbam {}\".format(r)\n",
    "            cmd = cmd + \" --inputbam {}\".format(i)\n",
    "            cmd = cmd + \" --output {}\".format(output_filename)\n",
    "            if positive is not None and negative is not None:\n",
    "                cmd = cmd + \" --annotations {} {} {}\".format(\n",
    "                    positive, negative, background_basic\n",
    "                )\n",
    "                cmd = cmd + \" --annotation_type {} {} {}\".format(\n",
    "                    'rmats', 'rmats', 'eric'\n",
    "                )\n",
    "            cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "            cmds.append(cmd)\n",
    "    Submitter(\n",
    "        cmds, \n",
    "        \"{}_NR_{}\".format(event, img_extension), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_NR_{}.sh'.format(\n",
    "            event, img_extension\n",
    "        ),\n",
    "        submit=True,\n",
    "        array=True,\n",
    "        walltime='4:00:00'\n",
    "    )\n",
    "# print(\"{} {} uIDs for which we don't have splicing data for: {}\".format(splice_condition, event, len(no_rnaseq)))\n",
    "# print(\"{} {} uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(splice_condition, event, len(no_rnaseq_yet)))\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the A3SS/A5SS splice events (positive and negative and controls) all together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a5ss_k562_all = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSall_0.5_20170401'\n",
    "a5ss_k562_basic = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSbasic_0.5_20170401'\n",
    "a5ss_k562_center = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SScenter_0.5_20170401'\n",
    "a5ss_k562_extension = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSextension_0.5_20170401'\n",
    "\n",
    "a3ss_k562_all = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSall_0.5_20170401'\n",
    "a3ss_k562_basic = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSbasic_0.5_20170401'\n",
    "a3ss_k562_center = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SScenter_0.5_20170401'\n",
    "a3ss_k562_extension = 'k562_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSextension_0.5_20170401'\n",
    "\n",
    "a5ss_hepg2_all = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSall_0.5_20170401'\n",
    "a5ss_hepg2_basic = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSbasic_0.5_20170401'\n",
    "a5ss_hepg2_center = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SScenter_0.5_20170401'\n",
    "a5ss_hepg2_extension = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA5SSextension_0.5_20170401'\n",
    "\n",
    "a3ss_hepg2_all = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSall_0.5_20170401'\n",
    "a3ss_hepg2_basic = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSbasic_0.5_20170401'\n",
    "a3ss_hepg2_center = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SScenter_0.5_20170401'\n",
    "a3ss_hepg2_extension = 'hepg2_brenton-graveley_ambiguous_bams_for_integrated_analysis.txt.nA3SSextension_0.5_20170401'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "density_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_density.py'\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "events = {\n",
    "    'a3ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a3ss/',\n",
    "    'a5ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a5ss/',\n",
    "}\n",
    "\n",
    "\n",
    "img_extension = 'svg'\n",
    "out_base = '/projects/ps-yeolab3/bay001/maps/current/'\n",
    "pos_splicing_suffix = 'positive.nr.txt'\n",
    "neg_splicing_suffix = 'negative.nr.txt'\n",
    "\n",
    "for event, annotation_dir in events.iteritems():\n",
    "    no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "    no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "    cmds = []\n",
    "    output_dir = os.path.join(out_base, '{}_nr'.format(event))\n",
    "    for uid in clip_df['uID']:\n",
    "        r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "        \n",
    "        if cell == 'K562':\n",
    "            if event == 'a3ss':\n",
    "                background_all = os.path.join(annotation_dir, a3ss_k562_all)\n",
    "                background_basic = os.path.join(annotation_dir, a3ss_k562_basic)\n",
    "                background_center = os.path.join(annotation_dir, a3ss_k562_center)\n",
    "                background_extension = os.path.join(annotation_dir, a3ss_k562_extension)\n",
    "            elif event == 'a5ss':\n",
    "                background_all = os.path.join(annotation_dir, a5ss_k562_all)\n",
    "                background_basic = os.path.join(annotation_dir, a5ss_k562_basic)\n",
    "                background_center = os.path.join(annotation_dir, a5ss_k562_center)\n",
    "                background_extension = os.path.join(annotation_dir, a5ss_k562_extension)\n",
    "            else:\n",
    "                print(event)\n",
    "        elif cell == 'HepG2':\n",
    "            if event == 'a3ss':\n",
    "                background_all = os.path.join(annotation_dir, a3ss_hepg2_all)\n",
    "                background_basic = os.path.join(annotation_dir, a3ss_hepg2_basic)\n",
    "                background_center = os.path.join(annotation_dir, a3ss_hepg2_center)\n",
    "                background_extension = os.path.join(annotation_dir, a3ss_hepg2_extension)\n",
    "            elif event == 'a5ss':\n",
    "                background_all = os.path.join(annotation_dir, a5ss_hepg2_all)\n",
    "                background_basic = os.path.join(annotation_dir, a5ss_hepg2_basic)\n",
    "                background_center = os.path.join(annotation_dir, a5ss_hepg2_center)\n",
    "                background_extension = os.path.join(annotation_dir, a5ss_hepg2_extension)\n",
    "            else:\n",
    "                print(event)\n",
    "        else:\n",
    "            print(cell)\n",
    "            \n",
    "        splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp, cell)\n",
    "        if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "            no_rnaseq.append(uid)\n",
    "        else:\n",
    "            positive, negative = get_annotations_from_splicing_prefix(\n",
    "                annotation_dir, \n",
    "                splicing_prefix,\n",
    "                pos_splicing_suffix=pos_splicing_suffix,\n",
    "                neg_splicing_suffix=neg_splicing_suffix\n",
    "            )\n",
    "            if(positive == None or negative == None):\n",
    "                no_rnaseq_yet.append(uid)\n",
    "            else:\n",
    "                if not (rbp in positive and rbp in negative):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                pos_prefix = os.path.basename(positive).split('-')[0]\n",
    "                neg_prefix = os.path.basename(negative).split('-')[0]\n",
    "                if not (pos_prefix in rbp and neg_prefix in rbp):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                for r in [r1, r2]:\n",
    "                    name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "                    output_filename = os.path.join(\n",
    "                        output_dir,\n",
    "                        name\n",
    "                    )\n",
    "                    cmd = \"python \" + density_runner\n",
    "                    cmd = cmd + \" --event {}\".format(event)\n",
    "                    cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                    cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                    cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                    if positive is not None and negative is not None:\n",
    "                        cmd = cmd + \" --annotations {} {} {} {} {} {}\".format(\n",
    "                            positive, negative, background_all, background_basic, background_center, background_extension\n",
    "                        )\n",
    "                        cmd = cmd + \" --annotation_type {} {} {} {} {} {}\".format(\n",
    "                            'rmats', 'rmats', 'eric', 'eric', 'eric', 'eric'\n",
    "                        )\n",
    "                    cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                    cmds.append(cmd)\n",
    "            # if(uid == '228'):\n",
    "            #     print(r1, r2, i, rbp, cell, annotation_dir, splicing_prefix, pos_splicing_suffix, neg_splicing_suffix)\n",
    "    \n",
    "    Submitter(\n",
    "        cmds, \n",
    "        \"{}_NR_{}\".format(event, img_extension), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_NR_{}.sh'.format(\n",
    "            event, img_extension\n",
    "        ),\n",
    "        submit=True,\n",
    "        array=True,\n",
    "        walltime='4:00:00'\n",
    "    )\n",
    "    # output the csv to missing folder \n",
    "    pd.Series(no_rnaseq_yet).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/{}-missing.txt'.format(event),index=None)\n",
    "    pd.Series(no_rnaseq).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/{}-nonexistant.txt'.format(event),index=None)\n",
    "\n",
    "# print(\"{} {} uIDs for which we don't have splicing data for: {}\".format(splice_condition, event, len(no_rnaseq)))\n",
    "# print(\"{} {} uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(splice_condition, event, len(no_rnaseq_yet)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a5ss_k562_all = 'K562-all-native-a5ss-events'\n",
    "a5ss_k562_basic = 'K562-shorter-isoform-in-majority-of-controls'\n",
    "a5ss_k562_center = 'K562-mixed-psi-isoform-in-majority-of-controls'\n",
    "a5ss_k562_extension = 'K562-longer-isoform-in-majority-of-controls'\n",
    "\n",
    "a3ss_k562_all = 'K562-all-native-a3ss-events'\n",
    "a3ss_k562_basic = 'K562-shorter-isoform-in-majority-of-controls'\n",
    "a3ss_k562_center = 'K562-mixed-psi-isoform-in-majority-of-controls'\n",
    "a3ss_k562_extension = 'K562-longer-isoform-in-majority-of-controls'\n",
    "\n",
    "a5ss_hepg2_all = 'HepG2-all-native-a5ss-events'\n",
    "a5ss_hepg2_basic = 'HepG2-shorter-isoform-in-majority-of-controls'\n",
    "a5ss_hepg2_center = 'HepG2-mixed-psi-isoform-in-majority-of-controls'\n",
    "a5ss_hepg2_extension = 'HepG2-longer-isoform-in-majority-of-controls'\n",
    "\n",
    "a3ss_hepg2_all = 'HepG2-all-native-a3ss-events'\n",
    "a3ss_hepg2_basic = 'HepG2-shorter-isoform-in-majority-of-controls'\n",
    "a3ss_hepg2_center = 'HepG2-mixed-psi-isoform-in-majority-of-controls'\n",
    "a3ss_hepg2_extension = 'HepG2-longer-isoform-in-majority-of-controls'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "density_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_density.py'\n",
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "events = {\n",
    "    'a3ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a3ss_renamed/',\n",
    "    'a5ss':'/projects/ps-yeolab3/bay001/maps/current_annotations/a5ss_renamed/',\n",
    "}\n",
    "\n",
    "\n",
    "img_extension = 'png'\n",
    "out_base = '/projects/ps-yeolab3/bay001/maps/current/'\n",
    "pos_splicing_suffix = '-longer-isoform-included-upon-knockdown'\n",
    "neg_splicing_suffix = '-shorter-isoform-included-upon-knockdown'\n",
    "\n",
    "for event, annotation_dir in events.iteritems():\n",
    "    no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "    no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "    cmds = []\n",
    "    output_dir = os.path.join(out_base, '{}_nr'.format(event))\n",
    "    for uid in clip_df['uID']:\n",
    "        r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "        \n",
    "        if cell == 'K562':\n",
    "            if event == 'a3ss':\n",
    "                background_all = os.path.join(annotation_dir, a3ss_k562_all)\n",
    "                background_basic = os.path.join(annotation_dir, a3ss_k562_basic)\n",
    "                background_center = os.path.join(annotation_dir, a3ss_k562_center)\n",
    "                background_extension = os.path.join(annotation_dir, a3ss_k562_extension)\n",
    "            elif event == 'a5ss':\n",
    "                background_all = os.path.join(annotation_dir, a5ss_k562_all)\n",
    "                background_basic = os.path.join(annotation_dir, a5ss_k562_basic)\n",
    "                background_center = os.path.join(annotation_dir, a5ss_k562_center)\n",
    "                background_extension = os.path.join(annotation_dir, a5ss_k562_extension)\n",
    "            else:\n",
    "                print(event)\n",
    "        elif cell == 'HepG2':\n",
    "            if event == 'a3ss':\n",
    "                background_all = os.path.join(annotation_dir, a3ss_hepg2_all)\n",
    "                background_basic = os.path.join(annotation_dir, a3ss_hepg2_basic)\n",
    "                background_center = os.path.join(annotation_dir, a3ss_hepg2_center)\n",
    "                background_extension = os.path.join(annotation_dir, a3ss_hepg2_extension)\n",
    "            elif event == 'a5ss':\n",
    "                background_all = os.path.join(annotation_dir, a5ss_hepg2_all)\n",
    "                background_basic = os.path.join(annotation_dir, a5ss_hepg2_basic)\n",
    "                background_center = os.path.join(annotation_dir, a5ss_hepg2_center)\n",
    "                background_extension = os.path.join(annotation_dir, a5ss_hepg2_extension)\n",
    "            else:\n",
    "                print(event)\n",
    "        else:\n",
    "            print(cell)\n",
    "            \n",
    "        splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp, cell)\n",
    "        if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "            no_rnaseq.append(uid)\n",
    "        else:\n",
    "            positive, negative = get_annotations_from_splicing_prefix(\n",
    "                annotation_dir, \n",
    "                splicing_prefix,\n",
    "                pos_splicing_suffix=pos_splicing_suffix,\n",
    "                neg_splicing_suffix=neg_splicing_suffix\n",
    "            )\n",
    "            if(positive == None or negative == None):\n",
    "                no_rnaseq_yet.append(uid)\n",
    "            else:\n",
    "                if not (rbp in positive and rbp in negative):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                pos_prefix = os.path.basename(positive).split('-')[0]\n",
    "                neg_prefix = os.path.basename(negative).split('-')[0]\n",
    "                if not (pos_prefix in rbp and neg_prefix in rbp):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                for r in [r1, r2]:\n",
    "                    name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "                    output_filename = os.path.join(\n",
    "                        output_dir,\n",
    "                        name\n",
    "                    )\n",
    "                    cmd = \"python \" + density_runner\n",
    "                    cmd = cmd + \" --event {}\".format(event)\n",
    "                    cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                    cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                    cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                    if positive is not None and negative is not None:\n",
    "                        cmd = cmd + \" --annotations {} {} {} {} {} {}\".format(\n",
    "                            positive, negative, background_all, background_basic, background_center, background_extension\n",
    "                        )\n",
    "                        cmd = cmd + \" --annotation_type {} {} {} {} {} {}\".format(\n",
    "                            'rmats', 'rmats', 'eric', 'eric', 'eric', 'eric'\n",
    "                        )\n",
    "                    cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                    cmds.append(cmd)\n",
    "            # if(uid == '228'):\n",
    "            #     print(r1, r2, i, rbp, cell, annotation_dir, splicing_prefix, pos_splicing_suffix, neg_splicing_suffix)\n",
    "    \n",
    "    Submitter(\n",
    "        cmds, \n",
    "        \"{}_NR_{}\".format(event, img_extension), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_NR_{}.sh'.format(\n",
    "            event, img_extension\n",
    "        ),\n",
    "        submit=False,\n",
    "        array=True,\n",
    "        walltime='4:00:00'\n",
    "    )\n",
    "    # output the csv to missing folder \n",
    "    pd.Series(no_rnaseq_yet).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/{}-missing.txt'.format(event),index=None)\n",
    "    pd.Series(no_rnaseq).to_csv('/home/bay001/projects/encode/analysis/tests/missing_rnaseq/{}-nonexistant.txt'.format(event),index=None)\n",
    "\n",
    "# print(\"{} {} uIDs for which we don't have splicing data for: {}\".format(splice_condition, event, len(no_rnaseq)))\n",
    "# print(\"{} {} uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(splice_condition, event, len(no_rnaseq_yet)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot RI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "density_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_density.py'\n",
    "\n",
    "events = {\n",
    "    'ri':'/projects/ps-yeolab3/bay001/maps/current_annotations/ri_renamed/'\n",
    "}\n",
    "\n",
    "\n",
    "img_extension = 'png'\n",
    "out_base = '/projects/ps-yeolab3/bay001/maps/current/'\n",
    "pos_splicing_suffix = '-included-upon-knockdown'\n",
    "neg_splicing_suffix = '-excluded-upon-knockdown'\n",
    "\n",
    "for event, annotation_dir in events.iteritems():\n",
    "    no_rnaseq = [] # uIDs for which we don't have rna seq expt ids for\n",
    "    no_rnaseq_yet = [] # uIDs for which we have an expt id, but haven't downloaded the data yet\n",
    "    cmds = []\n",
    "    output_dir = os.path.join(out_base, '{}_nr'.format(event))\n",
    "    for uid in clip_df['uID']:\n",
    "        r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "        if cell == 'HepG2':\n",
    "            background1 = os.path.join(annotation_dir, 'HepG2-constitutive-introns')\n",
    "            background2 = os.path.join(annotation_dir, 'HepG2-all-retained-introns')\n",
    "            background3 = os.path.join(annotation_dir, 'HepG2-greater-than-50-percent-spliced')\n",
    "            background4 = os.path.join(annotation_dir, 'HepG2-greater-than-50-percent-retained')\n",
    "        elif cell == 'K562':\n",
    "            background1 = os.path.join(annotation_dir, 'K562-constitutive-introns')\n",
    "            background2 = os.path.join(annotation_dir, 'K562-all-retained-introns')\n",
    "            background3 = os.path.join(annotation_dir, 'K562-greater-than-50-percent-spliced')\n",
    "            background4 = os.path.join(annotation_dir, 'K562-greater-than-50-percent-retained')\n",
    "        else:\n",
    "            print(cell)\n",
    "        splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp, cell)\n",
    "        if(splicing_prefix == \"NO_RNASEQ\"): # we don't have an rna seq expt for this clip:\n",
    "            no_rnaseq.append(uid)\n",
    "        else:\n",
    "            positive, negative = get_annotations_from_splicing_prefix(\n",
    "                annotation_dir, \n",
    "                splicing_prefix,\n",
    "                pos_splicing_suffix=pos_splicing_suffix,\n",
    "                neg_splicing_suffix=neg_splicing_suffix\n",
    "            )\n",
    "            if positive is not None and negative is not None:\n",
    "                if not (rbp in positive and rbp in negative):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                pos_prefix = os.path.basename(positive).split('-')[0]\n",
    "                neg_prefix = os.path.basename(negative).split('-')[0]\n",
    "                if not (pos_prefix in rbp and neg_prefix in rbp):\n",
    "                    print(\n",
    "                        'warning, these dont match: {}, {}, {}'.format(\n",
    "                            rbp, \n",
    "                            os.path.basename(positive),\n",
    "                            os.path.basename(negative)\n",
    "                        )\n",
    "                    )\n",
    "                for r in [r1, r2]:\n",
    "                    name = os.path.basename(r).replace('.bam','.{}'.format(img_extension))\n",
    "                    output_filename = os.path.join(\n",
    "                        output_dir,\n",
    "                        name\n",
    "                    )\n",
    "                    cmd = \"python \" + density_runner\n",
    "                    cmd = cmd + \" --event {}\".format(event)\n",
    "                    cmd = cmd + \" --ipbam {}\".format(r)\n",
    "                    cmd = cmd + \" --inputbam {}\".format(i)\n",
    "                    cmd = cmd + \" --output {}\".format(output_filename)\n",
    "                    if positive is not None and negative is not None:\n",
    "                        cmd = cmd + \" --annotations {} {} {} {} {} {}\".format(\n",
    "                            positive, negative, background1, background2, background3, background4\n",
    "                        )\n",
    "                        cmd = cmd + \" --annotation_type {} {} {} {} {} {}\".format(\n",
    "                            'rmats', 'rmats', 'eric', 'eric', 'eric', 'eric', \n",
    "                        )\n",
    "                    cmd = cmd + \" --chrom_sizes {}\".format(chrom_sizes)\n",
    "                    cmds.append(cmd)\n",
    "    Submitter(\n",
    "        cmds, \n",
    "        \"{}_NR_{}\".format(event, img_extension), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_NR_{}.sh'.format(\n",
    "            event, img_extension\n",
    "        ),\n",
    "        submit=False,\n",
    "        array=True,\n",
    "        walltime='4:00:00'\n",
    "    )\n",
    "print(\"{} uIDs for which we don't have splicing data for: {}\".format(event, len(no_rnaseq)))\n",
    "print(\"{} uIDs for which we have an RNASEQ ID, but we don't have the splicing data yet: {}\".format(event, len(no_rnaseq_yet)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "o = open('/home/bay001/projects/encode/analysis/run_ri_quickly.sh','w')\n",
    "for cmd in cmds:\n",
    "    o.write(cmd + '\\n')\n",
    "    \n",
    "o.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Plot the IDR peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('246', u'AUH', u'K562')\n",
      "('543', u'AGGF1', u'HepG2')\n",
      "('531', u'RPS5', u'K562')\n",
      "('417', u'POLR2G', u'K562')\n",
      "('358', u'GNL3', u'K562')\n",
      "('256', u'SLTM', u'K562')\n",
      "('540', u'RPS3', u'HepG2')\n",
      "('516', u'NOLC1', u'K562')\n",
      "('553', u'NOLC1', u'HepG2')\n",
      "('491', u'EXOSC5', u'K562')\n",
      "('684', u'ZC3H11A', u'K562')\n",
      "('415', u'TNRC6A', u'HepG2')\n",
      "('332', u'RBM27', u'K562')\n",
      "('444', u'LSM11', u'K562')\n",
      "('570', u'ZNF622', u'K562')\n",
      "('529', u'RPS24', u'K562')\n",
      "('614', u'RPS11', u'K562')\n",
      "('341', u'DGCR8', u'K562')\n",
      "('484', u'POLR2G', u'HepG2')\n",
      "('460', u'FASTKD2', u'K562')\n",
      "('275', u'CSTF2T', u'K562')\n",
      "('367', u'YWHAG', u'K562')\n",
      "('383', u'DGCR8', u'HepG2')\n",
      "('545', u'EIF3H', u'HepG2')\n",
      "('241', u'DDX42', u'K562')\n",
      "('390x4000', u'DDX20', u'HepG2')\n",
      "('641', u'ZC3H11A', u'HepG2')\n",
      "('679', u'RPL23A', u'HepG2')\n",
      "('366', u'TNRC6A', u'K562')\n",
      "('279', u'FAM120A', u'K562')\n"
     ]
    }
   ],
   "source": [
    "clip_manifest = '/home/bay001/projects/maps_20160420/permanent_data/ALLDATASETS_submittedonly.txt'\n",
    "clip_df = pd.read_table(clip_manifest)\n",
    "miso_annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/as_miso_renamed'\n",
    "idr_peak_dir = '/home/elvannostrand/data/clip/CLIPseq_analysis/ENCODE_FINALforpapers_20170325/IDR/'\n",
    "idr_peak_outdir = '/projects/ps-yeolab3/bay001/maps/current/idr_peaks_bg'\n",
    "program_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_peak.py'\n",
    "all_peaks = glob.glob(os.path.join(idr_peak_dir,'*0102merged.bed'))\n",
    "\n",
    "progress = tnrange(len(all_peaks))\n",
    "cmds = []\n",
    "for peak in all_peaks:\n",
    "    outfile = os.path.join(\n",
    "        idr_peak_outdir,\n",
    "        os.path.basename(peak).replace('merged.bed','merged.png')\n",
    "    )\n",
    "    uid = os.path.basename(peak).split('.')[0]\n",
    "    _, _, _, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "    splicing_prefix = get_rnaseq_splicing_prefix_from_rbpname(rnaseq_manifests, rbp, cell)\n",
    "    # print(splicing_prefix)\n",
    "    positive, negative = get_annotations_from_splicing_prefix(miso_annotation_dir, splicing_prefix)\n",
    "    \n",
    "    if positive is not None and negative is not None:\n",
    "        if cell == 'HepG2':\n",
    "            ce_background = os.path.join(miso_annotation_dir,'HepG2-constitutive-exons.miso')\n",
    "            nc_background = os.path.join(miso_annotation_dir,'HepG2-native-cassette-exons.miso')\n",
    "            ni_background = os.path.join(miso_annotation_dir,'HepG2-native-included-exons.miso')\n",
    "            ne_background = os.path.join(miso_annotation_dir,'HepG2-native-excluded-exons.miso')\n",
    "        elif cell == 'K562':\n",
    "            ce_background = os.path.join(miso_annotation_dir,'K562-constitutive-exons.miso')\n",
    "            nc_background = os.path.join(miso_annotation_dir,'K562-native-cassette-exons.miso')\n",
    "            ni_background = os.path.join(miso_annotation_dir,'K562-native-included-exons.miso')\n",
    "            ne_background = os.path.join(miso_annotation_dir,'K562-native-excluded-exons.miso')\n",
    "        else:\n",
    "            print('error')\n",
    "        cmd = \"python {} -i {} -o {} -m {} {} {} {} {} {}\".format(\n",
    "            program_runner,\n",
    "            peak,\n",
    "            outfile,\n",
    "            positive,\n",
    "            negative,\n",
    "            ce_background,\n",
    "            nc_background,\n",
    "            ni_background,\n",
    "            ne_background\n",
    "        )\n",
    "        # ! $cmd ####################################\n",
    "        cmds.append(cmd)\n",
    "    else:\n",
    "        print(uid, rbp, cell)\n",
    "    \n",
    "    progress.update(1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 151 tasks as an array-job.\n",
      "Wrote commands to /projects/ps-yeolab3/bay001/maps/bash_scripts/IDR_PNGS.sh.\n",
      "Submitted script to queue home-scrm.\n",
      " Job ID: 8400286\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8400286[].tscc-mgr.local\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<qtools.submitter.Submitter at 0x2abc7e63db10>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r",
      "          \r",
      "181/|/100%|| 181/181 [00:13<00:00, 51.20it/s]"
     ]
    }
   ],
   "source": [
    "Submitter(\n",
    "    cmds, \n",
    "    'idr_bg_peaks', \n",
    "    sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/IDR_PNGS.sh',\n",
    "    submit=True,\n",
    "    array=True,\n",
    "    walltime='2:00:00',\n",
    "    queue='home-scrm'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peak_dir = '/home/elvannostrand/data/clip/CLIPseq_analysis/ENCODE_CLIPperv2_20161120/'\n",
    "annotation_dir = '/projects/ps-yeolab3/bay001/maps/current_annotations/'\n",
    "program_runner = '/home/bay001/projects/codebase/rbp-maps/maps/plot_density.py'\n",
    "outpath = '/projects/ps-yeolab3/bay001/maps/current/'\n",
    "chrom_sizes = '/projects/ps-yeolab/genomes/hg19/hg19.chrom.sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "events = ['all_txStart','all_txStop','all_cdsStart']\n",
    "\n",
    "for event in events:\n",
    "    subtract_cmds = []\n",
    "    entropy_cmds = []\n",
    "    rawip_cmds = []\n",
    "    input_cmds = []\n",
    "    \n",
    "    output_dir = os.path.join(outpath, event)\n",
    "    for uid in clip_df['uID']:\n",
    "        r1, r2, i, rbp, cell = get_clip_file_from_uid(clip_df, uid)\n",
    "        if cell == 'k562':\n",
    "            annotation = os.path.join(annotation_dir, '{}_k562.bed'.format(event))\n",
    "        else:\n",
    "            annotation = os.path.join(annotation_dir, '{}_hepg2.bed'.format(event))\n",
    "\n",
    "        for r in [r1, r2]:\n",
    "            subtract_name = os.path.basename(r).replace('.bam','.norm_subtract.png')\n",
    "            entropy_name = os.path.basename(r).replace('.bam','.norm_entropy.png')\n",
    "            ip_name = os.path.basename(r).replace('.bam','.no_norm.png')\n",
    "            inp_name = os.path.basename(r).replace('.bam','.input.png')\n",
    "            \n",
    "            output_filename_subtract = os.path.join(output_dir, subtract_name)\n",
    "            output_filename_entropy = os.path.join(output_dir, entropy_name)\n",
    "            output_filename_ip = os.path.join(output_dir, ip_name)\n",
    "            output_filename_inp = os.path.join(output_dir, inp_name)\n",
    "            \n",
    "            cmd = \"python \" + program_runner\n",
    "            cmd = cmd + \" --event bed --ipbam {} --inputbam {} --annotations {} --annotation_type {} --intron_offset {} --exon_offset {} --scale --chrom_sizes {} \".format(\n",
    "                r, i, annotation, 'bed', 0, 0, chrom_sizes\n",
    "            )\n",
    "            \n",
    "            subtract_cmd = cmd + '--normalization_level 1 ' + '--output {}'.format(output_filename_subtract)\n",
    "            entropy_cmd = cmd + '--normalization_level 2 ' + '--output {}'.format(output_filename_entropy)\n",
    "            rawip_cmd = cmd + '--normalization_level 3 ' + '--output {}'.format(output_filename_ip)\n",
    "            input_cmd = cmd + '--normalization_level 4 ' + '--output {}'.format(output_filename_inp)\n",
    "            \n",
    "            subtract_cmds.append(subtract_cmd)\n",
    "            entropy_cmds.append(entropy_cmd)\n",
    "            rawip_cmds.append(rawip_cmd)\n",
    "            input_cmds.append(input_cmd)\n",
    "            \n",
    "    Submitter(\n",
    "        subtract_cmds, \n",
    "        \"{}_{}_PNGS\".format(event, 'subtract'), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_{}_PNGS.sh'.format(event, 'subtract'),\n",
    "        submit=False,\n",
    "        array=True,\n",
    "        walltime='2:00:00',\n",
    "        queue='condo'\n",
    "    )\n",
    "    Submitter(\n",
    "        entropy_cmds, \n",
    "        \"{}_{}_PNGS\".format(event, 'entropy'), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_{}_PNGS.sh'.format(event, 'entropy'),\n",
    "        submit=False,\n",
    "        array=True,\n",
    "        walltime='2:00:00',\n",
    "        queue='home-scrm'\n",
    "    )\n",
    "    Submitter(\n",
    "        rawip_cmds, \n",
    "        \"{}_{}_PNGS\".format(event, 'ip'), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_{}_PNGS.sh'.format(event, 'rawip'),\n",
    "        submit=False,\n",
    "        array=True,\n",
    "        walltime='2:00:00',\n",
    "        queue='home-scrm'\n",
    "    )\n",
    "    Submitter(\n",
    "        input_cmds, \n",
    "        \"{}_{}_PNGS\".format(event, 'input'), \n",
    "        sh='/projects/ps-yeolab3/bay001/maps/bash_scripts/{}_{}_PNGS.sh'.format(event, 'input'),\n",
    "        submit=True,\n",
    "        array=True,\n",
    "        walltime='2:00:00',\n",
    "        queue='home-scrm'\n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Will need to qsub these individually since there are tons of jobs\n",
    "- all_cdsStart_entropy_PNGS.sh  \n",
    "- all_cdsStart_rawip_PNGS.sh  \n",
    "- all_cdsStart_subtract_PNGS.sh  \n",
    "- all_txStart_entropy_PNGS.sh  \n",
    "- all_txStart_rawip_PNGS.sh  \n",
    "- all_txStart_subtract_PNGS.sh  \n",
    "- all_txStop_entropy_PNGS.sh  \n",
    "- all_txStop_rawip_PNGS.sh  \n",
    "- all_txStop_subtract_PNGS.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verify RBP SE Maps\n",
    "- see if we have all the maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clip_df = pd.read_table(clip_manifest)\n",
    "\n",
    "hepg2_rnaseq_df = pd.read_table(rnaseq_manifests['HepG2'])\n",
    "k562_rnaseq_df = pd.read_table(rnaseq_manifests['K562'])\n",
    "\n",
    "hepg2_clip_df = clip_df[clip_df['Cell line']=='HepG2']\n",
    "k562_clip_df = clip_df[clip_df['Cell line']=='K562']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hepg2_rbps = pd.merge(hepg2_clip_df, hepg2_rnaseq_df, how='outer', left_on='RBP', right_on='Official_RBP').dropna()\n",
    "k562_rbps = pd.merge(k562_clip_df, k562_rnaseq_df, how='outer', left_on='RBP', right_on='Official_RBP').dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos_annotations = glob.glob('/projects/ps-yeolab3/bay001/maps/current_annotations/se/*.positive.nr.txt')\n",
    "neg_annotations = glob.glob('/projects/ps-yeolab3/bay001/maps/current_annotations/se/*.negative.nr.txt')\n",
    "\n",
    "def get_rbp(f):\n",
    "    lst = os.path.basename(f).split('-')\n",
    "    return '{}-{}'.format(lst[0], lst[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos_rbp = set([get_rbp(f) for f in pos_annotations])\n",
    "neg_rbp = set([get_rbp(f) for f in neg_annotations])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allrbp = (pos_rbp.intersection(neg_rbp))\n",
    "print(\"We have annotations for: {} rbps\".format(len(allrbp)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "widgets": {
   "state": {
    "569a4a5b3e314a91ab2b92dcfc9868e4": {
     "views": [
      {
       "cell_index": 29
      }
     ]
    }
   },
   "version": "1.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
