{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic download from encode script.\n",
    "- Current stats as of 2-28:\n",
    "    - 526 Hg19 bams total (254 HepG2, 272 K562).  \n",
    "    - 474 Hg19 bams (230 HepG2, 244 K562)\n",
    "    - 52 control Hg19 bams (24 HepG2, 28 K562)\n",
    "    - 2 missing possible controls (TIA1 K562 and PPIL4 K562)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import urllib\n",
    "import json\n",
    "import requests\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import qtools\n",
    "from tqdm import tnrange, tqdm_notebook\n",
    "\n",
    "pd.set_option(\"display.max_columns\",500)\n",
    "host = 'https://www.encodeproject.org'\n",
    "experiments = \"https://www.encodeproject.org/experiments/\"\n",
    "date = '3-9-2017'\n",
    "annotated_manifest_hepg2 = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_HepG2.csv'\n",
    "annotated_manifest_k562 = '/home/bay001/projects/maps_20160420/permanent_data/RNASeq_final_exp_list_K562.csv'\n",
    "downloaded_dir = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/'\n",
    "aligner = 'ambiguous' # graveley lab pipeline doesn't adhere to the same file name structure as encode processing pipeline, but we trust that it's all tophat-aligned.\n",
    "lab = 'brenton-graveley' # or 'encode-processing-pipeline'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Do some checks on how many samples we have from the two lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('number of k562 rnaseq: ', 224)\n",
      "('number of RBPs in K562', 224)\n",
      "('number of Expts in K562', 224)\n",
      "('number of Expts in Hepg2', 228)\n",
      "('number of hepg2 rnaseq: ', 228)\n",
      "('number of RBPs in Hepg2', 228)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Official_RBP</th>\n",
       "      <th>GENCODE_v19_ID</th>\n",
       "      <th>target</th>\n",
       "      <th>RBP</th>\n",
       "      <th>RNA-Seq on HepG2</th>\n",
       "      <th>EXP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>ENSG00000090861.11</td>\n",
       "      <td>AARS-human</td>\n",
       "      <td>AARS</td>\n",
       "      <td>ENCSR547NWD</td>\n",
       "      <td>AARS-BGHLV17-HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>ENSG00000108270.6</td>\n",
       "      <td>AATF-human</td>\n",
       "      <td>AATF</td>\n",
       "      <td>ENCSR424YSV</td>\n",
       "      <td>AATF-BGHLV14-HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCF1</td>\n",
       "      <td>ENSG00000204574.8</td>\n",
       "      <td>ABCF1-human</td>\n",
       "      <td>ABCF1</td>\n",
       "      <td>ENCSR610VTA</td>\n",
       "      <td>ABCF1-BGHLV30-HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACO1</td>\n",
       "      <td>ENSG00000122729.14</td>\n",
       "      <td>ACO1-human</td>\n",
       "      <td>ACO1</td>\n",
       "      <td>ENCSR511SYK</td>\n",
       "      <td>ACO1-BGHLV31-HepG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADAR</td>\n",
       "      <td>ENSG00000160710.11</td>\n",
       "      <td>ADAR-human</td>\n",
       "      <td>ADAR</td>\n",
       "      <td>ENCSR104OLN</td>\n",
       "      <td>ADAR-BGHLV20-HepG2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Official_RBP      GENCODE_v19_ID       target    RBP RNA-Seq on HepG2  \\\n",
       "0         AARS  ENSG00000090861.11   AARS-human   AARS      ENCSR547NWD   \n",
       "1         AATF   ENSG00000108270.6   AATF-human   AATF      ENCSR424YSV   \n",
       "2        ABCF1   ENSG00000204574.8  ABCF1-human  ABCF1      ENCSR610VTA   \n",
       "3         ACO1  ENSG00000122729.14   ACO1-human   ACO1      ENCSR511SYK   \n",
       "4         ADAR  ENSG00000160710.11   ADAR-human   ADAR      ENCSR104OLN   \n",
       "\n",
       "                   EXP  \n",
       "0   AARS-BGHLV17-HepG2  \n",
       "1   AATF-BGHLV14-HepG2  \n",
       "2  ABCF1-BGHLV30-HepG2  \n",
       "3   ACO1-BGHLV31-HepG2  \n",
       "4   ADAR-BGHLV20-HepG2  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hepg2_df = pd.read_table(annotated_manifest_hepg2)\n",
    "k562_df = pd.read_table(annotated_manifest_k562)\n",
    "\n",
    "print(\"number of k562 rnaseq: \",k562_df.shape[0])\n",
    "print(\"number of RBPs in K562\",len(set(k562_df['Official_RBP'])))\n",
    "print(\"number of Expts in K562\",len(set(k562_df['RNA-Seq on K562'])))\n",
    "print(\"number of Expts in Hepg2\",len(set(hepg2_df['RNA-Seq on HepG2'])))\n",
    "print(\"number of hepg2 rnaseq: \",hepg2_df.shape[0])\n",
    "print(\"number of RBPs in Hepg2\",len(set(hepg2_df['Official_RBP'])))\n",
    "hepg2_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def star_or_tophat(link):\n",
    "    \"\"\"\n",
    "    from the link, reads and returns whether or not STAR or TOPHAT are in the filename.\n",
    "    \"\"\"\n",
    "    # print(\"LINK\",os.path.basename(link))\n",
    "    if os.path.basename(link).find('star') > -1:\n",
    "        return 'STAR'\n",
    "    if os.path.basename(link).find('tophat') > -1:\n",
    "        return 'TOPHAT'\n",
    "    return 'ambiguous'\n",
    "\n",
    "def exists(fpath, current):\n",
    "    \"\"\"\n",
    "    returns whether or not the file (fpath) exists in the directory(current)\n",
    "    \"\"\"\n",
    "    base = os.path.basename(fpath.rstrip())\n",
    "    if not os.path.exists(os.path.join(current,fpath)):\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "def exists_and_return_fullpath(fpath, current):\n",
    "    \"\"\"\n",
    "    returns whether or not the file (fpath) exists in the directory(current)\n",
    "    \"\"\"\n",
    "    base = os.path.basename(fpath.rstrip())\n",
    "    if not os.path.exists(os.path.join(current,fpath)):\n",
    "        return False\n",
    "    else:\n",
    "        return os.path.join(current,fpath)\n",
    "    \n",
    "def get_bams_from_expt_id(\n",
    "    expt_id, assembly, aligner, lab\n",
    "):\n",
    "    \"\"\"\n",
    "    Given an expt id, return a list: [rbp name, [rep1bam, rep2bam], control_expt_id]\n",
    "    \n",
    "    control_expt_id is None if the expt_id given to this function is itself a control.\n",
    "    \n",
    "    \"\"\"\n",
    "    sample_bams = []\n",
    "    control_expts = []\n",
    "    \n",
    "    experiments = \"https://www.encodeproject.org/experiments/\"\n",
    "    url = experiments+expt_id+\"/?format=json\"\n",
    "    response = urllib.urlopen(url)\n",
    "    data = json.loads(response.read())\n",
    "    if 'code' in data.keys():\n",
    "        next\n",
    "    else:\n",
    "        for i in range(0,len(data['files'])):\n",
    "            \n",
    "            if (\n",
    "                (host+data['files'][i]['href']).endswith('bam') & \n",
    "                (data['files'][i]['output_type'] == u'alignments') &\n",
    "                (data['files'][i]['lab'][u'name'] == lab) &\n",
    "                (star_or_tophat(data['files'][i]['submitted_file_name']) == aligner)\n",
    "            ):\n",
    "                if(data['files'][i]['assembly'] == assembly):\n",
    "                    sample_bams.append(\n",
    "                        {\n",
    "                            os.path.basename(data['files'][i]['href']):data['files'][i]['md5sum']\n",
    "                        }\n",
    "                    )\n",
    "        try:\n",
    "            control_expts.append(data['possible_controls'][0]['accession'])\n",
    "        except IndexError:\n",
    "            pass\n",
    "            # print(\"this is a control\")\n",
    "    if(len(control_expts) > 1):\n",
    "        print(\"Warning, this expt {} has more than 1 associated control expt\".format(expt_id))\n",
    "    return data['target']['label'], sample_bams, control_expts\n",
    "\n",
    "def make_expt_dataframe_from_expt_list(expt_list, assembly='hg19', aligner='STAR', lab='encode-processing-pipeline'):\n",
    "    \"\"\"\n",
    "    Given a list of expt ids, return the reps and controls (and md5sums in case we need to check)\n",
    "    \"\"\"\n",
    "    x = tnrange(len(expt_list))\n",
    "    samples = defaultdict(dict)\n",
    "    for h in expt_list:\n",
    "        x.update(1)\n",
    "        sname, s, control_expt = get_bams_from_expt_id(\n",
    "            h, assembly, aligner, lab\n",
    "        ) # gets the sample name, sample, and control expt id\n",
    "        cname, c, _ = get_bams_from_expt_id(\n",
    "            control_expt[0], assembly, aligner, lab\n",
    "        ) # gets the control name, control\n",
    "        for i in range(0,len(s)):\n",
    "            samples_key = 'expt_rep{}'.format(i+1)\n",
    "            samples_md5sum_key = 'expt_rep{}_md5sum'.format(i+1)\n",
    "            control_key = 'control_rep{}'.format(i+1)\n",
    "            controls_md5sum_key = 'control_rep{}_md5sum'.format(i+1)\n",
    "            name_key = 'name'\n",
    "            samples[h][name_key] = sname\n",
    "            samples[h][samples_key] = s[i].keys()[0]\n",
    "            samples[h][control_key] = c[i].keys()[0]\n",
    "            samples[h][samples_md5sum_key] = s[i].values()[0]\n",
    "            samples[h][controls_md5sum_key] = c[i].values()[0]\n",
    "    bams = pd.DataFrame(samples).T\n",
    "    return bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Make a list of expt IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "hepg2_expts = list(set(hepg2_df['RNA-Seq on HepG2']))\n",
    "k562_expts = list(set(k562_df['RNA-Seq on K562']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "(228, 9)\n",
      "(224, 9)\n"
     ]
    }
   ],
   "source": [
    "hepg2_bams = make_expt_dataframe_from_expt_list(hepg2_expts)\n",
    "k562_bams = make_expt_dataframe_from_expt_list(k562_expts)\n",
    "print(hepg2_bams.shape)\n",
    "print(k562_bams.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "hepg2_bams.to_csv(\n",
    "    '/projects/ps-yeolab3/encode/hepg2_{}_{}_bams_for_integrated_analysis.txt'.format(lab, aligner),sep='\\t'\n",
    ")\n",
    "k562_bams.to_csv(\n",
    "    '/projects/ps-yeolab3/encode/k562_{}_{}_bams_for_integrated_analysis.txt'.format(lab, aligner),sep='\\t'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Check md5sum and file existance.\n",
    "- If I've done this once, I don't need to do it again. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def already_checked_md5sum(filepath, md5sum_checked_file):\n",
    "    \"\"\"\n",
    "    given filepaths list, check the md5sum_checked_file to see if any file has been checked.\n",
    "    Parameters\n",
    "    ----------\n",
    "    filepath : list\n",
    "        list of files\n",
    "    md5sum_checked_file : file\n",
    "        dataframe of files\n",
    "        \n",
    "    returns list of files that haven't been checked.\n",
    "    \"\"\"\n",
    "    exists_in_file = ! grep $filepath $md5sum_checked_file\n",
    "    if len(exists_in_file) > 0:\n",
    "        if \"no such file\" not in exists_in_file[0]:\n",
    "            # print(exists_in_file)\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def is_md5sum_equal(row, rep, downloaded_dir):\n",
    "    \"\"\"\n",
    "    Takes a row in a bams dataframe and the rep prefix and determines\n",
    "    whether or not the md5sum is equal to what has been downloaded.\n",
    "    \"\"\"\n",
    "    website_md5sum = row['{}_md5sum'.format(rep)]\n",
    "    filepath = exists_and_return_fullpath(row[rep],downloaded_dir)\n",
    "    calculated_md5sum = ! md5sum $filepath\n",
    "    if website_md5sum != calculated_md5sum[0].split(' ')[0]:\n",
    "        print('TROUBLE: {}, {}, {}'.format(filepath, website_md5sum, calculated_md5sum))\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "def check_all_bams_for_existance(bams, downloaded_dir, to_download_file='download.sh'):\n",
    "    \"\"\"\n",
    "    Writes and prints the files that don't exist to a file\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bams : pandas.DataFrame\n",
    "        dataframe of bams with columns defined in cols.\n",
    "    to_download_file : string\n",
    "        file to download (download.sh)\n",
    "        \n",
    "    \"\"\"\n",
    "    encodeproject_prefix = 'https://www.encodeproject.org/files/'\n",
    "    o = open(to_download_file,'w')\n",
    "    cols = ['expt_rep1','expt_rep2','control_rep1','control_rep2']\n",
    "    for unused,row in bams.iterrows():\n",
    "        for c in cols:\n",
    "            if not(exists(row[c],downloaded_dir)):\n",
    "                # print(row[c], \" doesn't exist.\")\n",
    "                print(\"mv /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/{} ./\".format(row[c]))\n",
    "                o.write(\n",
    "                    encodeproject_prefix + \"{}/@@download/{}\\n\".format(\n",
    "                        os.path.splitext(row[c])[0],\n",
    "                        row[c]\n",
    "                    )\n",
    "                )\n",
    "    o.close()\n",
    "    \n",
    "def check_md5sums(bams, downloaded_dir, md5sums_checked_file):\n",
    "    \"\"\"\n",
    "    Checks the md5sums of the bams file\n",
    "    \"\"\"\n",
    "    cols = ['expt_rep1','expt_rep2','control_rep1','control_rep2']\n",
    "    x = tnrange(bams.shape[0])\n",
    "    \n",
    "    for _, row in bams.iterrows():\n",
    "        x.update(1)\n",
    "        for t in tnrange(len(cols), leave=False):\n",
    "            c = cols[t]\n",
    "            filepath = exists_and_return_fullpath(row[c], downloaded_dir)\n",
    "            if filepath:\n",
    "                if not already_checked_md5sum(\n",
    "                    os.path.join(downloaded_dir,row[c]),\n",
    "                    md5sums_checked_file\n",
    "                ):\n",
    "                    if is_md5sum_equal(row, c, downloaded_dir):\n",
    "                        with open(md5sums_checked_file,'a') as o:\n",
    "                            o.write(os.path.join(downloaded_dir,row[c]) + '\\n')\n",
    "                    else:\n",
    "                        print(row[c], \"needs to be downloaded again.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for existence of file, otherwise create a list to download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# use: xargs -n 1 curl -O -L < FILE to download these...\n",
    "def check_all_bams():\n",
    "    download_hepg2_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/to_download_{}_hepg2_{}_{}'.format(aligner,lab,date)\n",
    "    download_k562_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/to_download_{}_k562_{}_{}'.format(aligner,lab,date)\n",
    "    check_all_bams_for_existance(hepg2_bams, downloaded_dir, download_hepg2_file)\n",
    "    check_all_bams_for_existance(k562_bams, downloaded_dir, download_k562_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "check_all_bams()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check md5sum of the ones we need to download (or just all of them)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# periods indicate row progress, commas indicate rep progress\n",
    "bams = k562_bams\n",
    "md5sums_checked_file = '/projects/ps-yeolab3/encode/rnaseq/shrna_knockdown/md5sums.graveley-tophat.checked'\n",
    "check_md5sums(bams, downloaded_dir, md5sums_checked_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control_rep1</th>\n",
       "      <th>control_rep1_md5sum</th>\n",
       "      <th>control_rep2</th>\n",
       "      <th>control_rep2_md5sum</th>\n",
       "      <th>expt_rep1</th>\n",
       "      <th>expt_rep1_md5sum</th>\n",
       "      <th>expt_rep2</th>\n",
       "      <th>expt_rep2_md5sum</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCSR000KYM</th>\n",
       "      <td>ENCFF309OOI.bam</td>\n",
       "      <td>006834b1e534abfc0a766464ce4f60d9</td>\n",
       "      <td>ENCFF209PLP.bam</td>\n",
       "      <td>b50208a4553de637758e93ceaa099569</td>\n",
       "      <td>ENCFF354JPP.bam</td>\n",
       "      <td>aa41d43a1f51b87c0691d0526f4f3ac5</td>\n",
       "      <td>ENCFF243AMZ.bam</td>\n",
       "      <td>dea667416a4d4c3c582637c2f9ad0f19</td>\n",
       "      <td>DDX3X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR000YYN</th>\n",
       "      <td>ENCFF848JJM.bam</td>\n",
       "      <td>c04240081b7bb137e69d298837baa648</td>\n",
       "      <td>ENCFF695KEA.bam</td>\n",
       "      <td>14e5ef8ccef0d53859e5f67375f4c13f</td>\n",
       "      <td>ENCFF640HPA.bam</td>\n",
       "      <td>7f4e51a76b95e8699789d5bb36c5df85</td>\n",
       "      <td>ENCFF065KHB.bam</td>\n",
       "      <td>84a492145ce5bf8f806aa2b7f9b30765</td>\n",
       "      <td>AKAP8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR004RGI</th>\n",
       "      <td>ENCFF178NNK.bam</td>\n",
       "      <td>7b1be9be800298f9f8958b2adc14090c</td>\n",
       "      <td>ENCFF098BEA.bam</td>\n",
       "      <td>d3be652a5058aacf2e9d95311bcba657</td>\n",
       "      <td>ENCFF922IOC.bam</td>\n",
       "      <td>a2588d2a5a56edeb4f96ea6f4c17d4d9</td>\n",
       "      <td>ENCFF459YMO.bam</td>\n",
       "      <td>22e6c3066888702577d180c635d0280b</td>\n",
       "      <td>RPS10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR007XKL</th>\n",
       "      <td>ENCFF985IXD.bam</td>\n",
       "      <td>2e7f51c61c5c545dfba6906772ac5716</td>\n",
       "      <td>ENCFF461TSD.bam</td>\n",
       "      <td>9e3ce9878e3352b44855f4a2d1c66d37</td>\n",
       "      <td>ENCFF878JKR.bam</td>\n",
       "      <td>4ee27925a7f0f3ce95964686da0a8927</td>\n",
       "      <td>ENCFF466OZP.bam</td>\n",
       "      <td>c69b58950871a86055b9f24744add8c6</td>\n",
       "      <td>NFX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR023HWI</th>\n",
       "      <td>ENCFF804JHE.bam</td>\n",
       "      <td>f14e222eebd2030a3da0d086acd48bf7</td>\n",
       "      <td>ENCFF819ORB.bam</td>\n",
       "      <td>14d4a76c9237f92eff0e40f19ef4ffcb</td>\n",
       "      <td>ENCFF656QEN.bam</td>\n",
       "      <td>f409a646dd82e6b47c88869cd1921a55</td>\n",
       "      <td>ENCFF738SSD.bam</td>\n",
       "      <td>83fe79368dde05a9405107b333b0eca3</td>\n",
       "      <td>KHDRBS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR029LGJ</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF772MUC.bam</td>\n",
       "      <td>b9d28c8eddbe97c0db19681eb2cef1c5</td>\n",
       "      <td>ENCFF402FXS.bam</td>\n",
       "      <td>868de7091e2b4ab872cb07180b6c219f</td>\n",
       "      <td>DDX51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR034VBA</th>\n",
       "      <td>ENCFF669VSB.bam</td>\n",
       "      <td>997fae2a9785ac356107efc36f662d84</td>\n",
       "      <td>ENCFF490QNF.bam</td>\n",
       "      <td>43ae4f234533b09969780ec26dd956ef</td>\n",
       "      <td>ENCFF552SMO.bam</td>\n",
       "      <td>621654e8353221c1022db8aa72f66a85</td>\n",
       "      <td>ENCFF836HER.bam</td>\n",
       "      <td>259850b04aa20c46a845e7e745076957</td>\n",
       "      <td>HNRNPUL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR040FSN</th>\n",
       "      <td>ENCFF341FFK.bam</td>\n",
       "      <td>71bd5d702b7df954831b2b664461f1cd</td>\n",
       "      <td>ENCFF256JTT.bam</td>\n",
       "      <td>0173b3b380fe0cbcfa75a0068cdfbed9</td>\n",
       "      <td>ENCFF617NHT.bam</td>\n",
       "      <td>0e7c5274649aa448cbae6a437d0c6c83</td>\n",
       "      <td>ENCFF240PBV.bam</td>\n",
       "      <td>d45cbba792a437e4c1322bbc80e6b8ff</td>\n",
       "      <td>EIF4G2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR047AJA</th>\n",
       "      <td>ENCFF985IXD.bam</td>\n",
       "      <td>2e7f51c61c5c545dfba6906772ac5716</td>\n",
       "      <td>ENCFF461TSD.bam</td>\n",
       "      <td>9e3ce9878e3352b44855f4a2d1c66d37</td>\n",
       "      <td>ENCFF979CLD.bam</td>\n",
       "      <td>0e0e7f4c3cfbd19da90b98e607ed7f0a</td>\n",
       "      <td>ENCFF252MOZ.bam</td>\n",
       "      <td>08df33b474e55992dbc6fb89ce3daba2</td>\n",
       "      <td>SUB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR047EEG</th>\n",
       "      <td>ENCFF092XAP.bam</td>\n",
       "      <td>b02348cbc3ecb4852ca8266a7f519b89</td>\n",
       "      <td>ENCFF265ZZB.bam</td>\n",
       "      <td>3e64c51b808dcc83bc54eafe7e92d930</td>\n",
       "      <td>ENCFF199ZBI.bam</td>\n",
       "      <td>60bc0e4296a168560c15d4bd668d603b</td>\n",
       "      <td>ENCFF710FGW.bam</td>\n",
       "      <td>09011dd591c7fe9e2ad6466e3d60ac9d</td>\n",
       "      <td>PABPC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR047IUS</th>\n",
       "      <td>ENCFF239MNY.bam</td>\n",
       "      <td>7edd8ad75c39b93c840f5fa40832bfb2</td>\n",
       "      <td>ENCFF961SOT.bam</td>\n",
       "      <td>6338d2d5caec8d25d4bd64badf02723f</td>\n",
       "      <td>ENCFF167VEV.bam</td>\n",
       "      <td>3193914399052241d9bf51ce45ca566c</td>\n",
       "      <td>ENCFF854ZQF.bam</td>\n",
       "      <td>c7a234d56402a5cd82953e3b7285d4ab</td>\n",
       "      <td>HNRNPU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR047QHX</th>\n",
       "      <td>ENCFF847DQL.bam</td>\n",
       "      <td>7374cfece3ec94e0386bcd6453dc3bd0</td>\n",
       "      <td>ENCFF523LXJ.bam</td>\n",
       "      <td>63c3f4678e2a1c1fa30adcc41f3f1426</td>\n",
       "      <td>ENCFF074LFN.bam</td>\n",
       "      <td>f97107249e8a0d75f170b41f852058c9</td>\n",
       "      <td>ENCFF425GWW.bam</td>\n",
       "      <td>517170c8b8efa57ed5a5b9ab9802585b</td>\n",
       "      <td>SF3B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR048BWH</th>\n",
       "      <td>ENCFF239MNY.bam</td>\n",
       "      <td>7edd8ad75c39b93c840f5fa40832bfb2</td>\n",
       "      <td>ENCFF961SOT.bam</td>\n",
       "      <td>6338d2d5caec8d25d4bd64badf02723f</td>\n",
       "      <td>ENCFF395AZN.bam</td>\n",
       "      <td>8366822f54bac839f1292f61c846caf8</td>\n",
       "      <td>ENCFF244AJP.bam</td>\n",
       "      <td>8b4d6f5021dc6fc28ae4da13c97377e6</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR056QEW</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF496ARR.bam</td>\n",
       "      <td>aa87860db9b7173b5d7b2d6603d25cfb</td>\n",
       "      <td>ENCFF815KLH.bam</td>\n",
       "      <td>b6406e485ea7fb9b21bed157d97cd210</td>\n",
       "      <td>CIRBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR060KRD</th>\n",
       "      <td>ENCFF871DVA.bam</td>\n",
       "      <td>a8db8423d7ccd151b1a3072265136e7d</td>\n",
       "      <td>ENCFF746YUL.bam</td>\n",
       "      <td>3d7246d146f3589995917d876a5b7239</td>\n",
       "      <td>ENCFF542NHR.bam</td>\n",
       "      <td>10d092d0df13325ec19466d6226b83f3</td>\n",
       "      <td>ENCFF405TFQ.bam</td>\n",
       "      <td>29bb04acb8619257da66e68ea5cd6fb1</td>\n",
       "      <td>TROVE2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR066VOO</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF446HUH.bam</td>\n",
       "      <td>141ea1f9eae39b8bd461e475f7c84a24</td>\n",
       "      <td>ENCFF669MME.bam</td>\n",
       "      <td>52a68e458d86521b3f15d54ce74a3654</td>\n",
       "      <td>SRSF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR067LLB</th>\n",
       "      <td>ENCFF309OOI.bam</td>\n",
       "      <td>006834b1e534abfc0a766464ce4f60d9</td>\n",
       "      <td>ENCFF209PLP.bam</td>\n",
       "      <td>b50208a4553de637758e93ceaa099569</td>\n",
       "      <td>ENCFF084KMK.bam</td>\n",
       "      <td>1e83113ba4554d79de86c781ab427cfe</td>\n",
       "      <td>ENCFF626FFP.bam</td>\n",
       "      <td>30e1a2886acb7031ff042c2d376e4b2b</td>\n",
       "      <td>DDX24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR076PMZ</th>\n",
       "      <td>ENCFF528NYY.bam</td>\n",
       "      <td>fed572bb466eb4b68c12f3e62407be9a</td>\n",
       "      <td>ENCFF355KZZ.bam</td>\n",
       "      <td>f085cb33a80297e5ff8361e821620d7f</td>\n",
       "      <td>ENCFF751ZEN.bam</td>\n",
       "      <td>73c9b151ff6d7f4b6754e3f1e682df68</td>\n",
       "      <td>ENCFF988QUT.bam</td>\n",
       "      <td>7fb20b908c06eac4b34af7563f2bade9</td>\n",
       "      <td>EIF2S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR077BPR</th>\n",
       "      <td>ENCFF471AKF.bam</td>\n",
       "      <td>358a23234030e66511ca9f6c96143062</td>\n",
       "      <td>ENCFF083EEO.bam</td>\n",
       "      <td>5ce8a166c87db77ef153295a1763a563</td>\n",
       "      <td>ENCFF074MFZ.bam</td>\n",
       "      <td>42d1e23d00209241e55983ad01e92000</td>\n",
       "      <td>ENCFF419DWP.bam</td>\n",
       "      <td>e8fe4c2a2fac5c85dabbc1eb07f585e9</td>\n",
       "      <td>ESF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR079IPT</th>\n",
       "      <td>ENCFF092XAP.bam</td>\n",
       "      <td>b02348cbc3ecb4852ca8266a7f519b89</td>\n",
       "      <td>ENCFF265ZZB.bam</td>\n",
       "      <td>3e64c51b808dcc83bc54eafe7e92d930</td>\n",
       "      <td>ENCFF887PDR.bam</td>\n",
       "      <td>a1d235d24195c71232da83ec81970d8d</td>\n",
       "      <td>ENCFF531JMM.bam</td>\n",
       "      <td>baedd04b8de408ee40d99f28dee2e4ee</td>\n",
       "      <td>TBRG4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR079LMZ</th>\n",
       "      <td>ENCFF804JHE.bam</td>\n",
       "      <td>f14e222eebd2030a3da0d086acd48bf7</td>\n",
       "      <td>ENCFF819ORB.bam</td>\n",
       "      <td>14d4a76c9237f92eff0e40f19ef4ffcb</td>\n",
       "      <td>ENCFF266DIJ.bam</td>\n",
       "      <td>0fb80bc274ac64f358fc2593b17d63e0</td>\n",
       "      <td>ENCFF874NAQ.bam</td>\n",
       "      <td>68d9918b88e22790ee626b2601fdc3ee</td>\n",
       "      <td>DNAJC21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR081XRA</th>\n",
       "      <td>ENCFF092XAP.bam</td>\n",
       "      <td>b02348cbc3ecb4852ca8266a7f519b89</td>\n",
       "      <td>ENCFF265ZZB.bam</td>\n",
       "      <td>3e64c51b808dcc83bc54eafe7e92d930</td>\n",
       "      <td>ENCFF960VQB.bam</td>\n",
       "      <td>515e535314fc420c4f69c81769deb37b</td>\n",
       "      <td>ENCFF375HRX.bam</td>\n",
       "      <td>e50db65422799b49e2eaf26dcb9681a9</td>\n",
       "      <td>SF3B4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR082UWF</th>\n",
       "      <td>ENCFF264YZI.bam</td>\n",
       "      <td>4898dc907fd3e95e8a5f8ad8dc32b90b</td>\n",
       "      <td>ENCFF717TRR.bam</td>\n",
       "      <td>42d79c5fadbd8fbc225cbafa1876afaf</td>\n",
       "      <td>ENCFF543LNI.bam</td>\n",
       "      <td>8695cc9ceb336f99c41f2fd2cb282dc4</td>\n",
       "      <td>ENCFF939VXA.bam</td>\n",
       "      <td>cdbb13dc41e59d50aa5effff3cb07273</td>\n",
       "      <td>STIP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR096WVE</th>\n",
       "      <td>ENCFF822WNV.bam</td>\n",
       "      <td>f131a54b6213943acf373258dda18064</td>\n",
       "      <td>ENCFF611RKG.bam</td>\n",
       "      <td>b5ce85e4b9f546e6109c32884f6ce01b</td>\n",
       "      <td>ENCFF436KRH.bam</td>\n",
       "      <td>699a3a9bbc0aa12f3ac1135bc5cd22d2</td>\n",
       "      <td>ENCFF852UMT.bam</td>\n",
       "      <td>558655c8ece7f28bc4b7728b974b265f</td>\n",
       "      <td>EIF3H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR098NHI</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF296EEA.bam</td>\n",
       "      <td>7ca79d21c10ef1ed854806643f70286e</td>\n",
       "      <td>ENCFF076NJV.bam</td>\n",
       "      <td>14b03ea13b2196ee05769d431e272d53</td>\n",
       "      <td>RPS19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR101OPF</th>\n",
       "      <td>ENCFF341FFK.bam</td>\n",
       "      <td>71bd5d702b7df954831b2b664461f1cd</td>\n",
       "      <td>ENCFF256JTT.bam</td>\n",
       "      <td>0173b3b380fe0cbcfa75a0068cdfbed9</td>\n",
       "      <td>ENCFF041TND.bam</td>\n",
       "      <td>f056824e7ee3506a00044fc87040e18b</td>\n",
       "      <td>ENCFF526SEE.bam</td>\n",
       "      <td>d8e15e813b02ad11cdfd81726012d3d4</td>\n",
       "      <td>SUCLG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR112YTD</th>\n",
       "      <td>ENCFF804JHE.bam</td>\n",
       "      <td>f14e222eebd2030a3da0d086acd48bf7</td>\n",
       "      <td>ENCFF819ORB.bam</td>\n",
       "      <td>14d4a76c9237f92eff0e40f19ef4ffcb</td>\n",
       "      <td>ENCFF215RZE.bam</td>\n",
       "      <td>4915c4c05dc614288922ddabfb3883d1</td>\n",
       "      <td>ENCFF569ZVG.bam</td>\n",
       "      <td>09b71e6bdab6ea5dd3718cd497c5f3df</td>\n",
       "      <td>SLBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR113HRG</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF990JSY.bam</td>\n",
       "      <td>6de62e421c57403edd02b237439b5aa6</td>\n",
       "      <td>ENCFF792MXQ.bam</td>\n",
       "      <td>4e1d3690e8c460ab128df396ccb55715</td>\n",
       "      <td>SRSF9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR113PYX</th>\n",
       "      <td>ENCFF264YZI.bam</td>\n",
       "      <td>4898dc907fd3e95e8a5f8ad8dc32b90b</td>\n",
       "      <td>ENCFF717TRR.bam</td>\n",
       "      <td>42d79c5fadbd8fbc225cbafa1876afaf</td>\n",
       "      <td>ENCFF749NCC.bam</td>\n",
       "      <td>afecc3088c7955351e3420e244d84e26</td>\n",
       "      <td>ENCFF265PUP.bam</td>\n",
       "      <td>ddbb4e06422de2cceaae080a11272175</td>\n",
       "      <td>CNOT7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR117WLY</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF822OHH.bam</td>\n",
       "      <td>d14305259e78f7ae63500078a7b4d7a3</td>\n",
       "      <td>ENCFF811UAO.bam</td>\n",
       "      <td>55fb3e4b289256d647032fa41ddf5750</td>\n",
       "      <td>EFTUD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR874DVZ</th>\n",
       "      <td>ENCFF803MMD.bam</td>\n",
       "      <td>efe4e02c8d90b1b33062a78bbc227e2a</td>\n",
       "      <td>ENCFF351XSJ.bam</td>\n",
       "      <td>47707adf3087ee33e434869eec056c9a</td>\n",
       "      <td>ENCFF902MXK.bam</td>\n",
       "      <td>19074713fbe6d66bc443bdd3da3fdbcb</td>\n",
       "      <td>ENCFF688VZD.bam</td>\n",
       "      <td>555d5b2c939b2491bc374c5c06650152</td>\n",
       "      <td>GLRX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR874ZLI</th>\n",
       "      <td>ENCFF848JJM.bam</td>\n",
       "      <td>c04240081b7bb137e69d298837baa648</td>\n",
       "      <td>ENCFF695KEA.bam</td>\n",
       "      <td>14e5ef8ccef0d53859e5f67375f4c13f</td>\n",
       "      <td>ENCFF218TZW.bam</td>\n",
       "      <td>1edb71b8af92e8efedc49f907293542f</td>\n",
       "      <td>ENCFF385UCY.bam</td>\n",
       "      <td>381dc67be69b9c4e944a342d58c00a4c</td>\n",
       "      <td>CCDC124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR885YOI</th>\n",
       "      <td>ENCFF584WAF.bam</td>\n",
       "      <td>7eab15d31530d19bb4320f2cd5cdfe11</td>\n",
       "      <td>ENCFF674JPQ.bam</td>\n",
       "      <td>d9018956d9512cee07183feac285de7f</td>\n",
       "      <td>ENCFF932CGR.bam</td>\n",
       "      <td>1c6e7c0bc837b62d5ec78a4c3b945461</td>\n",
       "      <td>ENCFF789INJ.bam</td>\n",
       "      <td>aca38270657f2373e15bdfc12dc9559c</td>\n",
       "      <td>CSTF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR891AXF</th>\n",
       "      <td>ENCFF178NNK.bam</td>\n",
       "      <td>7b1be9be800298f9f8958b2adc14090c</td>\n",
       "      <td>ENCFF098BEA.bam</td>\n",
       "      <td>d3be652a5058aacf2e9d95311bcba657</td>\n",
       "      <td>ENCFF749HGR.bam</td>\n",
       "      <td>ae2ebb380a63e1fc9b1d90706ab36c34</td>\n",
       "      <td>ENCFF406LTN.bam</td>\n",
       "      <td>354c8ae25bc271071c8d7175a8babf3c</td>\n",
       "      <td>SSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR891DYO</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF691WZS.bam</td>\n",
       "      <td>3b9d06733ae0a9fc663e6b0437a491ba</td>\n",
       "      <td>ENCFF643FXI.bam</td>\n",
       "      <td>076be026f34af55f66abc8aae2a1bc91</td>\n",
       "      <td>PES1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR898OPN</th>\n",
       "      <td>ENCFF528NYY.bam</td>\n",
       "      <td>fed572bb466eb4b68c12f3e62407be9a</td>\n",
       "      <td>ENCFF355KZZ.bam</td>\n",
       "      <td>f085cb33a80297e5ff8361e821620d7f</td>\n",
       "      <td>ENCFF225ZFL.bam</td>\n",
       "      <td>9fa7df632da45424f15196fa7dcdb5c8</td>\n",
       "      <td>ENCFF666NAH.bam</td>\n",
       "      <td>c89265c4d3d8240e68d3a0acc87750cb</td>\n",
       "      <td>RBM17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR902WSK</th>\n",
       "      <td>ENCFF471AKF.bam</td>\n",
       "      <td>358a23234030e66511ca9f6c96143062</td>\n",
       "      <td>ENCFF083EEO.bam</td>\n",
       "      <td>5ce8a166c87db77ef153295a1763a563</td>\n",
       "      <td>ENCFF266SNR.bam</td>\n",
       "      <td>bcf100fdab2ab34febe5aff01d77a741</td>\n",
       "      <td>ENCFF630HEN.bam</td>\n",
       "      <td>c6dc5d4fb303736e1c8dc7b59eabc8ac</td>\n",
       "      <td>SSRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR904BCZ</th>\n",
       "      <td>ENCFF092XAP.bam</td>\n",
       "      <td>b02348cbc3ecb4852ca8266a7f519b89</td>\n",
       "      <td>ENCFF265ZZB.bam</td>\n",
       "      <td>3e64c51b808dcc83bc54eafe7e92d930</td>\n",
       "      <td>ENCFF793LEQ.bam</td>\n",
       "      <td>5df5df615fd203cf31eb024ebb328014</td>\n",
       "      <td>ENCFF939XGF.bam</td>\n",
       "      <td>6d83c671780f8e41fd5f9c3f8d74bca2</td>\n",
       "      <td>RAVER1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR904CJQ</th>\n",
       "      <td>ENCFF092XAP.bam</td>\n",
       "      <td>b02348cbc3ecb4852ca8266a7f519b89</td>\n",
       "      <td>ENCFF265ZZB.bam</td>\n",
       "      <td>3e64c51b808dcc83bc54eafe7e92d930</td>\n",
       "      <td>ENCFF667QFW.bam</td>\n",
       "      <td>7fa34d6e851dabeff2affc5485163b87</td>\n",
       "      <td>ENCFF318PFL.bam</td>\n",
       "      <td>5bcb90162ad3ef2ccc773fd8ec3a49d2</td>\n",
       "      <td>U2AF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR906RHU</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF287ZBH.bam</td>\n",
       "      <td>570bf22a79119a052938db1b04810603</td>\n",
       "      <td>ENCFF142DJX.bam</td>\n",
       "      <td>adc8cbd7a2f18162909cc5505b79be1d</td>\n",
       "      <td>SRSF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR907UTB</th>\n",
       "      <td>ENCFF985IXD.bam</td>\n",
       "      <td>2e7f51c61c5c545dfba6906772ac5716</td>\n",
       "      <td>ENCFF461TSD.bam</td>\n",
       "      <td>9e3ce9878e3352b44855f4a2d1c66d37</td>\n",
       "      <td>ENCFF097OIM.bam</td>\n",
       "      <td>2ad27ccd7c20ccd0c85b97c0f8cc5d60</td>\n",
       "      <td>ENCFF304ZSG.bam</td>\n",
       "      <td>6c3fb51f0e4f57c547a03a23bef662e5</td>\n",
       "      <td>DAZAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR911DGK</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF251OKX.bam</td>\n",
       "      <td>9dd41a09d713073329ba5896bb7467ff</td>\n",
       "      <td>ENCFF407PAK.bam</td>\n",
       "      <td>c4959ba18e96d36550067dcbe34f78e3</td>\n",
       "      <td>TFIP11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR916WOI</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF497GZB.bam</td>\n",
       "      <td>c0ae8397fa0c305f596cae62d96873e0</td>\n",
       "      <td>ENCFF529ZHW.bam</td>\n",
       "      <td>d063a40258df81222e50ecb4541edd09</td>\n",
       "      <td>TRA2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR921KDS</th>\n",
       "      <td>ENCFF719PLT.bam</td>\n",
       "      <td>ffbb2204bcab4774832401893e91061d</td>\n",
       "      <td>ENCFF275ZEJ.bam</td>\n",
       "      <td>33370761b6b46fc03f19b6f3e13511c0</td>\n",
       "      <td>ENCFF451UBP.bam</td>\n",
       "      <td>2e5f20b57833f026801b3e5d5e792e2d</td>\n",
       "      <td>ENCFF441MQQ.bam</td>\n",
       "      <td>0b0e42513f0ba5112c1985b48758df64</td>\n",
       "      <td>RCC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR925RNE</th>\n",
       "      <td>ENCFF362TIW.bam</td>\n",
       "      <td>2135657d1727c78ac2aba8176d6323f2</td>\n",
       "      <td>ENCFF742XEQ.bam</td>\n",
       "      <td>5739adf7b6b6165e9a2bb9a17c3a5310</td>\n",
       "      <td>ENCFF821YFS.bam</td>\n",
       "      <td>a0e3394d086f31e645d3b24473bc01f7</td>\n",
       "      <td>ENCFF246KQB.bam</td>\n",
       "      <td>ec457fe555f6ec387e7a8c0fe9a1f3cc</td>\n",
       "      <td>SERBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR927TSP</th>\n",
       "      <td>ENCFF719PLT.bam</td>\n",
       "      <td>ffbb2204bcab4774832401893e91061d</td>\n",
       "      <td>ENCFF275ZEJ.bam</td>\n",
       "      <td>33370761b6b46fc03f19b6f3e13511c0</td>\n",
       "      <td>ENCFF057FXC.bam</td>\n",
       "      <td>506cba8c995933270d0309ba48bacf6b</td>\n",
       "      <td>ENCFF953IZH.bam</td>\n",
       "      <td>da69132c7ea9b318b233bf7be0761f00</td>\n",
       "      <td>TIAL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR927XBT</th>\n",
       "      <td>ENCFF719PLT.bam</td>\n",
       "      <td>ffbb2204bcab4774832401893e91061d</td>\n",
       "      <td>ENCFF275ZEJ.bam</td>\n",
       "      <td>33370761b6b46fc03f19b6f3e13511c0</td>\n",
       "      <td>ENCFF693ZJX.bam</td>\n",
       "      <td>551a0392cade2a0739fbbfb6579bae60</td>\n",
       "      <td>ENCFF204APB.bam</td>\n",
       "      <td>a6b0a23a6f0d1e2123f63603d560f751</td>\n",
       "      <td>NUSAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR943LIB</th>\n",
       "      <td>ENCFF669VSB.bam</td>\n",
       "      <td>997fae2a9785ac356107efc36f662d84</td>\n",
       "      <td>ENCFF490QNF.bam</td>\n",
       "      <td>43ae4f234533b09969780ec26dd956ef</td>\n",
       "      <td>ENCFF167AFA.bam</td>\n",
       "      <td>0ebb6f86c63122beef1716a993066a30</td>\n",
       "      <td>ENCFF193SZZ.bam</td>\n",
       "      <td>4ff32e805f86a74f5568acfd9e4e1513</td>\n",
       "      <td>SNRNP200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR945GUR</th>\n",
       "      <td>ENCFF848JJM.bam</td>\n",
       "      <td>c04240081b7bb137e69d298837baa648</td>\n",
       "      <td>ENCFF695KEA.bam</td>\n",
       "      <td>14e5ef8ccef0d53859e5f67375f4c13f</td>\n",
       "      <td>ENCFF337TOV.bam</td>\n",
       "      <td>e9f206bf795127cedcc20900063f0e92</td>\n",
       "      <td>ENCFF414BQP.bam</td>\n",
       "      <td>8bf7422814bfd2a5b02d2f93ea920fa3</td>\n",
       "      <td>NAA15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR947OIM</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF016WOB.bam</td>\n",
       "      <td>dcb2ead998c0a39ea2746e31ce2ea81f</td>\n",
       "      <td>ENCFF535LRB.bam</td>\n",
       "      <td>c239ace2a2ffd38702a3b4dce7bcb008</td>\n",
       "      <td>RBM22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR952QDQ</th>\n",
       "      <td>ENCFF847DQL.bam</td>\n",
       "      <td>7374cfece3ec94e0386bcd6453dc3bd0</td>\n",
       "      <td>ENCFF523LXJ.bam</td>\n",
       "      <td>63c3f4678e2a1c1fa30adcc41f3f1426</td>\n",
       "      <td>ENCFF605WVQ.bam</td>\n",
       "      <td>626acee742c5fd5385d7918cfa4dfcc3</td>\n",
       "      <td>ENCFF371TVY.bam</td>\n",
       "      <td>1480adf72f3817db30575c1620320c61</td>\n",
       "      <td>METAP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR952RRH</th>\n",
       "      <td>ENCFF970PYT.bam</td>\n",
       "      <td>0a31eb661ad3bd5e16c03e5bc7caae2a</td>\n",
       "      <td>ENCFF019XTR.bam</td>\n",
       "      <td>2cd62e27919be8a273e484bd9fdb5692</td>\n",
       "      <td>ENCFF342OBS.bam</td>\n",
       "      <td>5b03bc17a8128e52cc838f4ab16ee1d7</td>\n",
       "      <td>ENCFF519XDG.bam</td>\n",
       "      <td>685a67ae00026fc05d7c666f4f064fa0</td>\n",
       "      <td>IGF2BP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR954HAY</th>\n",
       "      <td>ENCFF264YZI.bam</td>\n",
       "      <td>4898dc907fd3e95e8a5f8ad8dc32b90b</td>\n",
       "      <td>ENCFF717TRR.bam</td>\n",
       "      <td>42d79c5fadbd8fbc225cbafa1876afaf</td>\n",
       "      <td>ENCFF302IKH.bam</td>\n",
       "      <td>506e2e2e6fc378bcfe5fe383ddfcd1ef</td>\n",
       "      <td>ENCFF488MVU.bam</td>\n",
       "      <td>d75cfc2683f51a0fe04253a7781d69a3</td>\n",
       "      <td>SART3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR958NDU</th>\n",
       "      <td>ENCFF985IXD.bam</td>\n",
       "      <td>2e7f51c61c5c545dfba6906772ac5716</td>\n",
       "      <td>ENCFF461TSD.bam</td>\n",
       "      <td>9e3ce9878e3352b44855f4a2d1c66d37</td>\n",
       "      <td>ENCFF062BJD.bam</td>\n",
       "      <td>d1a4d652cf79fc5da7c6a9d552e37d46</td>\n",
       "      <td>ENCFF836SGI.bam</td>\n",
       "      <td>fb47f84207a8625d73c8429b8cf04902</td>\n",
       "      <td>HLTF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR961YAG</th>\n",
       "      <td>ENCFF309OOI.bam</td>\n",
       "      <td>006834b1e534abfc0a766464ce4f60d9</td>\n",
       "      <td>ENCFF209PLP.bam</td>\n",
       "      <td>b50208a4553de637758e93ceaa099569</td>\n",
       "      <td>ENCFF067HJU.bam</td>\n",
       "      <td>01ea66f0af664f22acfc2829aa67701c</td>\n",
       "      <td>ENCFF211TCF.bam</td>\n",
       "      <td>e86966f0ca77facf5a3e23a1e7e77cb1</td>\n",
       "      <td>EIF4A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR963RLK</th>\n",
       "      <td>ENCFF341FFK.bam</td>\n",
       "      <td>71bd5d702b7df954831b2b664461f1cd</td>\n",
       "      <td>ENCFF256JTT.bam</td>\n",
       "      <td>0173b3b380fe0cbcfa75a0068cdfbed9</td>\n",
       "      <td>ENCFF125TRY.bam</td>\n",
       "      <td>178541d6beb54ffa9520bff37edfe20e</td>\n",
       "      <td>ENCFF236LSH.bam</td>\n",
       "      <td>5215d582d345555b249236d2b2dacb26</td>\n",
       "      <td>APOBEC3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR967QNT</th>\n",
       "      <td>ENCFF847DQL.bam</td>\n",
       "      <td>7374cfece3ec94e0386bcd6453dc3bd0</td>\n",
       "      <td>ENCFF523LXJ.bam</td>\n",
       "      <td>63c3f4678e2a1c1fa30adcc41f3f1426</td>\n",
       "      <td>ENCFF122EWD.bam</td>\n",
       "      <td>a0d8ac21a28fbfae3993b78e4d7c3d0d</td>\n",
       "      <td>ENCFF152ZSQ.bam</td>\n",
       "      <td>a7174bf579eb5598c49c08c65e95508b</td>\n",
       "      <td>GPKOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR973QSV</th>\n",
       "      <td>ENCFF985IXD.bam</td>\n",
       "      <td>2e7f51c61c5c545dfba6906772ac5716</td>\n",
       "      <td>ENCFF461TSD.bam</td>\n",
       "      <td>9e3ce9878e3352b44855f4a2d1c66d37</td>\n",
       "      <td>ENCFF509JMI.bam</td>\n",
       "      <td>3a51c0f50a8a50b2c15f99d4ed0a78a4</td>\n",
       "      <td>ENCFF437YII.bam</td>\n",
       "      <td>1f14fc46203bf78bb751a111db534b5b</td>\n",
       "      <td>AATF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR978CSQ</th>\n",
       "      <td>ENCFF239MNY.bam</td>\n",
       "      <td>7edd8ad75c39b93c840f5fa40832bfb2</td>\n",
       "      <td>ENCFF961SOT.bam</td>\n",
       "      <td>6338d2d5caec8d25d4bd64badf02723f</td>\n",
       "      <td>ENCFF403CZS.bam</td>\n",
       "      <td>b4081ded594504582272fe1795d35898</td>\n",
       "      <td>ENCFF085ZVA.bam</td>\n",
       "      <td>3bb1915140afe5dd27a729492f6cb158</td>\n",
       "      <td>PKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCSR984CLJ</th>\n",
       "      <td>ENCFF341FFK.bam</td>\n",
       "      <td>71bd5d702b7df954831b2b664461f1cd</td>\n",
       "      <td>ENCFF256JTT.bam</td>\n",
       "      <td>0173b3b380fe0cbcfa75a0068cdfbed9</td>\n",
       "      <td>ENCFF079PJT.bam</td>\n",
       "      <td>17f954cb3b25aafc2441baab7b65f5c9</td>\n",
       "      <td>ENCFF144BCM.bam</td>\n",
       "      <td>87325cbdf7c098259f835f4826acc8cc</td>\n",
       "      <td>CCAR2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>224 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                control_rep1               control_rep1_md5sum  \\\n",
       "ENCSR000KYM  ENCFF309OOI.bam  006834b1e534abfc0a766464ce4f60d9   \n",
       "ENCSR000YYN  ENCFF848JJM.bam  c04240081b7bb137e69d298837baa648   \n",
       "ENCSR004RGI  ENCFF178NNK.bam  7b1be9be800298f9f8958b2adc14090c   \n",
       "ENCSR007XKL  ENCFF985IXD.bam  2e7f51c61c5c545dfba6906772ac5716   \n",
       "ENCSR023HWI  ENCFF804JHE.bam  f14e222eebd2030a3da0d086acd48bf7   \n",
       "ENCSR029LGJ  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "ENCSR034VBA  ENCFF669VSB.bam  997fae2a9785ac356107efc36f662d84   \n",
       "ENCSR040FSN  ENCFF341FFK.bam  71bd5d702b7df954831b2b664461f1cd   \n",
       "ENCSR047AJA  ENCFF985IXD.bam  2e7f51c61c5c545dfba6906772ac5716   \n",
       "ENCSR047EEG  ENCFF092XAP.bam  b02348cbc3ecb4852ca8266a7f519b89   \n",
       "ENCSR047IUS  ENCFF239MNY.bam  7edd8ad75c39b93c840f5fa40832bfb2   \n",
       "ENCSR047QHX  ENCFF847DQL.bam  7374cfece3ec94e0386bcd6453dc3bd0   \n",
       "ENCSR048BWH  ENCFF239MNY.bam  7edd8ad75c39b93c840f5fa40832bfb2   \n",
       "ENCSR056QEW  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR060KRD  ENCFF871DVA.bam  a8db8423d7ccd151b1a3072265136e7d   \n",
       "ENCSR066VOO  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR067LLB  ENCFF309OOI.bam  006834b1e534abfc0a766464ce4f60d9   \n",
       "ENCSR076PMZ  ENCFF528NYY.bam  fed572bb466eb4b68c12f3e62407be9a   \n",
       "ENCSR077BPR  ENCFF471AKF.bam  358a23234030e66511ca9f6c96143062   \n",
       "ENCSR079IPT  ENCFF092XAP.bam  b02348cbc3ecb4852ca8266a7f519b89   \n",
       "ENCSR079LMZ  ENCFF804JHE.bam  f14e222eebd2030a3da0d086acd48bf7   \n",
       "ENCSR081XRA  ENCFF092XAP.bam  b02348cbc3ecb4852ca8266a7f519b89   \n",
       "ENCSR082UWF  ENCFF264YZI.bam  4898dc907fd3e95e8a5f8ad8dc32b90b   \n",
       "ENCSR096WVE  ENCFF822WNV.bam  f131a54b6213943acf373258dda18064   \n",
       "ENCSR098NHI  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "ENCSR101OPF  ENCFF341FFK.bam  71bd5d702b7df954831b2b664461f1cd   \n",
       "ENCSR112YTD  ENCFF804JHE.bam  f14e222eebd2030a3da0d086acd48bf7   \n",
       "ENCSR113HRG  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR113PYX  ENCFF264YZI.bam  4898dc907fd3e95e8a5f8ad8dc32b90b   \n",
       "ENCSR117WLY  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "...                      ...                               ...   \n",
       "ENCSR874DVZ  ENCFF803MMD.bam  efe4e02c8d90b1b33062a78bbc227e2a   \n",
       "ENCSR874ZLI  ENCFF848JJM.bam  c04240081b7bb137e69d298837baa648   \n",
       "ENCSR885YOI  ENCFF584WAF.bam  7eab15d31530d19bb4320f2cd5cdfe11   \n",
       "ENCSR891AXF  ENCFF178NNK.bam  7b1be9be800298f9f8958b2adc14090c   \n",
       "ENCSR891DYO  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "ENCSR898OPN  ENCFF528NYY.bam  fed572bb466eb4b68c12f3e62407be9a   \n",
       "ENCSR902WSK  ENCFF471AKF.bam  358a23234030e66511ca9f6c96143062   \n",
       "ENCSR904BCZ  ENCFF092XAP.bam  b02348cbc3ecb4852ca8266a7f519b89   \n",
       "ENCSR904CJQ  ENCFF092XAP.bam  b02348cbc3ecb4852ca8266a7f519b89   \n",
       "ENCSR906RHU  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR907UTB  ENCFF985IXD.bam  2e7f51c61c5c545dfba6906772ac5716   \n",
       "ENCSR911DGK  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "ENCSR916WOI  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR921KDS  ENCFF719PLT.bam  ffbb2204bcab4774832401893e91061d   \n",
       "ENCSR925RNE  ENCFF362TIW.bam  2135657d1727c78ac2aba8176d6323f2   \n",
       "ENCSR927TSP  ENCFF719PLT.bam  ffbb2204bcab4774832401893e91061d   \n",
       "ENCSR927XBT  ENCFF719PLT.bam  ffbb2204bcab4774832401893e91061d   \n",
       "ENCSR943LIB  ENCFF669VSB.bam  997fae2a9785ac356107efc36f662d84   \n",
       "ENCSR945GUR  ENCFF848JJM.bam  c04240081b7bb137e69d298837baa648   \n",
       "ENCSR947OIM  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR952QDQ  ENCFF847DQL.bam  7374cfece3ec94e0386bcd6453dc3bd0   \n",
       "ENCSR952RRH  ENCFF970PYT.bam  0a31eb661ad3bd5e16c03e5bc7caae2a   \n",
       "ENCSR954HAY  ENCFF264YZI.bam  4898dc907fd3e95e8a5f8ad8dc32b90b   \n",
       "ENCSR958NDU  ENCFF985IXD.bam  2e7f51c61c5c545dfba6906772ac5716   \n",
       "ENCSR961YAG  ENCFF309OOI.bam  006834b1e534abfc0a766464ce4f60d9   \n",
       "ENCSR963RLK  ENCFF341FFK.bam  71bd5d702b7df954831b2b664461f1cd   \n",
       "ENCSR967QNT  ENCFF847DQL.bam  7374cfece3ec94e0386bcd6453dc3bd0   \n",
       "ENCSR973QSV  ENCFF985IXD.bam  2e7f51c61c5c545dfba6906772ac5716   \n",
       "ENCSR978CSQ  ENCFF239MNY.bam  7edd8ad75c39b93c840f5fa40832bfb2   \n",
       "ENCSR984CLJ  ENCFF341FFK.bam  71bd5d702b7df954831b2b664461f1cd   \n",
       "\n",
       "                control_rep2               control_rep2_md5sum  \\\n",
       "ENCSR000KYM  ENCFF209PLP.bam  b50208a4553de637758e93ceaa099569   \n",
       "ENCSR000YYN  ENCFF695KEA.bam  14e5ef8ccef0d53859e5f67375f4c13f   \n",
       "ENCSR004RGI  ENCFF098BEA.bam  d3be652a5058aacf2e9d95311bcba657   \n",
       "ENCSR007XKL  ENCFF461TSD.bam  9e3ce9878e3352b44855f4a2d1c66d37   \n",
       "ENCSR023HWI  ENCFF819ORB.bam  14d4a76c9237f92eff0e40f19ef4ffcb   \n",
       "ENCSR029LGJ  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "ENCSR034VBA  ENCFF490QNF.bam  43ae4f234533b09969780ec26dd956ef   \n",
       "ENCSR040FSN  ENCFF256JTT.bam  0173b3b380fe0cbcfa75a0068cdfbed9   \n",
       "ENCSR047AJA  ENCFF461TSD.bam  9e3ce9878e3352b44855f4a2d1c66d37   \n",
       "ENCSR047EEG  ENCFF265ZZB.bam  3e64c51b808dcc83bc54eafe7e92d930   \n",
       "ENCSR047IUS  ENCFF961SOT.bam  6338d2d5caec8d25d4bd64badf02723f   \n",
       "ENCSR047QHX  ENCFF523LXJ.bam  63c3f4678e2a1c1fa30adcc41f3f1426   \n",
       "ENCSR048BWH  ENCFF961SOT.bam  6338d2d5caec8d25d4bd64badf02723f   \n",
       "ENCSR056QEW  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR060KRD  ENCFF746YUL.bam  3d7246d146f3589995917d876a5b7239   \n",
       "ENCSR066VOO  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR067LLB  ENCFF209PLP.bam  b50208a4553de637758e93ceaa099569   \n",
       "ENCSR076PMZ  ENCFF355KZZ.bam  f085cb33a80297e5ff8361e821620d7f   \n",
       "ENCSR077BPR  ENCFF083EEO.bam  5ce8a166c87db77ef153295a1763a563   \n",
       "ENCSR079IPT  ENCFF265ZZB.bam  3e64c51b808dcc83bc54eafe7e92d930   \n",
       "ENCSR079LMZ  ENCFF819ORB.bam  14d4a76c9237f92eff0e40f19ef4ffcb   \n",
       "ENCSR081XRA  ENCFF265ZZB.bam  3e64c51b808dcc83bc54eafe7e92d930   \n",
       "ENCSR082UWF  ENCFF717TRR.bam  42d79c5fadbd8fbc225cbafa1876afaf   \n",
       "ENCSR096WVE  ENCFF611RKG.bam  b5ce85e4b9f546e6109c32884f6ce01b   \n",
       "ENCSR098NHI  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "ENCSR101OPF  ENCFF256JTT.bam  0173b3b380fe0cbcfa75a0068cdfbed9   \n",
       "ENCSR112YTD  ENCFF819ORB.bam  14d4a76c9237f92eff0e40f19ef4ffcb   \n",
       "ENCSR113HRG  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR113PYX  ENCFF717TRR.bam  42d79c5fadbd8fbc225cbafa1876afaf   \n",
       "ENCSR117WLY  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "...                      ...                               ...   \n",
       "ENCSR874DVZ  ENCFF351XSJ.bam  47707adf3087ee33e434869eec056c9a   \n",
       "ENCSR874ZLI  ENCFF695KEA.bam  14e5ef8ccef0d53859e5f67375f4c13f   \n",
       "ENCSR885YOI  ENCFF674JPQ.bam  d9018956d9512cee07183feac285de7f   \n",
       "ENCSR891AXF  ENCFF098BEA.bam  d3be652a5058aacf2e9d95311bcba657   \n",
       "ENCSR891DYO  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "ENCSR898OPN  ENCFF355KZZ.bam  f085cb33a80297e5ff8361e821620d7f   \n",
       "ENCSR902WSK  ENCFF083EEO.bam  5ce8a166c87db77ef153295a1763a563   \n",
       "ENCSR904BCZ  ENCFF265ZZB.bam  3e64c51b808dcc83bc54eafe7e92d930   \n",
       "ENCSR904CJQ  ENCFF265ZZB.bam  3e64c51b808dcc83bc54eafe7e92d930   \n",
       "ENCSR906RHU  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR907UTB  ENCFF461TSD.bam  9e3ce9878e3352b44855f4a2d1c66d37   \n",
       "ENCSR911DGK  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "ENCSR916WOI  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR921KDS  ENCFF275ZEJ.bam  33370761b6b46fc03f19b6f3e13511c0   \n",
       "ENCSR925RNE  ENCFF742XEQ.bam  5739adf7b6b6165e9a2bb9a17c3a5310   \n",
       "ENCSR927TSP  ENCFF275ZEJ.bam  33370761b6b46fc03f19b6f3e13511c0   \n",
       "ENCSR927XBT  ENCFF275ZEJ.bam  33370761b6b46fc03f19b6f3e13511c0   \n",
       "ENCSR943LIB  ENCFF490QNF.bam  43ae4f234533b09969780ec26dd956ef   \n",
       "ENCSR945GUR  ENCFF695KEA.bam  14e5ef8ccef0d53859e5f67375f4c13f   \n",
       "ENCSR947OIM  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR952QDQ  ENCFF523LXJ.bam  63c3f4678e2a1c1fa30adcc41f3f1426   \n",
       "ENCSR952RRH  ENCFF019XTR.bam  2cd62e27919be8a273e484bd9fdb5692   \n",
       "ENCSR954HAY  ENCFF717TRR.bam  42d79c5fadbd8fbc225cbafa1876afaf   \n",
       "ENCSR958NDU  ENCFF461TSD.bam  9e3ce9878e3352b44855f4a2d1c66d37   \n",
       "ENCSR961YAG  ENCFF209PLP.bam  b50208a4553de637758e93ceaa099569   \n",
       "ENCSR963RLK  ENCFF256JTT.bam  0173b3b380fe0cbcfa75a0068cdfbed9   \n",
       "ENCSR967QNT  ENCFF523LXJ.bam  63c3f4678e2a1c1fa30adcc41f3f1426   \n",
       "ENCSR973QSV  ENCFF461TSD.bam  9e3ce9878e3352b44855f4a2d1c66d37   \n",
       "ENCSR978CSQ  ENCFF961SOT.bam  6338d2d5caec8d25d4bd64badf02723f   \n",
       "ENCSR984CLJ  ENCFF256JTT.bam  0173b3b380fe0cbcfa75a0068cdfbed9   \n",
       "\n",
       "                   expt_rep1                  expt_rep1_md5sum  \\\n",
       "ENCSR000KYM  ENCFF354JPP.bam  aa41d43a1f51b87c0691d0526f4f3ac5   \n",
       "ENCSR000YYN  ENCFF640HPA.bam  7f4e51a76b95e8699789d5bb36c5df85   \n",
       "ENCSR004RGI  ENCFF922IOC.bam  a2588d2a5a56edeb4f96ea6f4c17d4d9   \n",
       "ENCSR007XKL  ENCFF878JKR.bam  4ee27925a7f0f3ce95964686da0a8927   \n",
       "ENCSR023HWI  ENCFF656QEN.bam  f409a646dd82e6b47c88869cd1921a55   \n",
       "ENCSR029LGJ  ENCFF772MUC.bam  b9d28c8eddbe97c0db19681eb2cef1c5   \n",
       "ENCSR034VBA  ENCFF552SMO.bam  621654e8353221c1022db8aa72f66a85   \n",
       "ENCSR040FSN  ENCFF617NHT.bam  0e7c5274649aa448cbae6a437d0c6c83   \n",
       "ENCSR047AJA  ENCFF979CLD.bam  0e0e7f4c3cfbd19da90b98e607ed7f0a   \n",
       "ENCSR047EEG  ENCFF199ZBI.bam  60bc0e4296a168560c15d4bd668d603b   \n",
       "ENCSR047IUS  ENCFF167VEV.bam  3193914399052241d9bf51ce45ca566c   \n",
       "ENCSR047QHX  ENCFF074LFN.bam  f97107249e8a0d75f170b41f852058c9   \n",
       "ENCSR048BWH  ENCFF395AZN.bam  8366822f54bac839f1292f61c846caf8   \n",
       "ENCSR056QEW  ENCFF496ARR.bam  aa87860db9b7173b5d7b2d6603d25cfb   \n",
       "ENCSR060KRD  ENCFF542NHR.bam  10d092d0df13325ec19466d6226b83f3   \n",
       "ENCSR066VOO  ENCFF446HUH.bam  141ea1f9eae39b8bd461e475f7c84a24   \n",
       "ENCSR067LLB  ENCFF084KMK.bam  1e83113ba4554d79de86c781ab427cfe   \n",
       "ENCSR076PMZ  ENCFF751ZEN.bam  73c9b151ff6d7f4b6754e3f1e682df68   \n",
       "ENCSR077BPR  ENCFF074MFZ.bam  42d1e23d00209241e55983ad01e92000   \n",
       "ENCSR079IPT  ENCFF887PDR.bam  a1d235d24195c71232da83ec81970d8d   \n",
       "ENCSR079LMZ  ENCFF266DIJ.bam  0fb80bc274ac64f358fc2593b17d63e0   \n",
       "ENCSR081XRA  ENCFF960VQB.bam  515e535314fc420c4f69c81769deb37b   \n",
       "ENCSR082UWF  ENCFF543LNI.bam  8695cc9ceb336f99c41f2fd2cb282dc4   \n",
       "ENCSR096WVE  ENCFF436KRH.bam  699a3a9bbc0aa12f3ac1135bc5cd22d2   \n",
       "ENCSR098NHI  ENCFF296EEA.bam  7ca79d21c10ef1ed854806643f70286e   \n",
       "ENCSR101OPF  ENCFF041TND.bam  f056824e7ee3506a00044fc87040e18b   \n",
       "ENCSR112YTD  ENCFF215RZE.bam  4915c4c05dc614288922ddabfb3883d1   \n",
       "ENCSR113HRG  ENCFF990JSY.bam  6de62e421c57403edd02b237439b5aa6   \n",
       "ENCSR113PYX  ENCFF749NCC.bam  afecc3088c7955351e3420e244d84e26   \n",
       "ENCSR117WLY  ENCFF822OHH.bam  d14305259e78f7ae63500078a7b4d7a3   \n",
       "...                      ...                               ...   \n",
       "ENCSR874DVZ  ENCFF902MXK.bam  19074713fbe6d66bc443bdd3da3fdbcb   \n",
       "ENCSR874ZLI  ENCFF218TZW.bam  1edb71b8af92e8efedc49f907293542f   \n",
       "ENCSR885YOI  ENCFF932CGR.bam  1c6e7c0bc837b62d5ec78a4c3b945461   \n",
       "ENCSR891AXF  ENCFF749HGR.bam  ae2ebb380a63e1fc9b1d90706ab36c34   \n",
       "ENCSR891DYO  ENCFF691WZS.bam  3b9d06733ae0a9fc663e6b0437a491ba   \n",
       "ENCSR898OPN  ENCFF225ZFL.bam  9fa7df632da45424f15196fa7dcdb5c8   \n",
       "ENCSR902WSK  ENCFF266SNR.bam  bcf100fdab2ab34febe5aff01d77a741   \n",
       "ENCSR904BCZ  ENCFF793LEQ.bam  5df5df615fd203cf31eb024ebb328014   \n",
       "ENCSR904CJQ  ENCFF667QFW.bam  7fa34d6e851dabeff2affc5485163b87   \n",
       "ENCSR906RHU  ENCFF287ZBH.bam  570bf22a79119a052938db1b04810603   \n",
       "ENCSR907UTB  ENCFF097OIM.bam  2ad27ccd7c20ccd0c85b97c0f8cc5d60   \n",
       "ENCSR911DGK  ENCFF251OKX.bam  9dd41a09d713073329ba5896bb7467ff   \n",
       "ENCSR916WOI  ENCFF497GZB.bam  c0ae8397fa0c305f596cae62d96873e0   \n",
       "ENCSR921KDS  ENCFF451UBP.bam  2e5f20b57833f026801b3e5d5e792e2d   \n",
       "ENCSR925RNE  ENCFF821YFS.bam  a0e3394d086f31e645d3b24473bc01f7   \n",
       "ENCSR927TSP  ENCFF057FXC.bam  506cba8c995933270d0309ba48bacf6b   \n",
       "ENCSR927XBT  ENCFF693ZJX.bam  551a0392cade2a0739fbbfb6579bae60   \n",
       "ENCSR943LIB  ENCFF167AFA.bam  0ebb6f86c63122beef1716a993066a30   \n",
       "ENCSR945GUR  ENCFF337TOV.bam  e9f206bf795127cedcc20900063f0e92   \n",
       "ENCSR947OIM  ENCFF016WOB.bam  dcb2ead998c0a39ea2746e31ce2ea81f   \n",
       "ENCSR952QDQ  ENCFF605WVQ.bam  626acee742c5fd5385d7918cfa4dfcc3   \n",
       "ENCSR952RRH  ENCFF342OBS.bam  5b03bc17a8128e52cc838f4ab16ee1d7   \n",
       "ENCSR954HAY  ENCFF302IKH.bam  506e2e2e6fc378bcfe5fe383ddfcd1ef   \n",
       "ENCSR958NDU  ENCFF062BJD.bam  d1a4d652cf79fc5da7c6a9d552e37d46   \n",
       "ENCSR961YAG  ENCFF067HJU.bam  01ea66f0af664f22acfc2829aa67701c   \n",
       "ENCSR963RLK  ENCFF125TRY.bam  178541d6beb54ffa9520bff37edfe20e   \n",
       "ENCSR967QNT  ENCFF122EWD.bam  a0d8ac21a28fbfae3993b78e4d7c3d0d   \n",
       "ENCSR973QSV  ENCFF509JMI.bam  3a51c0f50a8a50b2c15f99d4ed0a78a4   \n",
       "ENCSR978CSQ  ENCFF403CZS.bam  b4081ded594504582272fe1795d35898   \n",
       "ENCSR984CLJ  ENCFF079PJT.bam  17f954cb3b25aafc2441baab7b65f5c9   \n",
       "\n",
       "                   expt_rep2                  expt_rep2_md5sum      name  \n",
       "ENCSR000KYM  ENCFF243AMZ.bam  dea667416a4d4c3c582637c2f9ad0f19     DDX3X  \n",
       "ENCSR000YYN  ENCFF065KHB.bam  84a492145ce5bf8f806aa2b7f9b30765     AKAP8  \n",
       "ENCSR004RGI  ENCFF459YMO.bam  22e6c3066888702577d180c635d0280b     RPS10  \n",
       "ENCSR007XKL  ENCFF466OZP.bam  c69b58950871a86055b9f24744add8c6      NFX1  \n",
       "ENCSR023HWI  ENCFF738SSD.bam  83fe79368dde05a9405107b333b0eca3   KHDRBS1  \n",
       "ENCSR029LGJ  ENCFF402FXS.bam  868de7091e2b4ab872cb07180b6c219f     DDX51  \n",
       "ENCSR034VBA  ENCFF836HER.bam  259850b04aa20c46a845e7e745076957  HNRNPUL1  \n",
       "ENCSR040FSN  ENCFF240PBV.bam  d45cbba792a437e4c1322bbc80e6b8ff    EIF4G2  \n",
       "ENCSR047AJA  ENCFF252MOZ.bam  08df33b474e55992dbc6fb89ce3daba2      SUB1  \n",
       "ENCSR047EEG  ENCFF710FGW.bam  09011dd591c7fe9e2ad6466e3d60ac9d    PABPC4  \n",
       "ENCSR047IUS  ENCFF854ZQF.bam  c7a234d56402a5cd82953e3b7285d4ab    HNRNPU  \n",
       "ENCSR047QHX  ENCFF425GWW.bam  517170c8b8efa57ed5a5b9ab9802585b     SF3B1  \n",
       "ENCSR048BWH  ENCFF244AJP.bam  8b4d6f5021dc6fc28ae4da13c97377e6   HNRNPA1  \n",
       "ENCSR056QEW  ENCFF815KLH.bam  b6406e485ea7fb9b21bed157d97cd210     CIRBP  \n",
       "ENCSR060KRD  ENCFF405TFQ.bam  29bb04acb8619257da66e68ea5cd6fb1    TROVE2  \n",
       "ENCSR066VOO  ENCFF669MME.bam  52a68e458d86521b3f15d54ce74a3654     SRSF1  \n",
       "ENCSR067LLB  ENCFF626FFP.bam  30e1a2886acb7031ff042c2d376e4b2b     DDX24  \n",
       "ENCSR076PMZ  ENCFF988QUT.bam  7fb20b908c06eac4b34af7563f2bade9    EIF2S2  \n",
       "ENCSR077BPR  ENCFF419DWP.bam  e8fe4c2a2fac5c85dabbc1eb07f585e9      ESF1  \n",
       "ENCSR079IPT  ENCFF531JMM.bam  baedd04b8de408ee40d99f28dee2e4ee     TBRG4  \n",
       "ENCSR079LMZ  ENCFF874NAQ.bam  68d9918b88e22790ee626b2601fdc3ee   DNAJC21  \n",
       "ENCSR081XRA  ENCFF375HRX.bam  e50db65422799b49e2eaf26dcb9681a9     SF3B4  \n",
       "ENCSR082UWF  ENCFF939VXA.bam  cdbb13dc41e59d50aa5effff3cb07273     STIP1  \n",
       "ENCSR096WVE  ENCFF852UMT.bam  558655c8ece7f28bc4b7728b974b265f     EIF3H  \n",
       "ENCSR098NHI  ENCFF076NJV.bam  14b03ea13b2196ee05769d431e272d53     RPS19  \n",
       "ENCSR101OPF  ENCFF526SEE.bam  d8e15e813b02ad11cdfd81726012d3d4    SUCLG1  \n",
       "ENCSR112YTD  ENCFF569ZVG.bam  09b71e6bdab6ea5dd3718cd497c5f3df      SLBP  \n",
       "ENCSR113HRG  ENCFF792MXQ.bam  4e1d3690e8c460ab128df396ccb55715     SRSF9  \n",
       "ENCSR113PYX  ENCFF265PUP.bam  ddbb4e06422de2cceaae080a11272175     CNOT7  \n",
       "ENCSR117WLY  ENCFF811UAO.bam  55fb3e4b289256d647032fa41ddf5750    EFTUD2  \n",
       "...                      ...                               ...       ...  \n",
       "ENCSR874DVZ  ENCFF688VZD.bam  555d5b2c939b2491bc374c5c06650152     GLRX3  \n",
       "ENCSR874ZLI  ENCFF385UCY.bam  381dc67be69b9c4e944a342d58c00a4c   CCDC124  \n",
       "ENCSR885YOI  ENCFF789INJ.bam  aca38270657f2373e15bdfc12dc9559c     CSTF2  \n",
       "ENCSR891AXF  ENCFF406LTN.bam  354c8ae25bc271071c8d7175a8babf3c       SSB  \n",
       "ENCSR891DYO  ENCFF643FXI.bam  076be026f34af55f66abc8aae2a1bc91      PES1  \n",
       "ENCSR898OPN  ENCFF666NAH.bam  c89265c4d3d8240e68d3a0acc87750cb     RBM17  \n",
       "ENCSR902WSK  ENCFF630HEN.bam  c6dc5d4fb303736e1c8dc7b59eabc8ac     SSRP1  \n",
       "ENCSR904BCZ  ENCFF939XGF.bam  6d83c671780f8e41fd5f9c3f8d74bca2    RAVER1  \n",
       "ENCSR904CJQ  ENCFF318PFL.bam  5bcb90162ad3ef2ccc773fd8ec3a49d2     U2AF2  \n",
       "ENCSR906RHU  ENCFF142DJX.bam  adc8cbd7a2f18162909cc5505b79be1d     SRSF5  \n",
       "ENCSR907UTB  ENCFF304ZSG.bam  6c3fb51f0e4f57c547a03a23bef662e5    DAZAP1  \n",
       "ENCSR911DGK  ENCFF407PAK.bam  c4959ba18e96d36550067dcbe34f78e3    TFIP11  \n",
       "ENCSR916WOI  ENCFF529ZHW.bam  d063a40258df81222e50ecb4541edd09     TRA2A  \n",
       "ENCSR921KDS  ENCFF441MQQ.bam  0b0e42513f0ba5112c1985b48758df64      RCC2  \n",
       "ENCSR925RNE  ENCFF246KQB.bam  ec457fe555f6ec387e7a8c0fe9a1f3cc    SERBP1  \n",
       "ENCSR927TSP  ENCFF953IZH.bam  da69132c7ea9b318b233bf7be0761f00     TIAL1  \n",
       "ENCSR927XBT  ENCFF204APB.bam  a6b0a23a6f0d1e2123f63603d560f751    NUSAP1  \n",
       "ENCSR943LIB  ENCFF193SZZ.bam  4ff32e805f86a74f5568acfd9e4e1513  SNRNP200  \n",
       "ENCSR945GUR  ENCFF414BQP.bam  8bf7422814bfd2a5b02d2f93ea920fa3     NAA15  \n",
       "ENCSR947OIM  ENCFF535LRB.bam  c239ace2a2ffd38702a3b4dce7bcb008     RBM22  \n",
       "ENCSR952QDQ  ENCFF371TVY.bam  1480adf72f3817db30575c1620320c61    METAP2  \n",
       "ENCSR952RRH  ENCFF519XDG.bam  685a67ae00026fc05d7c666f4f064fa0   IGF2BP2  \n",
       "ENCSR954HAY  ENCFF488MVU.bam  d75cfc2683f51a0fe04253a7781d69a3     SART3  \n",
       "ENCSR958NDU  ENCFF836SGI.bam  fb47f84207a8625d73c8429b8cf04902      HLTF  \n",
       "ENCSR961YAG  ENCFF211TCF.bam  e86966f0ca77facf5a3e23a1e7e77cb1    EIF4A3  \n",
       "ENCSR963RLK  ENCFF236LSH.bam  5215d582d345555b249236d2b2dacb26  APOBEC3C  \n",
       "ENCSR967QNT  ENCFF152ZSQ.bam  a7174bf579eb5598c49c08c65e95508b     GPKOW  \n",
       "ENCSR973QSV  ENCFF437YII.bam  1f14fc46203bf78bb751a111db534b5b      AATF  \n",
       "ENCSR978CSQ  ENCFF085ZVA.bam  3bb1915140afe5dd27a729492f6cb158       PKM  \n",
       "ENCSR984CLJ  ENCFF144BCM.bam  87325cbdf7c098259f835f4826acc8cc     CCAR2  \n",
       "\n",
       "[224 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Run FeatureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "498\n",
      "508\n"
     ]
    }
   ],
   "source": [
    "all_k562_bams = set(\n",
    "    k562_bams['control_rep1'].append(\n",
    "        k562_bams['control_rep2'].append(\n",
    "            k562_bams['expt_rep1'].append(\n",
    "                k562_bams['expt_rep2']\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")\n",
    "all_hepg2_bams = set(\n",
    "    hepg2_bams['control_rep1'].append(\n",
    "        hepg2_bams['control_rep2'].append(\n",
    "            hepg2_bams['expt_rep1'].append(\n",
    "                hepg2_bams['expt_rep2']\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")\n",
    "print(len(all_k562_bams))\n",
    "print(len(all_hepg2_bams))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 224 tasks as an array-job.\n",
      "Wrote commands to /home/bay001/projects/encode/analysis/featureCounts/individual_expts/_k562.sh.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8000665[].tscc-mgr.local\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Submitted script to queue home-scrm.\n",
      " Job ID: 8000665\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8000667[].tscc-mgr.local\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "running 228 tasks as an array-job.\n",
      "Wrote commands to /home/bay001/projects/encode/analysis/featureCounts/individual_expts/_hepg2.sh.\n",
      "Submitted script to queue home-scrm.\n",
      " Job ID: 8000667\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPLP0_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF534AWV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF134QUL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF992YGD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF126RPY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SNRNP200_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF831XLU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF194QUZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DNAJC2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF582HPV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF601DMY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FXR1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF865GTN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF569MFV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HLTF_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF120YPA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF791RLG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SART3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF172AYP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF099GGP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/APOBEC3C_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF887JLA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF772ENG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AKAP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF426DHN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF108HHU.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NPM1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF590AIM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF698DFA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PCBP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF979AOI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF918BEM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PA2G4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF148JEE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF393QIH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NCBP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF974BIC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF664TKJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TRA2A_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF593MYQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF578EXH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CALR_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF379PKN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF304DEE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF495QHS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF418UCJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FAM120A_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF459FNF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF681TRC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPC_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF031QZG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF531AUK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TIA1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF690DUO.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF143PAE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ESF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF501BAP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF252NEG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PTBP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF155FIK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF527FPF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF516CJR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF562PGS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/G3BP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF012GYQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF433ERW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CCAR1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF772LMC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF711NDL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF654XYV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF900XSL.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ZRANB2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF095LYX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF563FGQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SMN1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF781TAZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF269ZGN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRSF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF707GZQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF223XHU.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ADAR_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF087IGP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF841CBS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/MARK2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF393QCL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF865DXQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF2S2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF137QNH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF084VUF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SAFB2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF205GZW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF434WKD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FIP1L1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF551ACF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF546TSA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GNB2L1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF426ZZV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF357QYT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DKC1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF265NMT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712RNK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPS3A_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF406FCK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF597LPT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/STAU1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF365QAH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF501MSW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX6_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF343WSX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF401NSP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SF3B4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF987GMJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF283MXG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF4G2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF125SRX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF723TPD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRFBP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715VOF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515ZFQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF275ZLW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF438XUW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPL_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF749XEN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF384ABD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRP68_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF544JIA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF402SKG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NUSAP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF492TDD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052IMO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPA1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF339VWZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF032FGV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/KIF1C_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF693JST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF729YOG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SLTM_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF077NLP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF899LRG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ASCC1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF074RKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF491BPO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PUM2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF547WSA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF476YIR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX27_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF754YET.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF495SBM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DAZAP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF208PUO.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF808EHG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX28_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF545CXX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF088AZS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM27_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF377BGQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF457JDA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CIRBP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF299XEY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF949KOV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CCAR2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF066MMV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF039KBZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HSPD1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF282FHV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF394ADM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/IGF2BP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF534AWV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF134QUL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326QKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF384APO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CKAP4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF860SKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF219MPL.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UTP18_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF115BQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715AGI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF576LMK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF219OWH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CNOT7_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF983DDC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF690GXX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SSB_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715VOF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515ZFQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF166COZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF196ESE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CDC40_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF843QOP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF432WNR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF900XKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF800BJN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SUPT6H_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF517QNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF120VMQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GTF2F1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF570WZG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF241EAJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PUS1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF377BWN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF087JNB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX24_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF940KMT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF202KRA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPU_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF556EJB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF935RCR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TRIM56_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF301ZPR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF669FDH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX19B_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF532JTC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF117UAS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF4B_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF501TAD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF165UVA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM34_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF292QAQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF813XMA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UPF2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF578YZK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF334NRZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM22_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715VOF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515ZFQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776RHY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF475XDR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/QKI_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF262GAC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610CID.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SBDS_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF433WPM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF488ZBV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/XRN2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF407VNE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF322LTW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPAB_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF821PAW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF366UBV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NAA15_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF880YEJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515TAC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ILF2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF943PEC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF157XBJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/U2AF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF882SMH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF811FWH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SF3A3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF115BQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715AGI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF291FDL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF373JTP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRSF3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF263ZLE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF808OJZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/BUD13_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF960YDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF101CXE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DNAJC21_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF698MTR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF531GWL.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM17_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF665YCW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF397HZZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX47_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF459DPA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF918YBN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF249GRD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF031QPS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FTO_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF623FEY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF309YKH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SND1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF758JDY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF283VKH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AUH_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715VOF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515ZFQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF271WBV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF085JXU.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPS10_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF196RKN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF964BRZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SSRP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF502YFG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF286DGD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UBE2L3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF872CVZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF442NSN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AATF_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF159JHT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF800WEJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TIAL1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF468DJE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF604SZR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PABPC4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF929OLP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF907GVB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NUP35_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF611WGJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF888WPC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF941QUJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF537EKO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TUFM_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF904ZSQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF997JOB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RRP9_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF943GWU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF819AEP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EEF2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF591NXT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF062AVO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/IGF2BP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF563UBV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF139QDN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/BCLAF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF254QYF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF775APS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX21_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF276ZGA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF419TNF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPS19_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF764OON.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF558WIP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPLL_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF049UJQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF846LPR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/MATR3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF387CQS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF878ELZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/YBX3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF118VJE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF121VWZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/XRCC6_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF624WSP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF707UZQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF4G1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF692EUM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF648QEW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ACO1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF995CGT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF681MQG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NKRF_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF230ATW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF292YKI.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ZNF622_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF115BQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715AGI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF012WHT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF674BBE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SLBP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628MBB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF123CNA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TARDBP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF246URB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF498VOW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PRPF6_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF216RVH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF121FVJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EWSR1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF811NAN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF584DDC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AGO1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF807YHN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF475KTX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/KRR1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF151QED.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF417WJR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AARS_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF604SNM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF255RNW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/BCCIP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF596JEB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF022PCS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RECQL_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF333ROV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF397KTR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TFIP11_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF788JGG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF132PEP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RAVER1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF226GRM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF488AVD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NUFIP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF301WWZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF102UOY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CPSF7_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF387KJD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF970UME.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EXOSC9_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF077TZR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF363NEI.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRSF9_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF222VIH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF879FTB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX59_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF201XQE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF277DTA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPF_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF534AWV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF134QUL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF876TOA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF269FXL.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM15_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF017OAA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF836ZPZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF598XVT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF418OZB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF843QOP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF432WNR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF934JKB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF264LLP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM25_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF349KHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF962XCB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ABCF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF015TIY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF596YKH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PPIG_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF329FDS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF965NWJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EFTUD2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF718GZU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF944XRD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/U2AF2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF992OHI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF057UPR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/LARP7_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF936OPH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF228WNK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NSUN2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF437CON.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF151XPE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HDGF_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF803TTT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF359JYW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SNRNP70_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF080EFD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF717XEO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PCBP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF540GDB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF182FFM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX3X_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF201FHG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF977MZX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FKBP4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF635BOE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF377GZT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NOL12_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF453VAW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF329LUC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF164JOP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF887DJY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NONO_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF743IBK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF905RMN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PUF60_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF450QLW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF584LUS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PKM_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF556NSH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF617SLZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPD_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF549PGZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF616QND.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPS2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF791MAQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF945DBT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GRSF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF993GPN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF017PWT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CPSF6_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF806WSX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF713OCI.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PHF6_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF229WSC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF421IAK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UCHL5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF509OED.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF447JFQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RCC2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF296JDF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF028BLN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UPF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF115BQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715AGI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF578QNC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF484REW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPUL1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF721EPM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF468MLK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CELF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF464STX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF082FOH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NIP7_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF652PYP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF035RWF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/MTPAP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF761AIG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF748NRY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPL23A_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF321SLT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF783ONX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/IGF2BP3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF536RKP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF930KJO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM47_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF594AMB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF121EPB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ATP5C1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF652MXA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF817IOG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FASTKD2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF144PMD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF792EZN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPA0_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF707FIH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF352GXN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FASTKD1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF150SWH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF545NCH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/XRCC5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF599PLA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF356YKU.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TBRG4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF088YZA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF557WZU.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/LARP4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF772LMC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF711NDL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF183BFP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF583GGM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PSIP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF099LRS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF863XOG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/MAGOH_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF809RUA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF629OTS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FUBP3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF435FCA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF222QRV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF974DGK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF642RQE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBM39_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF907LAX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF161VUQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RBFOX2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF946VPZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF347ERZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPA2B1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF575VSD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF273EDC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GEMIN5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF962VUU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF500VFR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF3G_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF430WMT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF614QOI.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/XPO5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF045PJY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF580HEE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRSF5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF341XGN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF946EQQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SFPQ_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF988VWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QHC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF153PUW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF474BUK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/KHDRBS1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF227TTE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF354CDR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF3D_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF526QIM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF121JBQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AKAP8L_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF823MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF579IAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF754BCQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF377DAR.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF103PVI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF519YTY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF290HIA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF243LRE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SUCLG1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF346DRN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF519DOO.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CSTF2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF116TYA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF387LNP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SERBP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF459VBK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF670IQK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PARN_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF859SVA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF185BCI.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SUGP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF384IPG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF716BPQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RPS5_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF438JSV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF321UEF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ETF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715VOF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF515ZFQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF611HBN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF298PYT.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SRSF7_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF534AWV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF134QUL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF757NZN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF220CPX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/KHSRP_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF624GLA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF461CJW.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GRWD1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF681EYY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF695NHA.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PPIL4_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF291VAA.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF178YXP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DHX30_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF460AAE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF058QDT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF180JBY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF502YPG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPK_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF589WIS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF033EZX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF2S1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF287DXE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF188MUD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/STIP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF562CPJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF132LSH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PNPT1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF881HFB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF447EEQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/LSM11_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF245MMG.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF101FLQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SF3B1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF686NSC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF911MWC.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/MSI2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF600ERV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF932BOB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FMR1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF035LSX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF045GQX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/RTF1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF341XOS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF872WLN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/UTP3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF115BQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF715AGI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF318MDE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF622ENH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PABPC1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF748GLV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF397CTG.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PES1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF374PNQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF959LXB.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX52_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF246WQT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF756RDZ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CSTF2T_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF383QRK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF712XJM.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF352MST.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF187JOD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/BOP1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF998VON.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF722HDV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/FUS_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF162PMY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF382DXD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/LIN28B_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF505HKR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF142SQX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF170MXT.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF732IMS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/CEBPZ_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF391FXZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF925SHM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/NELFE_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF269RNF.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF957GUJ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ILF3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF481PMW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF610JDV.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF415JGN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF603VPF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/G3BP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF035NBS.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF562NOS.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PUM1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF829TUN.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF350VVR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF656HFZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF005EIM.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TROVE2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF565GTX.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF099TTE.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/EIF4A3_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF052HTH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF857QSU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF224ALJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF037BCP.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/AKAP8_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF776ARB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF975SQY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF093STD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF285YDY.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DDX55_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF799DYJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF935XLF.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/GPKOW_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF231NKH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF893QAU.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF270TJD.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF235AOH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/DROSHA_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF843QOP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF432WNR.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF954PQI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF523OXV.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/METAP2_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF628UPI.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF561UZL.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF095WYC.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF682JKD.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/HNRNPM_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF124DWE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF774FAK.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF833UFJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF303DFN.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SUPV3L1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF192VFW.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF078QOX.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SMNDC1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF455QQB.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF360CYQ.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/SUB1_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF499UUZ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF326XKY.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF785OWH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF147AZH.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/PRPF8_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF038LUH.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF705LNP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF556JQE.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF082BNK.bam',\n",
       " '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts -s 2 --tmpDir /home/bay001/projects/encode/analysis/featureCounts/individual_expts/ -p -a /projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf -o /home/bay001/projects/encode/analysis/featureCounts/individual_expts/TAF15_hepg2.counts.txt /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF571ARP.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF755GKJ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF970XIQ.bam /projects/ps-yeolab3/encode/rnaseq/shrna_knockdown_graveley_tophat/ENCFF166NGR.bam']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 0\n",
    "jobs = ['_k562','_hepg2']\n",
    "# jobs = ['_hepg2']\n",
    "for bams in [k562_bams, hepg2_bams]:\n",
    "    cmds = []\n",
    "    for unused,row in bams.iterrows():\n",
    "        p = '/home/bay001/software/subread-1.5.1-Linux-x86_64/bin/featureCounts'\n",
    "        s = '/home/bay001/projects/encode/analysis/featureCounts/individual_expts/'\n",
    "        o = '/home/bay001/projects/encode/analysis/featureCounts/individual_expts/{}{}.counts.txt'.format(row['name'],jobs[i])\n",
    "        a = '/projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf'\n",
    "        ctrl_rep1 = exists_and_return_fullpath(row['control_rep1'],downloaded_dir)\n",
    "        ctrl_rep2 = exists_and_return_fullpath(row['control_rep2'],downloaded_dir)\n",
    "        expt_rep1 = exists_and_return_fullpath(row['expt_rep1'],downloaded_dir)\n",
    "        expt_rep2 = exists_and_return_fullpath(row['expt_rep2'],downloaded_dir)\n",
    "        cmd = '{} -s 2 --tmpDir {} -p -a {} -o {} {} {} {} {}'.format(\n",
    "            p, s, a, o, ctrl_rep1, ctrl_rep2, expt_rep1, expt_rep2\n",
    "        )\n",
    "        cmds.append(cmd)\n",
    "    qtools.Submitter(\n",
    "        cmds, jobs[i], array=True, nodes=1, ppn=1, walltime='4:00:00', submit=True, queue='home-scrm', sh=s+'{}.sh'.format(jobs[i])\n",
    "    )\n",
    "    i = i + 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annotation = '/projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf'\n",
    "k562_outfile = '/home/bay001/projects/encode/permanent_data/_k562_hg19_featurecounts_intersect_CLIP_RNASEQ.txt'\n",
    "hepg2_outfile = '/home/bay001/projects/encode/permanent_data/_hepg2_hg19_featurecounts_intersect_CLIP_RNASEQ.txt'\n",
    "\n",
    "hepg2_cmd = \"featureCounts -T 8 -a {} -s 2 -p -o {}\".format(annotation, hepg2_outfile)\n",
    "for bam in all_hepg2_bams:\n",
    "    hepg2_cmd = hepg2_cmd + '{} '.format(os.path.join(exists_and_return_fullpath(bam, downloaded_dir)))\n",
    "\"\"\"qtools.Submitter(\n",
    "    hepg2_cmd, 'k562_featureCounts', array=False, nodes=1, ppn=8, walltime='48:00:00', submit=True, queue='home-scrm'\n",
    ")\"\"\"\n",
    "\n",
    "k562_cmd = \"featureCounts -T 8 -a {} -s 2 -p -o {}\".format(annotation, k562_outfile)\n",
    "for bam in all_k562_bams:\n",
    "    k562_cmd = k562_cmd + '{} '.format(os.path.join(exists_and_return_fullpath(bam, downloaded_dir)))\n",
    "\"\"\"qtools.Submitter(\n",
    "    cmd, 'hepg2_featureCounts', array=False, nodes=1, ppn=8, walltime='48:00:00', submit=True, queue='home-scrm'\n",
    ")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# see what's going on...\n",
    "jobid = 7893144\n",
    "for i in range(0,86):\n",
    "    ! qpeek $jobid[$i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
